// @tibco-tcstk/tc-core-lib/esm2015/lib/models/tc-component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class TcComponent {
    /*
    xs	'(max-width: 599px)'
    sm	'(min-width: 600px) and (max-width: 959px)'
    md	'(min-width: 960px) and (max-width: 1279px)'
    lg	'(min-width: 1280px) and (max-width: 1919px)'
    xl	'(min-width: 1920px) and (max-width: 5000px)'
    ltSm	'(max-width: 599px)'
    ltMd	'(max-width: 959px)'
    ltLg	'(max-width: 1279px)'
    ltXl	'(max-width: 1919px)'
    gtXs	'(min-width: 600px)'
    gtSm	'(min-width: 960px)'
    gtMd	'(min-width: 1280px)'
    gtLg	'(min-width: 1920px)'
     */
    /**
     * @return {?}
     */
    get xs() {
        return this.width <= 599;
    }
    /**
     * @return {?}
     */
    get sm() {
        return (this.width >= 600 && this.width <= 1279);
    }
    /**
     * @return {?}
     */
    get lg() {
        return (this.width >= 1280 && this.width <= 1919);
    }
    /**
     * @return {?}
     */
    get xl() {
        return (this.width >= 1920);
    }
    /**
     * @return {?}
     */
    get ltSm() {
        return (this.width <= 599);
    }
    /**
     * @return {?}
     */
    get ltMd() {
        return (this.width <= 959);
    }
    /**
     * @return {?}
     */
    get ltLg() {
        return (this.width <= 1279);
    }
    /**
     * @return {?}
     */
    get ltXl() {
        return (this.width <= 1919);
    }
    /**
     * @return {?}
     */
    get gtXs() {
        return (this.width >= 600);
    }
    /**
     * @return {?}
     */
    get gtSm() {
        return (this.width >= 960);
    }
    /**
     * @return {?}
     */
    get gtMd() {
        return (this.width >= 1280);
    }
    /**
     * @return {?}
     */
    get gtLg() {
        return (this.width >= 1920);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    TcComponent.prototype.width;
}
