// @tibco-tcstk/tc-core-lib/esm2015/lib/models/tc-login.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/* Models used by login and authorization services */
export class Group {
    /**
     * @param {?} id
     * @param {?} type
     */
    constructor(id, type) {
        this.id = id;
        this.type = type;
    }
}
if (false) {
    /** @type {?} */
    Group.prototype.id;
    /** @type {?} */
    Group.prototype.type;
}
export class Claim {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Claim.prototype.email;
    /** @type {?} */
    Claim.prototype.firstName;
    /** @type {?} */
    Claim.prototype.id;
    /** @type {?} */
    Claim.prototype.lastName;
    /** @type {?} */
    Claim.prototype.sandboxes;
    /** @type {?} */
    Claim.prototype.subscriptionId;
    /** @type {?} */
    Claim.prototype.username;
    /** @type {?} */
    Claim.prototype.primaryProductionSandbox;
}
export class Sandbox {
    /**
     * @param {?} groups
     * @param {?} id
     * @param {?} type
     * @param {?} name
     * @param {?} subscriptionId
     * @param {?} ownerId
     */
    constructor(groups, id, type, name, subscriptionId, ownerId) {
        this.groups = groups;
        this.id = id;
        this.type = type;
        this.name = name;
        this.subscriptionId = subscriptionId;
        this.ownerId = ownerId;
    }
}
if (false) {
    /** @type {?} */
    Sandbox.prototype.groups;
    /** @type {?} */
    Sandbox.prototype.id;
    /** @type {?} */
    Sandbox.prototype.type;
    /** @type {?} */
    Sandbox.prototype.name;
    /** @type {?} */
    Sandbox.prototype.subscriptionId;
    /** @type {?} */
    Sandbox.prototype.ownerId;
}
export class SandboxList {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        (/** @type {?} */ (this)).sandboxes = [];
        Object.assign((/** @type {?} */ (this)).sandboxes, input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    SandboxList.prototype.sandboxes;
}
export class AccessToken {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    AccessToken.prototype.access_token;
    /** @type {?} */
    AccessToken.prototype.token_type;
    /** @type {?} */
    AccessToken.prototype.expires_in;
}
export class AuthInfo {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    AuthInfo.prototype.userName;
    /** @type {?} */
    AuthInfo.prototype.firstName;
    /** @type {?} */
    AuthInfo.prototype.lastName;
    /** @type {?} */
    AuthInfo.prototype.userId;
    /** @type {?} */
    AuthInfo.prototype.ts;
    /** @type {?} */
    AuthInfo.prototype.orgName;
}
export class Subscription {
    /**
     * @param {?} accountId
     * @param {?} accountDisplayName
     * @param {?} ownerName
     * @param {?} loggedInUserRole
     * @param {?} regions
     */
    constructor(accountId, accountDisplayName, ownerName, loggedInUserRole, regions) {
        this.accountId = accountId;
        this.accountDisplayName = accountDisplayName;
        this.ownerName = ownerName;
        this.loggedInUserRole = loggedInUserRole;
        this.regions = regions;
    }
}
if (false) {
    /** @type {?} */
    Subscription.prototype.accountId;
    /** @type {?} */
    Subscription.prototype.accountDisplayName;
    /** @type {?} */
    Subscription.prototype.ownerName;
    /** @type {?} */
    Subscription.prototype.loggedInUserRole;
    /** @type {?} */
    Subscription.prototype.regions;
}
export class Owner {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Owner.prototype.email;
    /** @type {?} */
    Owner.prototype.firstName;
    /** @type {?} */
    Owner.prototype.lastName;
    /** @type {?} */
    Owner.prototype.role;
}
export class AccountInfo {
}
if (false) {
    /** @type {?} */
    AccountInfo.prototype.accountDisplayName;
    /** @type {?} */
    AccountInfo.prototype.accountId;
    /** @type {?} */
    AccountInfo.prototype.loggedInUserRole;
    /** @type {?} */
    AccountInfo.prototype.ownerInfo;
    /** @type {?} */
    AccountInfo.prototype.regions;
}
export class AccountsInfo {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        (/** @type {?} */ (this)).accountInfos = [];
        Object.assign((/** @type {?} */ (this)).accountInfos, input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    AccountsInfo.prototype.accountInfos;
}
export class LoginPrefill {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    LoginPrefill.prototype.emailId;
    /** @type {?} */
    LoginPrefill.prototype.clientId;
}
