// @tibco-tcstk/tc-core-lib/esm5/lib/components/tibco-cloud-new-element/tibco-cloud-new-element.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
var TibcoCloudNewElementComponent = /** @class */ (function () {
    function TibcoCloudNewElementComponent(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        this.elementType = data.resourceType;
    }
    /**
     * @return {?}
     */
    TibcoCloudNewElementComponent.prototype.onCreate = /**
     * @return {?}
     */
    function () {
        this.dialogRef.close({ event: 'submit', id: this.id, name: this.name });
    };
    TibcoCloudNewElementComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tc-tibco-cloud-new-element',
                    template: "<div class=\"tcs-dialog-outer-div\" fxFill >\n    <mat-toolbar class=\"tcs-case-creation-dialog-toolbar\">\n        <div fxLayout=\"row\" fxLayoutAlign=\"space-between center\" fxFlex>\n            <span class=\"tcs-case-creation-dialog-toolbar-text\">Create New {{elementType}}</span>\n        </div>\n    </mat-toolbar>\n    <div mat-dialog-content class=\"tcs-case-creation-dialog\" fxLayout=\"row\" fxLayoutAlign=\"space-around none\">\n        <mat-form-field style=\"min-width: 200px\">\n            <input matInput placeholder=\"Id\" [(ngModel)]=\"id\">\n        </mat-form-field>\n        <mat-form-field style=\"min-width: 400px\">\n            <input matInput placeholder=\"Name\" [(ngModel)]=\"name\">\n        </mat-form-field>\n    </div>\n    <div class=\"tcs-spacer\"></div>\n    <mat-dialog-actions class=\"tcs-dialog-actions\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n        <button mat-button mat-flat-button color=\"secondary\">Cancel</button>\n        <button mat-button mat-flat-button color=\"primary\" (click)=\"onCreate()\">Create</button>\n    </mat-dialog-actions>\n</div>",
                    styles: [".tcs-dialog-outer-div{height:100%;display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}"]
                }] }
    ];
    /** @nocollapse */
    TibcoCloudNewElementComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    return TibcoCloudNewElementComponent;
}());
export { TibcoCloudNewElementComponent };
if (false) {
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.id;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.name;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.elementType;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.dialogRef;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.data;
}
