// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-settings-roles/live-apps-settings-roles.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { TibcoCloudSettingsGeneralComponent, RoleAttribute, TcGeneralConfigService, TibcoCloudNewElementComponent } from '@tibco-tcstk/tc-core-lib';
import { ActivatedRoute } from '@angular/router';
import { MatSnackBar, MatDialog } from '@angular/material';
/**
 * Configuration of roles
 *
 * ![alt-text](../live-apps-settings-roles.png "")
 *
 * \@example <tcla-live-apps-settings-roles></tcla-live-apps-settings-roles>
 */
var LiveAppsSettingsRolesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsSettingsRolesComponent, _super);
    function LiveAppsSettingsRolesComponent(route, generalConfigService, snackBar, dialog) {
        var _this = _super.call(this, route, generalConfigService, snackBar) || this;
        _this.route = route;
        _this.generalConfigService = generalConfigService;
        _this.snackBar = snackBar;
        _this.dialog = dialog;
        _this.deleteRoleFunction = (/**
         * @return {?}
         */
        function () {
            _this.generalConfig.roles = _this.generalConfig.roles.filter((/**
             * @param {?} element
             * @return {?}
             */
            function (element) { return element.id != _this.currentRole.id; }));
        });
        _this.selectedRole = (/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            _this.currentRole = role;
        });
        _this.createRoleFunction = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var dialogRef = _this.dialog.open(TibcoCloudNewElementComponent, {
                width: '50%',
                height: '30%',
                maxWidth: '100vw',
                maxHeight: '100vh',
                panelClass: 'tcs-style-dialog',
                data: { resourceType: 'Role' }
            });
            dialogRef.afterClosed().subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                if (result) {
                    /** @type {?} */
                    var newRole = new RoleAttribute().deserialize({ id: result.id, display: result.name, priority: 1 });
                    _this.generalConfig.roles.push(newRole);
                }
            }));
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsSettingsRolesComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.generalConfig = _super.prototype.getRoute.call(this).snapshot.data.generalConfigHolder;
        this.roles = _super.prototype.getRoute.call(this).snapshot.data.allRoles;
        this.groups = _super.prototype.getRoute.call(this).snapshot.data.allGroups;
        this.claims = _super.prototype.getRoute.call(this).snapshot.data.claims;
        this.sandboxId = Number(this.claims.primaryProductionSandbox.id).valueOf();
    };
    LiveAppsSettingsRolesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-settings-roles',
                    template: "<div fxLayout=\"column\" fxFill>\n    <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\" [headerText]=\"'Role Configuration'\">\n    </tc-tibco-cloud-widget-header>\n    <div fxFlex style=\"padding: 20px; overflow: hidden\" fxLayout=\"column\">\n        <div style=\"overflow: auto;\">\n            <div fxLayout=\"column\" fxFlex>\n                <p>You can configure roles by here:</p>\n                <br>\n                <!-- Role specific -->\n                <mat-accordion>\n                    <mat-expansion-panel *ngFor=\"let role of generalConfig.roles\" (opened)=\"selectedRole(role)\">\n                        <mat-expansion-panel-header>\n                            <mat-panel-title>{{role.display}}</mat-panel-title>\n                        </mat-expansion-panel-header>\n                        <div fxLayout=\"column\">\n                            <div fxLayoutGap=\"10px\">\n                                <mat-form-field fxFlex=\"20\" style=\"min-width: 100px;\">\n                                    <input matInput placeholder=\"Id\" [(ngModel)]=\"role.id\" readonly>\n                                </mat-form-field>\n                                <mat-form-field fxFlex=\"40\" style=\"min-width: 400px;\">\n                                    <input matInput placeholder=\"Display Name\" [(ngModel)]=\"role.display\">\n                                </mat-form-field>\n                                <mat-form-field fxFlex=\"40\" style=\"min-width: 400px;\">\n                                    <mat-select [disableOptionCentering]=\"true\" placeholder=\"Live Apps Group\" [(value)]=\"role.group\">\n                                        <mat-option *ngFor=\"let group of groups.groups\" [value]=\"group.name\">\n                                            <span matTooltip=\"{{group.description}}\" matTooltipShowDelay=\"1000\"\n                                                matTooltipPosition=\"below\">{{group.name}}</span>\n                                        </mat-option>\n                                    </mat-select>\n                                </mat-form-field>\n                            </div>\n                            <div fxLayoutGap=\"5px\">\n                                <mat-form-field fxFlex=\"20\">\n                                    <input matInput type=\"number\" min=\"0\" max=\"10\" placeholder=\"Priority\" [(ngModel)]=\"role.priority\">\n                                </mat-form-field>\n                                <mat-slide-toggle [(ngModel)]=\"role.configuration\">Configuration role</mat-slide-toggle>\n                            </div>\n                        </div>\n                    </mat-expansion-panel>\n                </mat-accordion>\n            </div>\n\n        </div>\n        <div fxFlex class=\"tcs-filler-panel\"></div>\n        <div fxLayout=\"row\" fxLayoutAlign=\"end end\" fxLayoutGap=\"10px\" style=\"min-height: 50px\">\n            <button mat-raised-button color=\"secondary\" [disabled]=\"!currentRole\"\n                (click)=\"deleteRoleFunction()\">Delete</button>\n            <button mat-raised-button color=\"secondary\" (click)=\"createRoleFunction()\">New</button>\n            <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n        </div>\n    </div>\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsSettingsRolesComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: TcGeneralConfigService },
        { type: MatSnackBar },
        { type: MatDialog }
    ]; };
    return LiveAppsSettingsRolesComponent;
}(TibcoCloudSettingsGeneralComponent));
export { LiveAppsSettingsRolesComponent };
if (false) {
    /** @type {?} */
    LiveAppsSettingsRolesComponent.prototype.roles;
    /** @type {?} */
    LiveAppsSettingsRolesComponent.prototype.groups;
    /** @type {?} */
    LiveAppsSettingsRolesComponent.prototype.currentRole;
    /** @type {?} */
    LiveAppsSettingsRolesComponent.prototype.deleteRoleFunction;
    /** @type {?} */
    LiveAppsSettingsRolesComponent.prototype.selectedRole;
    /** @type {?} */
    LiveAppsSettingsRolesComponent.prototype.createRoleFunction;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsRolesComponent.prototype.route;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsRolesComponent.prototype.generalConfigService;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsRolesComponent.prototype.snackBar;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsRolesComponent.prototype.dialog;
}
