// @tibco-tcstk/tc-core-lib/esm5/lib/components/tibco-cloud-setting-menu-entry/tibco-cloud-setting-menu-entry.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
/**
 * Renders the menu options for each config menu
 *
 * ![alt-text](../tibco-cloud-setting-menu-entry.png "Image")
 *
 * \@example <tc-tibco-cloud-setting-menu-entry></tc-tibco-cloud-setting-menu-entry>
 */
var TibcoCloudSettingMenuEntryComponent = /** @class */ (function () {
    function TibcoCloudSettingMenuEntryComponent() {
        /**
         * ~event configureOption : Option Clicked
         * ~payload string : emits Id of option selected in config main page (options)
         */
        this.configureOption = new EventEmitter();
    }
    /**
     * @param {?} id
     * @return {?}
     */
    TibcoCloudSettingMenuEntryComponent.prototype.viewButtonClick = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.configureOption.emit(id);
    };
    /**
     * @return {?}
     */
    TibcoCloudSettingMenuEntryComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    TibcoCloudSettingMenuEntryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tc-tibco-cloud-setting-menu-entry',
                    template: "<div fxLayout=\"row\" fxLayoutAlign=\"start start\">\n    <div class=\"padding-icon\">\n        <mat-icon svgIcon=\"{{icon}}\"></mat-icon>\n    </div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"start start\">\n        <div class=\"underlying-capability-text\">{{title}}</div>\n        <div class=\"starters-custom-layout-box-option\" *ngFor=\"let option of options\" (click)=\"viewButtonClick(title + ' ' + option)\">\n            <span class=\"underlying-capability-link\">{{option}}</span>\n        </div>\n    </div>\n</div>\n",
                    styles: [".starters-custom-layout-box-option:hover{cursor:pointer}.underlying-capability-text{font-size:14px;font-weight:600;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:.3px;color:#062e79}.underlying-capability-link{font-size:12px;font-weight:400;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:.3px;color:#0081cb;padding-bottom:20px}:host ::ng-deep .padding-icon{padding-right:15px}"]
                }] }
    ];
    /** @nocollapse */
    TibcoCloudSettingMenuEntryComponent.ctorParameters = function () { return []; };
    TibcoCloudSettingMenuEntryComponent.propDecorators = {
        icon: [{ type: Input }],
        title: [{ type: Input }],
        options: [{ type: Input }],
        configureOption: [{ type: Output }]
    };
    return TibcoCloudSettingMenuEntryComponent;
}());
export { TibcoCloudSettingMenuEntryComponent };
if (false) {
    /**
     * icon name (svg key - needs to be registered)
     * @type {?}
     */
    TibcoCloudSettingMenuEntryComponent.prototype.icon;
    /**
     * page title comes from config resolver
     * @type {?}
     */
    TibcoCloudSettingMenuEntryComponent.prototype.title;
    /**
     * RenderedFormComponent: (options from third party API). TibcoCloudSettingMenuEntryComponent -Check with MC - not sure we need to document this.
     * @type {?}
     */
    TibcoCloudSettingMenuEntryComponent.prototype.options;
    /**
     * ~event configureOption : Option Clicked
     * ~payload string : emits Id of option selected in config main page (options)
     * @type {?}
     */
    TibcoCloudSettingMenuEntryComponent.prototype.configureOption;
}
