// @tibco-tcstk/tc-core-lib/esm2015/lib/components/tibco-cloud-upload-dialog/tibco-cloud-upload-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, Inject, EventEmitter } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
export class TibcoCloudUploadDialogComponent {
    // public fileText: string;
    /**
     * @param {?} dialogRef
     * @param {?} data
     */
    constructor(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        this.fileevent = new EventEmitter();
        this.fileToUpload = undefined;
        this.description = undefined;
        this.uploadFile = (/**
         * @return {?}
         */
        () => {
            if (this.fileToUpload) {
                this.fileevent.emit({
                    file: this.fileToUpload,
                    description: this.description,
                    inputData: this.data
                });
                this.dialogRef.close();
            }
        });
    }
    /**
     * @param {?} description
     * @return {?}
     */
    setFileDescription(description) {
        this.description = description;
    }
    /**
     * @param {?} files
     * @return {?}
     */
    attachFile(files) {
        this.fileToUpload = files.item(0);
    }
    /**
     * @return {?}
     */
    onNoClick() {
        this.dialogRef.close();
    }
}
TibcoCloudUploadDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'tibco-cloud-upload-dialog',
                template: "<h1 mat-dialog-title>{{data.title}}</h1>\n<div mat-dialog-content class=\"tcs-state-icon-upload-dialog\">\n    <form #fileUploadForm=\"ngForm\" fxFlex fxLayout=\"column\">\n        <div fxLayout=\"row\" fxLayoutAlign=\"start space-between\">\n            <button type=\"button\" mat-raised-button (click)=\"fileInput.click()\">Choose File To Upload</button>\n            <div *ngIf=\"fileToUpload\">{{fileToUpload.name}}</div>\n            <!-- below input is hidden and enabled by above button -->\n            <input class=\"tcs-hidden-input\" (change)=\"attachFile($event.target.files)\" accept=\"{{data.allowedExtensions}}\" #fileInput\n                type=\"file\" id=\"file\">\n        </div>\n    </form>\n</div>\n<div mat-dialog-actions>\n    <button mat-button (click)=\"uploadFile()\" type=\"submit\" [disabled]=\"!fileToUpload\">Upload</button>\n    <button mat-button (click)=\"onNoClick()\">Cancel</button>\n</div>",
                styles: [".tcs-hidden-input{display:none}"]
            }] }
];
/** @nocollapse */
TibcoCloudUploadDialogComponent.ctorParameters = () => [
    { type: MatDialogRef },
    { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
TibcoCloudUploadDialogComponent.propDecorators = {
    fileevent: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TibcoCloudUploadDialogComponent.prototype.fileevent;
    /** @type {?} */
    TibcoCloudUploadDialogComponent.prototype.fileToUpload;
    /** @type {?} */
    TibcoCloudUploadDialogComponent.prototype.description;
    /** @type {?} */
    TibcoCloudUploadDialogComponent.prototype.uploadFile;
    /** @type {?} */
    TibcoCloudUploadDialogComponent.prototype.dialogRef;
    /** @type {?} */
    TibcoCloudUploadDialogComponent.prototype.data;
}
