// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-creator-dialog/live-apps-creator-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material';
import { CaseCreatorSelectionContext } from '../../models/tc-case-creator';
import { CaseRoute } from '../../models/liveappsdata';
/**
 * Dialog used on home screen to allow case creation (wraps other components)
 *
 * \@example <tcla-live-apps-creator-dialog></tcla-live-apps-creator-dialog>
 */
var LiveAppsCreatorDialogComponent = /** @class */ (function () {
    function LiveAppsCreatorDialogComponent(dialogRef, data) {
        var _this = this;
        this.dialogRef = dialogRef;
        this.data = data;
        this.handleCaseCreated = (/**
         * @param {?} createdCase
         * @return {?}
         */
        function (createdCase) {
            _this.createdCase = createdCase;
        });
        this.openCase = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var caseRoute = new CaseRoute().deserialize({ caseRef: _this.createdCase.caseReference, appId: _this.application.applicationId, typeId: _this.application.id });
            _this.dialogRef.close(caseRoute);
        });
        this.closeDialog = (/**
         * @return {?}
         */
        function () {
            _this.dialogRef.close();
        });
        this.application = data.application;
        this.sandboxId = data.sandboxId;
        this.initialData = data.initialData;
        this.customFormDefs = data.customFormDefs;
    }
    LiveAppsCreatorDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-creator-dialog',
                    template: "<div class=\"tcs-dialog-outer-div\">\n  <mat-toolbar class=\"tcs-case-creation-dialog-toolbar\">\n    <div fxLayout=\"row\" fxLayoutAlign=\"space-between center\" fxFlex>\n      <span class=\"tcs-case-creation-dialog-toolbar-text\">Create New {{application.applicationName}}</span>\n    </div>\n    <mat-icon fxLayout=\"column\" fxLayoutAlign=\"start end\" class=\"tcs-case-creation-dialog-icons\" [svgIcon]=\"'tcs-close-icon'\" (click)=\"closeDialog()\"></mat-icon>\n  </mat-toolbar>\n  <div mat-dialog-content class=\"tcs-case-creation-dialog\" fxLayout=\"column\">\n    <div *ngIf=\"!createdCase\" class=\"tcs-case-creation-dialog-details\">\n      <tcla-live-apps-creators [sandboxId]=\"sandboxId\" [applicationId]=\"application.applicationId\"\n                               [typeId]=\"application.id\" [dataOverride]=\"initialData\"\n                               [customFormDefs]=\"customFormDefs\"\n                               (caseCreated)=\"handleCaseCreated($event)\"></tcla-live-apps-creators>\n    </div>\n    <div *ngIf=\"createdCase\" class=\"tcs-case-created-confirmation-details\" fxLayout=\"row\" fxLayoutAlign=\"center start\">\n      <span class=\"tcs-case-created-confirmation-text\">New {{application.applicationName}}\n        case successfully submitted</span>\n      </div>\n  </div>\n  <div class=\"tcs-spacer\"></div>\n  <mat-dialog-actions *ngIf=\"createdCase\" class=\"tcs-dialog-actions\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n    <button mat-button mat-flat-button color=\"primary\" (click)=\"openCase()\">Open Case</button>\n    <button mat-button (click)=\"closeDialog()\">Close</button>\n  </mat-dialog-actions>\n</div>\n",
                    styles: [".tcs-dialog-outer-div{height:100%;display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column}.tcs-spacer{-webkit-box-flex:1;flex-grow:1}.tcs-case-creation-dialog-toolbar{padding-left:20px;padding-right:20px;background-color:#fff;border-bottom:1px solid #dcdcdc}.tcs-case-creation-dialog-toolbar-text{font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px}.tcs-case-creation-dialog-details{padding:20px}.tcs-case-created-confirmation-text{font-family:Source Sans Pro;font-size:16px;font-weight:400;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:normal;color:#b6b6b6}.tcs-case-creation-dialog-actions{padding:20px}:host ::ng-deep .tcs-dialog-actions.mat-dialog-actions{margin-bottom:0;padding:20px}.tcs-case-creation-dialog.mat-dialog-content{padding:0;margin:0}.tcs-case-created-confirmation-details{margin-top:20px}.tcs-case-creation-dialog-icons{height:16px;width:16px}.tcs-case-creation-dialog-icons:hover{cursor:pointer}:host ::ng-deep .tcs-case-creation-dialog-icons:hover .svg-content{fill:#0081cb}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCreatorDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: CaseCreatorSelectionContext, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    return LiveAppsCreatorDialogComponent;
}());
export { LiveAppsCreatorDialogComponent };
if (false) {
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.application;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.sandboxId;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.initialData;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.createdCase;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.customFormDefs;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.handleCaseCreated;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.openCase;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.closeDialog;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.dialogRef;
    /** @type {?} */
    LiveAppsCreatorDialogComponent.prototype.data;
}
