// angular6-json-schema-form/esm2015/lib/widget-library/widget-library.module.js
import * as tslib_1 from "tslib";
var WidgetLibraryModule_1;
import { BASIC_WIDGETS } from './index';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { JsonSchemaFormService } from '../json-schema-form.service';
import { NgModule } from '@angular/core';
import { OrderableDirective } from './orderable.directive';
let WidgetLibraryModule = WidgetLibraryModule_1 = class WidgetLibraryModule {
    static forRoot() {
        return {
            ngModule: WidgetLibraryModule_1,
            providers: [JsonSchemaFormService]
        };
    }
};
WidgetLibraryModule = WidgetLibraryModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [CommonModule, FormsModule, ReactiveFormsModule],
        declarations: [...BASIC_WIDGETS, OrderableDirective],
        exports: [...BASIC_WIDGETS, OrderableDirective],
        entryComponents: [...BASIC_WIDGETS],
        providers: [JsonSchemaFormService]
    })
], WidgetLibraryModule);
export { WidgetLibraryModule };
