// @tibco-tcstk/tc-liveapps-lib/esm5/lib/resolvers/all-roles.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { LiveAppsService } from '../services/live-apps.service';
import { GeneralConfigResolver, TcGeneralConfigService, TcSharedStateService } from '@tibco-tcstk/tc-core-lib';
import { map, switchMap } from 'rxjs/operators';
import { Roles } from '../models/tc-groups-data';
import { HttpClient } from '@angular/common/http';
import { Location } from '@angular/common';
var AllRolesResolver = /** @class */ (function () {
    function AllRolesResolver(sharedStateService, generalConfigService, http, liveapps, location, router) {
        this.sharedStateService = sharedStateService;
        this.generalConfigService = generalConfigService;
        this.http = http;
        this.liveapps = liveapps;
        this.location = location;
        this.router = router;
    }
    /**
     * @param {?} routeSnapshot
     * @return {?}
     */
    AllRolesResolver.prototype.resolve = /**
     * @param {?} routeSnapshot
     * @return {?}
     */
    function (routeSnapshot) {
        /** @type {?} */
        var claims$ = this.liveapps.getClaims()
            .pipe(map((/**
         * @param {?} claim
         * @return {?}
         */
        function (claim) {
            claim.sandboxes.forEach((/**
             * @param {?} sandbox
             * @return {?}
             */
            function (sandbox) {
                if (sandbox.type === 'Production') {
                    claim.primaryProductionSandbox = sandbox;
                }
            }));
            return claim;
        })));
        /** @type {?} */
        var generalConfigResolver = new GeneralConfigResolver(this.sharedStateService, this.generalConfigService, this.http, this.location, this.router);
        return claims$.pipe(switchMap((/**
         * @param {?} claiminfo
         * @return {?}
         */
        function (claiminfo) {
            generalConfigResolver.setSandbox(Number(claiminfo.primaryProductionSandbox.id));
            /** @type {?} */
            var generalConfig$ = generalConfigResolver.resolve().pipe(map((/**
             * @param {?} generalConfig
             * @return {?}
             */
            function (generalConfig) {
                return new Roles().deserialize({ roles: generalConfig.roles });
            })));
            return generalConfig$;
        })));
    };
    AllRolesResolver.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AllRolesResolver.ctorParameters = function () { return [
        { type: TcSharedStateService },
        { type: TcGeneralConfigService },
        { type: HttpClient },
        { type: LiveAppsService },
        { type: Location },
        { type: Router }
    ]; };
    return AllRolesResolver;
}());
export { AllRolesResolver };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AllRolesResolver.prototype.sharedStateService;
    /**
     * @type {?}
     * @private
     */
    AllRolesResolver.prototype.generalConfigService;
    /**
     * @type {?}
     * @private
     */
    AllRolesResolver.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AllRolesResolver.prototype.liveapps;
    /**
     * @type {?}
     * @private
     */
    AllRolesResolver.prototype.location;
    /**
     * @type {?}
     * @private
     */
    AllRolesResolver.prototype.router;
}
