// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-settings-forms/live-apps-settings-forms.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { LiveAppsConfig } from '../../models/tc-liveapps-config';
import { ActivatedRoute } from '@angular/router';
import { TcLiveAppsConfigService } from '../../services/tc-live-apps-config.service';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { MatSnackBar } from '@angular/material';
/**
 * Configuration of forms
 *
 * ![alt-text](../live-apps-settings-forms.png "")
 *
 * \@example <tcla-live-apps-settings-forms></tcla-live-apps-settings-forms>
 */
export class LiveAppsSettingsFormsComponent extends LiveAppsComponent {
    /**
     * @param {?} route
     * @param {?} liveAppsConfigService
     * @param {?} snackBar
     */
    constructor(route, liveAppsConfigService, snackBar) {
        super();
        this.route = route;
        this.liveAppsConfigService = liveAppsConfigService;
        this.snackBar = snackBar;
        this.runSaveFunction = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const liveAppsConfig = new LiveAppsConfig().deserialize({
                applicationIds: this.selectedAppIds,
                caseIconsFolderId: this.liveAppsConfig.caseIconsFolderId,
                documentAppId: this.liveAppsConfig.documentAppId,
                collaborationAppId: this.liveAppsConfig.documentAppId,
                legacyCreators: this.liveAppsConfig.legacyCreators,
                legacyActions: this.liveAppsConfig.legacyActions,
                legacyWorkitems: this.liveAppsConfig.legacyWorkitems,
                formsFramework: this.liveAppsConfig.formsFramework
            });
            this.liveAppsConfigService.updateLiveAppsConfig(this.sandboxId, this.generalConfig.uiAppId, liveAppsConfig, this.liveAppsConfig.id).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            result => {
                this.snackBar.open('Live Apps form configuration settings saved', 'OK', {
                    duration: 3000
                });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                this.snackBar.open('Error saving forms configuration', 'OK', {
                    duration: 3000
                });
            }));
        });
    }
    /**
     * @protected
     * @return {?}
     */
    getLiveAppsConfigService() {
        return this.liveAppsConfigService;
    }
    /**
     * @param {?} appIds
     * @return {?}
     */
    handleAppIdSelection(appIds) {
        console.log('Selected: ', appIds);
        this.selectedAppIds = appIds;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.generalConfig = this.route.snapshot.data.laConfigHolder.generalConfig;
        this.liveAppsConfig = this.route.snapshot.data.laConfigHolder.liveAppsConfig;
        this.claims = this.route.snapshot.data.claims;
        this.sandboxId = Number(this.claims.primaryProductionSandbox.id).valueOf();
        this.selectedAppIds = this.liveAppsConfig.applicationIds;
    }
}
LiveAppsSettingsFormsComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-settings-forms',
                template: "<div fxLayout=\"column\" class=\"tcs-live-apps-settings-pane\" fxFill>\n  <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\" [headerText]=\"'Live Apps Form Configuration'\">\n  </tc-tibco-cloud-widget-header>\n  <div fxFlex style=\"padding: 20px;\" fxLayout=\"column\">\n    <div fxLayout=\"row\" fxLayoutAlign=\"start start\" fxLayoutGap=\"5px\" style=\"height: 20px\">\n      <p class=\"tcs-settings-text\">Select framework for generated forms (only applies when NOT using legacy or custom forms):</p>\n    </div>\n    <div fxLayout=\"column\" style=\"padding: 10px;\">\n      <mat-radio-group [(ngModel)]=\"liveAppsConfig.formsFramework\" fxLayout=\"column\">\n        <mat-radio-button class=\"tcs-radio-label-content\" value=\"material-design\">Material Design</mat-radio-button>\n        <mat-radio-button class=\"tcs-radio-label-content\" value=\"bootstrap-4\">Bootstrap 4</mat-radio-button>\n      </mat-radio-group>\n    </div>\n    <div fxLayout=\"row\" fxLayoutAlign=\"start start\" fxLayoutGap=\"5px\" style=\"height: 20px\">\n      <p class=\"tcs-settings-text\">Select whether to use Live Apps Case Manager Forms for this application:</p>\n    </div>\n    <div fxLayout=\"column\" style=\"padding: 10px;\">\n      <mat-checkbox [(ngModel)]=\"liveAppsConfig.legacyCreators\">\n        <mat-label class=\"tcs-checkbox-label\">Creators</mat-label>\n      </mat-checkbox>\n      <mat-checkbox [(ngModel)]=\"liveAppsConfig.legacyActions\">\n        <mat-label class=\"tcs-checkbox-label\">Actions</mat-label>\n      </mat-checkbox>\n      <mat-checkbox [(ngModel)]=\"liveAppsConfig.legacyWorkitems\">\n        <mat-label class=\"tcs-checkbox-label\">Workitems</mat-label>\n      </mat-checkbox>\n    </div>\n    <div fxFlex class=\"tcs-filler-panel\"></div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n      <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n    </div>\n  </div>\n</div>\n",
                styles: [".tcs-card-config-widget-pane,.tcs-live-apps-summary-cards-settings{overflow:hidden}.tcs-checkbox-label,.tcs-settings-text{font-family:Source Sans Pro;font-size:14px;font-weight:400;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px}:host ::ng-deep .tcs-radio-label-content .mat-radio-label-content{font-family:Source Sans Pro;font-size:14px;font-weight:400;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px}"]
            }] }
];
/** @nocollapse */
LiveAppsSettingsFormsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: TcLiveAppsConfigService },
    { type: MatSnackBar }
];
if (false) {
    /** @type {?} */
    LiveAppsSettingsFormsComponent.prototype.sandboxId;
    /** @type {?} */
    LiveAppsSettingsFormsComponent.prototype.selectedAppIds;
    /** @type {?} */
    LiveAppsSettingsFormsComponent.prototype.liveAppsConfig;
    /** @type {?} */
    LiveAppsSettingsFormsComponent.prototype.generalConfig;
    /** @type {?} */
    LiveAppsSettingsFormsComponent.prototype.claims;
    /** @type {?} */
    LiveAppsSettingsFormsComponent.prototype.runSaveFunction;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsFormsComponent.prototype.route;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsFormsComponent.prototype.liveAppsConfigService;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsFormsComponent.prototype.snackBar;
}
