// @tibco-tcstk/tc-core-lib/esm2015/lib/resolvers/login-prefill.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/* Used to resolve the default login credentials (username/clientId) when using dev/local login route instead of tibco cloud */
/* Data is stored/retrieved from session storage */
import { Injectable } from '@angular/core';
import { LoginPrefill } from '../models/tc-login';
/** @type {?} */
export const EMAIL_ID_KEY = 'tcs-login-email-id';
/** @type {?} */
export const CLIENT_ID_KEY = 'tcs-login-client-id';
export class LoginPrefillResolver {
    constructor() {
        this.getClientId = (/**
         * @return {?}
         */
        () => {
            return localStorage.getItem(CLIENT_ID_KEY);
        });
        this.getEmail = (/**
         * @return {?}
         */
        () => {
            return localStorage.getItem(EMAIL_ID_KEY);
        });
    }
    /**
     * @return {?}
     */
    resolve() {
        return new LoginPrefill().deserialize({ emailId: this.getEmail(), clientId: this.getClientId() });
    }
}
LoginPrefillResolver.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LoginPrefillResolver.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    LoginPrefillResolver.prototype.getClientId;
    /**
     * @type {?}
     * @private
     */
    LoginPrefillResolver.prototype.getEmail;
}
