// @tibco-tcstk/tc-core-lib/esm2015/lib/components/tibco-cloud-widget-header/tibco-cloud-widget-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * Header bar for small widgets (like recent cases etc on home)
 *
 * ![alt-text](../tibco-cloud-widget-header.png "Image")
 *
 * \@example <tc-tibco-cloud-widget-header></tc-tibco-cloud-widget-header>
 */
export class TibcoCloudWidgetHeaderComponent {
    constructor() {
        /**
         * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
         */
        this.showHeader = this.showHeader ? this.showHeader : true;
        /**
         * ~event toolbarButtonEvent : Button Clicked
         * ~payload string : tells caller a button was clicked - outputs button Id
         */
        this.toolbarButtonEvent = new EventEmitter();
    }
    /**
     * @param {?} id
     * @return {?}
     */
    toolbarButtonClick(id) {
        this.toolbarButtonEvent.emit(id);
    }
}
TibcoCloudWidgetHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'tc-tibco-cloud-widget-header',
                template: "<div *ngIf=\"showHeader\" class=\"tcs-widget-header\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n  <div fxLayoutAlign=\"start center\">\n    <mat-icon class=\"tcs-widget-header-icon\" svgIcon=\"{{icon}}\"></mat-icon>\n    <div class=\"tcs-widget-header-text\">{{headerText}}</div>\n  </div>\n  <div class=\"tcs-widget-toolbar\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n    <div class=\"tcs-widget-toolbar-button-div\" *ngFor=\"let toolbarButton of toolbarButtons\">\n      <mat-icon *ngIf=\"toolbarButton.visible\" (click)=\"toolbarButtonClick(toolbarButton.id)\" class=\"tcs-widget-toolbar-button\" [matTooltip]=\"toolbarButton.tooltip\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\" [svgIcon]=\"toolbarButton.icon\"></mat-icon>\n    </div>\n  </div>\n</div>\n\n<div *ngIf=\"!showHeader\" class=\"tcs-widget-toolbar\" style=\"margin-top: 5px; margin-bottom: 5px\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n  <div class=\"tcs-widget-toolbar-button-div\" *ngFor=\"let toolbarButton of toolbarButtons\">\n    <mat-icon *ngIf=\"toolbarButton.visible\" (click)=\"toolbarButtonClick(toolbarButton.id)\" class=\"tcs-widget-toolbar-button\" [matTooltip]=\"toolbarButton.tooltip\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\" [svgIcon]=\"toolbarButton.icon\"></mat-icon>\n  </div>\n</div>\n",
                styles: [".tcs-widget-header-icon{height:24px;width:24px}.tcs-widget-toolbar-button{height:20px;width:20px;margin-left:5px;margin-right:5px}:host ::ng-deep .tcs-widget-toolbar-button:hover path.svg-content{fill:#0081cb;cursor:pointer}:host ::ng-deep .tcs-widget-toolbar-button:hover g.svg-content{fill:#0081cb;cursor:pointer}.tcs-widget-header{height:40px;border-radius:3px 3px 0 0;box-shadow:0 1px 2px 0 #dedede;padding-left:20px;padding-right:20px;margin-bottom:5px}.tcs-widget-header-text{margin-left:10px;font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px}"]
            }] }
];
/** @nocollapse */
TibcoCloudWidgetHeaderComponent.ctorParameters = () => [];
TibcoCloudWidgetHeaderComponent.propDecorators = {
    headerText: [{ type: Input }],
    icon: [{ type: Input }],
    toolbarButtons: [{ type: Input }],
    showHeader: [{ type: Input }],
    toolbarButtonEvent: [{ type: Output }]
};
if (false) {
    /**
     * Text shown in menu bar
     * @type {?}
     */
    TibcoCloudWidgetHeaderComponent.prototype.headerText;
    /**
     * icon name (svg key - needs to be registered)
     * @type {?}
     */
    TibcoCloudWidgetHeaderComponent.prototype.icon;
    /**
     * buttons to display in the menu bar
     * @type {?}
     */
    TibcoCloudWidgetHeaderComponent.prototype.toolbarButtons;
    /**
     * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
     * @type {?}
     */
    TibcoCloudWidgetHeaderComponent.prototype.showHeader;
    /**
     * ~event toolbarButtonEvent : Button Clicked
     * ~payload string : tells caller a button was clicked - outputs button Id
     * @type {?}
     */
    TibcoCloudWidgetHeaderComponent.prototype.toolbarButtonEvent;
}
