// @tibco-tcstk/tc-liveapps-lib/esm5/lib/services/tc-roles-service.ts.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var TcRolesService = /** @class */ (function () {
    function TcRolesService() {
        var _this = this;
        this.currentRole = undefined;
        this.hasRoleAccess = (/**
         * @param {?} roleReq
         * @param {?} roles
         * @return {?}
         */
        function (roleReq, roles) {
            /** @type {?} */
            var reqRole = roles.roles.find((/**
             * @param {?} role
             * @return {?}
             */
            function (role) {
                return (role.id === roleReq);
            }));
            return reqRole ? true : false;
        });
        this.hasButtonAccess = (/**
         * @param {?} buttonId
         * @param {?} access
         * @return {?}
         */
        function (buttonId, access) {
            return access.buttonIds.indexOf(buttonId) > -1;
        });
        this.amIConfigurator = (/**
         * @param {?} roles
         * @return {?}
         */
        function (roles) {
            /** @type {?} */
            var reqRole = roles.roles.find((/**
             * @param {?} role
             * @return {?}
             */
            function (role) {
                return (role.configuration === true);
            }));
            return reqRole ? true : false;
        });
        this.getCurrentRole = (/**
         * @return {?}
         */
        function () {
            return _this.currentRole;
        });
        this.setCurrentRole = (/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            _this.currentRole = role;
        });
    }
    /**
     * @param {?} roleId
     * @param {?} roles
     * @return {?}
     */
    TcRolesService.prototype.checkRole = /**
     * @param {?} roleId
     * @param {?} roles
     * @return {?}
     */
    function (roleId, roles) {
        return this.hasRoleAccess(roleId, roles);
    };
    /**
     * @param {?} buttonId
     * @param {?} access
     * @return {?}
     */
    TcRolesService.prototype.checkButton = /**
     * @param {?} buttonId
     * @param {?} access
     * @return {?}
     */
    function (buttonId, access) {
        return this.hasButtonAccess(buttonId, access);
    };
    TcRolesService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TcRolesService.ctorParameters = function () { return []; };
    /** @nocollapse */ TcRolesService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcRolesService_Factory() { return new TcRolesService(); }, token: TcRolesService, providedIn: "root" });
    return TcRolesService;
}());
export { TcRolesService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TcRolesService.prototype.currentRole;
    /**
     * @type {?}
     * @private
     */
    TcRolesService.prototype.hasRoleAccess;
    /**
     * @type {?}
     * @private
     */
    TcRolesService.prototype.hasButtonAccess;
    /** @type {?} */
    TcRolesService.prototype.amIConfigurator;
    /** @type {?} */
    TcRolesService.prototype.getCurrentRole;
    /** @type {?} */
    TcRolesService.prototype.setCurrentRole;
}
