// @tibco-tcstk/tc-core-lib/esm2015/lib/common/tc-core-topic-comm.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
export class MessageTopicService {
    constructor() {
        this.topics = [
            { name: 'sample.topic', subject: new BehaviorSubject('init') }
        ];
        console.log('Topic Message service Created...');
    }
    /**
     * @param {?} name
     * @param {?} message
     * @return {?}
     */
    sendMessage(name, message) {
        this.createTopicIfNotExists(name, { text: message });
        this.topics.filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x.name === name))[0].subject.next({ text: message });
    }
    /**
     * @param {?} name
     * @return {?}
     */
    clearMessages(name) {
        this.createTopicIfNotExists(name, { text: 'init' });
        // this.queues.filter(x => x.name === name)[0].subject.next('init');
    }
    /**
     * @param {?} name
     * @return {?}
     */
    getMessage(name) {
        this.createTopicIfNotExists(name, { text: 'init' });
        return this.topics.filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x.name === name))[0].subject.asObservable();
    }
    /**
     * @param {?} name
     * @param {?} init
     * @return {?}
     */
    createTopicIfNotExists(name, init) {
        /** @type {?} */
        let queueExist = false;
        for (const q of this.topics) {
            if (q.name === name) {
                // console.log('Queue Exist: ' + q.name);
                queueExist = true;
            }
        }
        if (!queueExist) {
            console.log('Creating Topic: ' + name);
            /** @type {?} */
            const tempTopic = { name: name, subject: new BehaviorSubject(init) };
            this.topics.push(tempTopic);
        }
    }
}
MessageTopicService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
MessageTopicService.ctorParameters = () => [];
/** @nocollapse */ MessageTopicService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MessageTopicService_Factory() { return new MessageTopicService(); }, token: MessageTopicService, providedIn: "root" });
if (false) {
    /** @type {?} */
    MessageTopicService.prototype.topics;
}
