// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-case-creator-widget/live-apps-case-creator-widget.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { CaseType } from '../../models/liveappsdata';
import { CustomFormDefs } from '@tibco-tcstk/tc-forms-lib';
/**
 * Wraps case creators component (high level component)
 *
 * \@example <tcla-live-apps-case-creator-widget></tcla-live-apps-case-creator-widget>
 */
var LiveAppsCaseCreatorWidgetComponent = /** @class */ (function () {
    function LiveAppsCaseCreatorWidgetComponent() {
        var _this = this;
        /**
         * ~event caseCreated : Case Creator started (process started)
         * ~payload ProcessId : ProcessId object output on case creation (details of process started)
         */
        this.caseCreated = new EventEmitter();
        // handle case created
        this.handleCaseCreation = (/**
         * @param {?} processId
         * @return {?}
         */
        function (processId) {
            _this.caseCreated.emit(processId);
        });
    }
    /**
     * @return {?}
     */
    LiveAppsCaseCreatorWidgetComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    LiveAppsCaseCreatorWidgetComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-case-creator-widget',
                    template: "<div>\n  <div fxLayout=\"row\" fxLayout=\"start stretch\">\n    <span style=\"width: 185px\">Select Case Creator: </span>\n    <tcla-live-apps-creators fxFlex [sandboxId]=\"sandboxId\" [applicationId]=\"application.id\" [typeId]=\"application.id\" (caseCreated)=\"handleCaseCreation($event)\"></tcla-live-apps-creators>\n  </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCaseCreatorWidgetComponent.ctorParameters = function () { return []; };
    LiveAppsCaseCreatorWidgetComponent.propDecorators = {
        application: [{ type: Input }],
        sandboxId: [{ type: Input }],
        uiAppId: [{ type: Input }],
        initialData: [{ type: Input }],
        customFormDefs: [{ type: Input }],
        caseCreated: [{ type: Output }]
    };
    return LiveAppsCaseCreatorWidgetComponent;
}());
export { LiveAppsCaseCreatorWidgetComponent };
if (false) {
    /**
     * CaseType model for the case type you want to run a case creator for (normally comes from an application selector component)
     * @type {?}
     */
    LiveAppsCaseCreatorWidgetComponent.prototype.application;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsCaseCreatorWidgetComponent.prototype.sandboxId;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsCaseCreatorWidgetComponent.prototype.uiAppId;
    /**
     * override the initial data for a case creator
     * @type {?}
     */
    LiveAppsCaseCreatorWidgetComponent.prototype.initialData;
    /**
     * Custom Form configuration file
     * @type {?}
     */
    LiveAppsCaseCreatorWidgetComponent.prototype.customFormDefs;
    /**
     * ~event caseCreated : Case Creator started (process started)
     * ~payload ProcessId : ProcessId object output on case creation (details of process started)
     * @type {?}
     */
    LiveAppsCaseCreatorWidgetComponent.prototype.caseCreated;
    /** @type {?} */
    LiveAppsCaseCreatorWidgetComponent.prototype.handleCaseCreation;
}
