// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-creators/live-apps-creators.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { LiveAppsCreatorSelectorComponent } from '../live-apps-creator-selector/live-apps-creator-selector.component';
import { CustomFormDefs } from '@tibco-tcstk/tc-forms-lib';
/**
 * Wraps case creator selection and execution of creator
 *
 * \@example <tcla-live-apps-creators></tcla-live-apps-creators>
 */
var LiveAppsCreatorsComponent = /** @class */ (function () {
    function LiveAppsCreatorsComponent() {
        var _this = this;
        /**
         * Enable legacy creators
         */
        this.legacyCreators = this.legacyCreators ? this.legacyCreators : false;
        /**
         * Allow override of forms framework
         * Options: bootstrap-4 or material-design
         */
        this.formsFramework = this.formsFramework ? this.formsFramework : 'material-design';
        /**
         * ~event caseCreated : Case Creator started (process started)
         * ~payload ProcessId : ProcessId object output on case creation (details of process started)
         */
        this.caseCreated = new EventEmitter();
        // handle form submit
        this.handleSubmit = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.caseCreated.emit(data);
            /*if (this.creatorSelector) {
              this.creatorSelector.reset();
            }*/
        });
        // handle case creator selection
        this.handleCreatorSelection = (/**
         * @param {?} process
         * @return {?}
         */
        function (process) {
            _this.selectedProcess = process;
        });
    }
    /**
     * @return {?}
     */
    LiveAppsCreatorsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        console.log(this.applicationId);
    };
    LiveAppsCreatorsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-creators',
                    template: "<div fxLayout=\"column\" fxFlexFill>\n  <!--tcla-live-apps-creator-selector #creatorSelector style=\"\" [sandboxId]=\"sandboxId\" [appId]=\"applicationId\" [typeId]=\"typeId\" (creatorSelection)=\"handleCreatorSelection($event)\"></tcla-live-apps-creator-selector-->\n  <tcla-live-apps-case-creators *ngIf=\"!selectedProcess\" #creatorSelector style=\"\" [sandboxId]=\"sandboxId\" [appId]=\"applicationId\" [typeId]=\"typeId\" (creatorClicked)=\"handleCreatorSelection($event)\"></tcla-live-apps-case-creators>\n  <tcla-live-apps-case-creator style=\"overflow: auto; height: 100%\" [formsFramework]=\"formsFramework\" [sandboxId]=\"sandboxId\" [applicationId]=\"applicationId\" [typeId]=\"typeId\" [process]=\"selectedProcess\" [dataOverride]=\"dataOverride\" [customFormDefs]=\"customFormDefs\" [legacyCreators]=\"legacyCreators\" (caseChanged)=\"handleSubmit($event)\"></tcla-live-apps-case-creator>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCreatorsComponent.ctorParameters = function () { return []; };
    LiveAppsCreatorsComponent.propDecorators = {
        creatorSelector: [{ type: ViewChild, args: ['creatorSelector', { static: false },] }],
        sandboxId: [{ type: Input }],
        applicationId: [{ type: Input }],
        typeId: [{ type: Input }],
        dataOverride: [{ type: Input }],
        customFormDefs: [{ type: Input }],
        legacyCreators: [{ type: Input }],
        formsFramework: [{ type: Input }],
        caseCreated: [{ type: Output }]
    };
    return LiveAppsCreatorsComponent;
}());
export { LiveAppsCreatorsComponent };
if (false) {
    /** @type {?} */
    LiveAppsCreatorsComponent.prototype.creatorSelector;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.sandboxId;
    /**
     * lA appId ??
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.applicationId;
    /**
     * The LA Application Type Id (generally 1)
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.typeId;
    /**
     * Data object that will be displayed on the form. Allows overriding over form data (eg. when selecting data in spotfire)
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.dataOverride;
    /**
     * Custom Form configuration file
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.customFormDefs;
    /**
     * Enable legacy creators
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.legacyCreators;
    /**
     * Allow override of forms framework
     * Options: bootstrap-4 or material-design
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.formsFramework;
    /**
     * ~event caseCreated : Case Creator started (process started)
     * ~payload ProcessId : ProcessId object output on case creation (details of process started)
     * @type {?}
     */
    LiveAppsCreatorsComponent.prototype.caseCreated;
    /** @type {?} */
    LiveAppsCreatorsComponent.prototype.selectedProcess;
    /** @type {?} */
    LiveAppsCreatorsComponent.prototype.handleSubmit;
    /** @type {?} */
    LiveAppsCreatorsComponent.prototype.handleCreatorSelection;
}
