// @tibco-tcstk/tc-core-lib/esm2015/lib/common/tc-core-queue-comm.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class MessageQueueService {
    constructor() {
        this.queues = [
            { name: 'sample.queue', subject: new Subject() }
        ];
        console.log('Queue Message service Created...');
    }
    /**
     * @param {?} name
     * @param {?} message
     * @return {?}
     */
    sendMessage(name, message) {
        this.createQueueIfNotExists(name);
        this.queues.filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x.name === name))[0].subject.next({ text: message });
    }
    /**
     * @param {?} name
     * @return {?}
     */
    clearMessages(name) {
        this.createQueueIfNotExists(name);
        this.queues.filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x.name === name))[0].subject.next();
    }
    /**
     * @param {?} name
     * @return {?}
     */
    getMessage(name) {
        this.createQueueIfNotExists(name);
        return this.queues.filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x.name === name))[0].subject.asObservable();
    }
    /**
     * @param {?} name
     * @return {?}
     */
    createQueueIfNotExists(name) {
        /** @type {?} */
        let queueExist = false;
        for (const q of this.queues) {
            if (q.name === name) {
                // console.log('Queue Exist: ' + q.name);
                queueExist = true;
            }
        }
        if (!queueExist) {
            console.log('Creating Queue: ' + name);
            /** @type {?} */
            const tempQueue = { name: name, subject: new Subject() };
            this.queues.push(tempQueue);
        }
    }
}
MessageQueueService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
MessageQueueService.ctorParameters = () => [];
/** @nocollapse */ MessageQueueService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MessageQueueService_Factory() { return new MessageQueueService(); }, token: MessageQueueService, providedIn: "root" });
if (false) {
    /** @type {?} */
    MessageQueueService.prototype.queues;
}
