// @tibco-tcstk/tc-core-lib/esm2015/lib/common/tc-core-common-functions.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HashLocationStrategy } from '@angular/common';
// @dynamic
export class TcCoreCommonFunctions {
    /**
     * @param {?} text
     * @return {?}
     */
    static escapeString(text) {
        return text.replace(/"/g, '\"');
    }
    /**
     * @param {?} size
     * @return {?}
     */
    static fileSizeToHuman(size) {
        /** @type {?} */
        const e = (Math.log(size) / Math.log(1e3)) | 0;
        return +(size / Math.pow(1e3, e)).toFixed(2) + ' ' + ('kMGTPEZY'[e - 1] || '') + 'B';
    }
    /**
     * @param {?} str
     * @return {?}
     */
    static camelCaseToWords(str) {
        /** @type {?} */
        const re = str.replace(/([a-z])([A-Z])/g, '$1 $2')
            .replace(/([A-Z])([a-z])/g, ' $1$2')
            .replace(/\ +/g, ' ');
        return re;
    }
    /**
     * @param {?} array
     * @return {?}
     */
    static convertToJSON(array) {
        /** @type {?} */
        const objArray = [];
        for (let i = 1; i < array.length; i++) {
            objArray[i - 1] = {};
            for (let k = 0; k < array[0].length && k < array[i].length; k++) {
                /** @type {?} */
                const key = array[0][k];
                objArray[i - 1][key] = array[i][k];
            }
        }
        return objArray;
    }
    /**
     * @param {?} array
     * @param {?} arrayLength
     * @return {?}
     */
    static transposeArray(array, arrayLength) {
        // console.log('transposing array) arrayLength: ' + arrayLength );
        /** @type {?} */
        const newArray = [];
        for (let i = 0; i < arrayLength; i++) {
            newArray.push([]);
        }
        for (let i = 0; i < array.length; i++) {
            for (let j = 0; j < arrayLength; j++) {
                newArray[j].push(array[i][j]);
            }
        }
        return newArray;
    }
}
TcCoreCommonFunctions.camelize = (/**
 * @param {?} str
 * @return {?}
 */
(str) => {
    /** @type {?} */
    let newStr = '';
    /** @type {?} */
    let newArr = [];
    if (str.indexOf('-') !== -1) {
        newArr = str.split('-');
        for (let i = 0; i < newArr.length; i++) {
            newArr[i] = newArr[i].charAt(0).toUpperCase() + newArr[i].substr(1);
        }
        newStr = newArr.join(' ');
    }
    return newStr;
});
TcCoreCommonFunctions.prepareUrlForNonStaticResource = (/**
 * @param {?} location
 * @param {?} str
 * @return {?}
 */
(location, str) => {
    // @ts-ignore
    if (location._platformStrategy instanceof HashLocationStrategy) {
        return str;
    }
    else {
        return str;
    }
});
TcCoreCommonFunctions.prepareUrlForStaticResource = (/**
 * @param {?} location
 * @param {?} str
 * @return {?}
 */
(location, str) => {
    // @ts-ignore
    if (location._platformStrategy instanceof HashLocationStrategy) {
        return str;
    }
    else {
        return location.prepareExternalUrl(str);
    }
});
if (false) {
    /** @type {?} */
    TcCoreCommonFunctions.camelize;
    /** @type {?} */
    TcCoreCommonFunctions.prepareUrlForNonStaticResource;
    /** @type {?} */
    TcCoreCommonFunctions.prepareUrlForStaticResource;
}
