// @tibco-tcstk/tc-forms-lib/esm5/lib/tc-forms-lib.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { MaterialDesignFrameworkModule } from 'angular6-json-schema-form';
import { Bootstrap4FrameworkModule } from 'angular6-json-schema-form';
import { RenderedFormComponent } from './components/rendered-form/rendered-form.component';
import { MatButtonModule, MatCardModule, MatCheckboxModule, MatDialogModule, MatFormFieldModule, MatIconModule, MatInputModule, MatListModule, MatMenuModule, MatOptionModule, MatSelectModule, MatToolbarModule, MatTooltipModule } from '@angular/material';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { BrowserModule } from '@angular/platform-browser';
var TcFormsLibModule = /** @class */ (function () {
    function TcFormsLibModule() {
    }
    TcFormsLibModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        RenderedFormComponent
                    ],
                    imports: [
                        BrowserModule,
                        BrowserAnimationsModule,
                        FlexLayoutModule,
                        FormsModule,
                        MatButtonModule,
                        MatCardModule,
                        MatCheckboxModule,
                        MatIconModule,
                        MatMenuModule,
                        MatSelectModule,
                        MatToolbarModule,
                        MaterialDesignFrameworkModule,
                        Bootstrap4FrameworkModule,
                        MatButtonModule,
                        MatCardModule,
                        MatCheckboxModule,
                        MatListModule,
                        MatFormFieldModule,
                        MatInputModule,
                        MatIconModule,
                        MatSelectModule,
                        MatOptionModule,
                        MatDialogModule,
                        MatMenuModule,
                        MatCardModule,
                        MatTooltipModule,
                        FormsModule,
                        FlexLayoutModule,
                        ScrollingModule,
                        FormsModule,
                        ReactiveFormsModule
                    ],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [
                        RenderedFormComponent
                    ],
                    entryComponents: []
                },] }
    ];
    return TcFormsLibModule;
}());
export { TcFormsLibModule };
