// angular6-json-schema-form/esm2015/lib/framework-library/no-framework/no-framework.module.js
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { Framework } from '../framework';
import { FrameworkLibraryService } from '../framework-library.service';
import { JsonSchemaFormService } from '../../json-schema-form.service';
import { NgModule } from '@angular/core';
import { NoFramework } from './no.framework';
import { NoFrameworkComponent } from './no-framework.component';
import { WidgetLibraryModule } from '../../widget-library/widget-library.module';
import { WidgetLibraryService } from '../../widget-library/widget-library.service';
// No framework - plain HTML controls (styles from form layout only)
let NoFrameworkModule = class NoFrameworkModule {
};
NoFrameworkModule = tslib_1.__decorate([
    NgModule({
        imports: [CommonModule, WidgetLibraryModule],
        declarations: [NoFrameworkComponent],
        exports: [NoFrameworkComponent],
        providers: [JsonSchemaFormService, FrameworkLibraryService, WidgetLibraryService,
            { provide: Framework, useClass: NoFramework, multi: true }
        ],
        entryComponents: [NoFrameworkComponent]
    })
], NoFrameworkModule);
export { NoFrameworkModule };
