// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-case-state-audit/live-apps-case-state-audit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { map, take, takeUntil } from 'rxjs/operators';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { TcCaseStatesService } from '../../services/tc-case-states.service';
/**
 * Provides case audit summary for states
 *
 * ![alt-text](../live-apps-case-state-audit.png "Image")
 *
 * \@example <tcla-live-apps-case-state-audit></tcla-live-apps-case-state-audit>
 */
var LiveAppsCaseStateAuditComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsCaseStateAuditComponent, _super);
    function LiveAppsCaseStateAuditComponent(caseStatesService) {
        var _this = _super.call(this) || this;
        _this.caseStatesService = caseStatesService;
        _this.refresh = (/**
         * @return {?}
         */
        function () {
            _this.caseStatesService.getCaseStateAuditWithTerminal(_this.caseRef, _this.sandboxId, _this.appId)
                .pipe(take(1), takeUntil(_this._destroyed$), map((/**
             * @param {?} auditeventlist
             * @return {?}
             */
            function (auditeventlist) {
                _this.auditEvents = auditeventlist.auditEvents;
                if (!_this.auditEvents || _this.auditEvents.length <= 0) {
                    console.error('Unable to create states audit view. Case Audit likely removed due to subscription retention period.');
                }
            }))).subscribe(null, (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { _this.errorMessage = 'Error retrieving case audit: ' + error.error.errorMsg; }));
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsCaseStateAuditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.refresh();
    };
    LiveAppsCaseStateAuditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-case-state-audit',
                    template: "<div class=\"tcs-state-audit-pane\" fxLayout=\"column\" fxFill style=\"overflow: hidden;\">\n  <tc-tibco-cloud-widget-header fxFlex=\"nogrow\" [headerText]=\"'States'\" [icon]=\"'tcs-case-state-audit-icon'\"></tc-tibco-cloud-widget-header>\n  <div *ngIf=\"auditEvents && auditEvents.length > 0\" class=\"tcs-state-audit-container\">\n    <div class=\"tcs-state-audit-details\" fxLayout=\"column\" fxLayoutAlign=\"start start\">\n      <div *ngFor=\"let state of auditEvents; first as isFirst; last as isLast\"\n           [ngClass]=\"isFirst ? 'tcs-mini-state-item-first' : (isLast ? 'tcs-mini-state-item-last' : 'tcs-mini-state-item-middle')\">\n        <div fxLayout=\"row\">\n          <mat-icon class=\"tcs-mini-state-icon\"\n                    [svgIcon]=\"isLast ? (state.isTerminal ? 'tcs-mini-state-terminal-completed' : 'tcs-mini-state-current') : 'tcs-mini-state-completed'\"></mat-icon>\n          <li class=\"tcs-mini-state-details\" fxLayout=\"column\" fxLayoutAlign=\"start start\"\n              ng-class=\"{'underline':smallContainer}\">\n            <div class=\"tcs-mini-state-primary-label\"><span>{{state.caseState.value}}</span></div>\n            <div class=\"tcs-mini-state-secondary-label\">\n              <span *ngIf=\"state.principalName\">By {{state.principalName.value}}</span>\n              <span class=\"tcs-mini-state-spanPadding\"> on {{state.creationTime.value | date: \"medium\"}}</span>\n            </div>\n          </li>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div *ngIf=\"!auditEvents || auditEvents.length <= 0\">\n    <span class=\"tc-no-audit-data\">Audit data not available</span>\n  </div>\n</div>\n",
                    styles: [".tcs-state-audit-pane{border-radius:3px;box-shadow:0 2px 4px 0 rgba(0,0,0,.33);background-color:#fff}.tcs-mini-state-item-middle{padding-top:30px;border-left:3px solid #062e79;margin-top:-10px}.tcs-mini-state-item-first{border-left:3px solid #062e79}.tcs-mini-state-item-last{margin-top:-10px;padding-top:30px;border-left:3px solid #062e79}.tcs-state-audit-container{height:100%;overflow-x:auto;padding:11px}.tc-no-audit-data{padding:20px;min-width:150px;font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;color:#000}.tcs-mini-state-item-last:after{content:'';display:-webkit-box;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;-webkit-box-align:center;align-items:center;position:relative;left:-46px;top:0;background-color:red;width:10px;height:0;margin-top:-12px}.tcs-mini-state-icon{position:relative;left:-13px}.tcs-state-audit-details{padding:10px 20px}.tcs-mini-state-details{margin-left:10px}.tcs-mini-state-primary-label{font-family:Source Sans Pro;font-size:14px;font-weight:600;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:.2px;color:#212121}.tcs-mini-state-secondary-label{font-family:Source Sans Pro;font-size:14px;font-weight:400;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:.2px;color:#b6b6b6}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCaseStateAuditComponent.ctorParameters = function () { return [
        { type: TcCaseStatesService }
    ]; };
    LiveAppsCaseStateAuditComponent.propDecorators = {
        caseRef: [{ type: Input }],
        sandboxId: [{ type: Input }],
        appId: [{ type: Input }]
    };
    return LiveAppsCaseStateAuditComponent;
}(LiveAppsComponent));
export { LiveAppsCaseStateAuditComponent };
if (false) {
    /**
     * The case reference
     * @type {?}
     */
    LiveAppsCaseStateAuditComponent.prototype.caseRef;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsCaseStateAuditComponent.prototype.sandboxId;
    /**
     * The LA Application Id
     * @type {?}
     */
    LiveAppsCaseStateAuditComponent.prototype.appId;
    /** @type {?} */
    LiveAppsCaseStateAuditComponent.prototype.auditEvents;
    /** @type {?} */
    LiveAppsCaseStateAuditComponent.prototype.errorMessage;
    /** @type {?} */
    LiveAppsCaseStateAuditComponent.prototype.refresh;
    /**
     * @type {?}
     * @private
     */
    LiveAppsCaseStateAuditComponent.prototype.caseStatesService;
}
