// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-applications/live-apps-applications.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { LiveAppsService } from '../../services/live-apps.service';
import { CaseType, CaseTypesList } from '../../models/liveappsdata';
import { take, takeUntil } from 'rxjs/operators';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
/**
 * Lists live apps applications in drop down list allowing selection of app
 *
 * ![alt-text](../live-apps-applications.png "Image")
 *
 * \@example <tcla-live-apps-applications></tcla-live-apps-applications>
 */
var LiveAppsApplicationsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsApplicationsComponent, _super);
    function LiveAppsApplicationsComponent(liveapps) {
        var _this = _super.call(this) || this;
        _this.liveapps = liveapps;
        /**
         * Application selected from dropdown (output)
         */
        _this.selectedApp = _this.selectedApp ? _this.selectedApp : new CaseType();
        /**
         * ~event selection : Value selected in child component
         * ~payload CaseType : type varies.  but is when something is selected in a drop down it is passed back to the caller
         */
        _this.selection = new EventEmitter();
        _this.applications = new CaseTypesList();
        _this.changeAppSelection = (/**
         * @param {?} appSelected
         * @return {?}
         */
        function (appSelected) {
            _this.selectedApp = _this.applications.casetypes.find((/**
             * @param {?} casetype
             * @return {?}
             */
            function (casetype) {
                return casetype.applicationId === appSelected.applicationId;
            }));
        });
        _this.selectApplication = (/**
         * @param {?} selectionEvent
         * @return {?}
         */
        function (selectionEvent) {
            _this.selectedApp = selectionEvent.source.value;
            _this.selection.emit(selectionEvent.source.value);
        });
        _this.refresh = (/**
         * @param {?} bypassCache
         * @return {?}
         */
        function (bypassCache) {
            _this.liveapps.getApplications(_this.sandboxId, _this.appIds, 100, bypassCache)
                .pipe(take(1), takeUntil(_this._destroyed$))
                .subscribe((/**
             * @param {?} applicationList
             * @return {?}
             */
            function (applicationList) {
                _this.applications = applicationList;
                if (_this.selectedApp.applicationId) {
                    _this.selectedApp = applicationList.casetypes.find((/**
                     * @param {?} casetype
                     * @return {?}
                     */
                    function (casetype) {
                        return casetype.applicationId === _this.selectedApp.applicationId;
                    }));
                    // this.selection.emit(this.selectedApp);
                }
                else 
                // select first as default
                if (applicationList.casetypes.length > 0 && _this.selectFirstApp) {
                    _this.selectedApp = applicationList.casetypes[0];
                    _this.selection.emit(applicationList.casetypes[0]);
                }
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { _this.errorMessage = 'Error retrieving applications: ' + error.error.errorMsg; }));
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsApplicationsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.refresh(false);
    };
    LiveAppsApplicationsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-applications',
                    template: "<div class=\"tcs-application-selector\" fxFlex style=\"margin-left: 10px;\">\n    <mat-label></mat-label>\n    <mat-select *ngIf=\"selectedApp\" class=\"tcs-application-selector-option\" [disableOptionCentering]=\"true\" placeholder=\"Application\" [(value)]=\"selectedApp\" (selectionChange)=\"selectApplication($event)\" matTooltip=\"{{selectedApp.applicationName}}\" matTooltipShowDelay=\"1000\" matTooltipPosition=\"below\">\n      <mat-option *ngFor=\"let application of applications.casetypes\" [value]=\"application\">\n        {{application.label | ellipsis : 20}}\n      </mat-option>\n    </mat-select>\n</div>\n\n\n\n<!--div *ngIf=\"errorMessage\">{{errorMessage}}</div>\n<div *ngIf=\"applications\" class=\"la-applications-div\" fxLayoutAlign=\"space-around start\">\n  <div class=\"la-applications-card\" fxLayout=\"column\">\n    <h3>Select application:</h3>\n    <div class=\"la-application-selection\">\n      <div class=\"la-available-sub\" (click)=\"selectApplication(application)\" *ngFor=\"let application of applications.casetypes\">\n        <div class=\"la-application-name\">{{application.applicationName}}</div>\n        <div fxLayout=\"column\" class=\"la-application-selection-dtl\" fxLayoutAlign=\"start start\">\n          <div>Application Id: {{application.applicationId}}</div>\n          <div>Type Id: {{application.id}}</div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div-->\n",
                    styles: ["::ng-deep .tcs-application-selector .mat-form-field-underline{display:none}:host ::ng-deep .mat-select-value{min-width:100px;max-width:100%;width:auto}.la-applications-div{margin-top:50px}.la-applications-card{width:500px}.la-available-sub:hover{box-shadow:0 7px 14px 0 rgba(0,0,0,.3);-webkit-transition:.2s ease-in-out;transition:.2s ease-in-out;cursor:pointer}.la-available-sub{border:1px solid #dedede;border-radius:4px;box-shadow:0 1px 4px 0 rgba(0,0,0,.1);padding:20px 31px 31px;margin-top:5px;margin-bottom:5px}.la-application-name{font-size:18px;font-weight:700}:host ::ng-deep .tcs-application-selector-option .mat-select-value span{font-family:Source Sans Pro;font-size:16px;font-weight:400;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;color:#0081cb}:host ::ng-deep .tcs-application-selector-option .mat-select-arrow-wrapper div{color:#0081cb}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsApplicationsComponent.ctorParameters = function () { return [
        { type: LiveAppsService }
    ]; };
    LiveAppsApplicationsComponent.propDecorators = {
        sandboxId: [{ type: Input }],
        appIds: [{ type: Input }],
        selectFirstApp: [{ type: Input }],
        selectedApp: [{ type: Input }],
        selection: [{ type: Output }]
    };
    return LiveAppsApplicationsComponent;
}(LiveAppsComponent));
export { LiveAppsApplicationsComponent };
if (false) {
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsApplicationsComponent.prototype.sandboxId;
    /**
     * The list of LA Application IDs you want to handle
     * @type {?}
     */
    LiveAppsApplicationsComponent.prototype.appIds;
    /**
     * Whether to auto select the first app in dropdown selector (eg search)
     * @type {?}
     */
    LiveAppsApplicationsComponent.prototype.selectFirstApp;
    /**
     * Application selected from dropdown (output)
     * @type {?}
     */
    LiveAppsApplicationsComponent.prototype.selectedApp;
    /**
     * ~event selection : Value selected in child component
     * ~payload CaseType : type varies.  but is when something is selected in a drop down it is passed back to the caller
     * @type {?}
     */
    LiveAppsApplicationsComponent.prototype.selection;
    /** @type {?} */
    LiveAppsApplicationsComponent.prototype.applications;
    /** @type {?} */
    LiveAppsApplicationsComponent.prototype.errorMessage;
    /** @type {?} */
    LiveAppsApplicationsComponent.prototype.changeAppSelection;
    /** @type {?} */
    LiveAppsApplicationsComponent.prototype.selectApplication;
    /** @type {?} */
    LiveAppsApplicationsComponent.prototype.refresh;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsApplicationsComponent.prototype.liveapps;
}
