// @tibco-tcstk/tc-core-lib/esm2015/lib/components/tibco-cloud-table/tibco-cloud-table-datasource.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataSource } from '@angular/cdk/collections';
import { map } from 'rxjs/operators';
import { of as observableOf, merge } from 'rxjs';
/**
 * Data source for the TibcoCloudTable view. This class should
 * encapsulate all logic for fetching and manipulating the displayed data
 * (including sorting, pagination, and filtering).
 */
export class TibcoCloudTableDataSource extends DataSource {
    /**
     * @param {?} paginator
     * @param {?} sort
     * @param {?} d
     */
    constructor(paginator, sort, d) {
        super();
        this.paginator = paginator;
        this.sort = sort;
        this.d = d;
        this.data = d;
        //this._updateChangeSubscription();
    }
    /**
     * Connect this data source to the table. The table will only update when
     * the returned stream emits new items.
     * @return {?} A stream of the items to be rendered.
     */
    connect() {
        // Combine everything that affects the rendered data into one update
        // stream for the data-table to consume.
        /** @type {?} */
        const dataMutations = [
            observableOf(this.data),
            this.paginator.page,
            this.sort.sortChange
        ];
        // Set the paginator's length
        this.paginator.length = this.data.length;
        return merge(...dataMutations).pipe(map((/**
         * @return {?}
         */
        () => {
            return this.getPagedData(this.getSortedData([...this.data]));
        })));
    }
    /**
     *  Called when the table is being destroyed. Use this function, to clean up
     * any open connections or free any held resources that were set up during connect.
     * @return {?}
     */
    disconnect() { }
    /**
     * Paginate the data (client-side). If you're using server-side pagination,
     * this would be replaced by requesting the appropriate data from the server.
     * @private
     * @param {?} data
     * @return {?}
     */
    getPagedData(data) {
        /** @type {?} */
        const startIndex = this.paginator.pageIndex * this.paginator.pageSize;
        return data.splice(startIndex, this.paginator.pageSize);
    }
    /**
     * Sort the data (client-side). If you're using server-side sorting,
     * this would be replaced by requesting the appropriate data from the server.
     * @private
     * @param {?} data
     * @return {?}
     */
    getSortedData(data) {
        if (!this.sort.active || this.sort.direction === '') {
            return data;
        }
        return data.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        (a, b) => {
            /** @type {?} */
            const isAsc = this.sort.direction === 'asc';
            // console.log('Sort ON: ' + this.sort.active + ' A Has Property: ' + a.hasOwnProperty(this.sort.active) + ' B Has Property: ' + b.hasOwnProperty(this.sort.active) ) ;
            if (a.hasOwnProperty(this.sort.active) && b.hasOwnProperty(this.sort.active)) {
                return compare(a[this.sort.active], b[this.sort.active], isAsc);
            }
            else {
                return 0;
            }
        }));
    }
}
if (false) {
    /** @type {?} */
    TibcoCloudTableDataSource.prototype.data;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudTableDataSource.prototype.paginator;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudTableDataSource.prototype.sort;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudTableDataSource.prototype.d;
}
/**
 * Simple sort comparator for example ID/Name columns (for client-side sorting).
 * @param {?} a
 * @param {?} b
 * @param {?} isAsc
 * @return {?}
 */
function compare(a, b, isAsc) {
    return (a < b ? -1 : 1) * (isAsc ? 1 : -1);
}
