// @tibco-tcstk/tc-core-lib/esm5/lib/common/tc-core-topic-comm.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
var MessageTopicService = /** @class */ (function () {
    function MessageTopicService() {
        this.topics = [
            { name: 'sample.topic', subject: new BehaviorSubject('init') }
        ];
        console.log('Topic Message service Created...');
    }
    /**
     * @param {?} name
     * @param {?} message
     * @return {?}
     */
    MessageTopicService.prototype.sendMessage = /**
     * @param {?} name
     * @param {?} message
     * @return {?}
     */
    function (name, message) {
        this.createTopicIfNotExists(name, { text: message });
        this.topics.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.name === name; }))[0].subject.next({ text: message });
    };
    /**
     * @param {?} name
     * @return {?}
     */
    MessageTopicService.prototype.clearMessages = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        this.createTopicIfNotExists(name, { text: 'init' });
        // this.queues.filter(x => x.name === name)[0].subject.next('init');
    };
    /**
     * @param {?} name
     * @return {?}
     */
    MessageTopicService.prototype.getMessage = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        this.createTopicIfNotExists(name, { text: 'init' });
        return this.topics.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.name === name; }))[0].subject.asObservable();
    };
    /**
     * @param {?} name
     * @param {?} init
     * @return {?}
     */
    MessageTopicService.prototype.createTopicIfNotExists = /**
     * @param {?} name
     * @param {?} init
     * @return {?}
     */
    function (name, init) {
        var e_1, _a;
        /** @type {?} */
        var queueExist = false;
        try {
            for (var _b = tslib_1.__values(this.topics), _c = _b.next(); !_c.done; _c = _b.next()) {
                var q = _c.value;
                if (q.name === name) {
                    // console.log('Queue Exist: ' + q.name);
                    queueExist = true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (!queueExist) {
            console.log('Creating Topic: ' + name);
            /** @type {?} */
            var tempTopic = { name: name, subject: new BehaviorSubject(init) };
            this.topics.push(tempTopic);
        }
    };
    MessageTopicService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    MessageTopicService.ctorParameters = function () { return []; };
    /** @nocollapse */ MessageTopicService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MessageTopicService_Factory() { return new MessageTopicService(); }, token: MessageTopicService, providedIn: "root" });
    return MessageTopicService;
}());
export { MessageTopicService };
if (false) {
    /** @type {?} */
    MessageTopicService.prototype.topics;
}
