// @tibco-tcstk/tc-core-lib/esm5/lib/common/tc-core-queue-comm.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
var MessageQueueService = /** @class */ (function () {
    function MessageQueueService() {
        this.queues = [
            { name: 'sample.queue', subject: new Subject() }
        ];
        console.log('Queue Message service Created...');
    }
    /**
     * @param {?} name
     * @param {?} message
     * @return {?}
     */
    MessageQueueService.prototype.sendMessage = /**
     * @param {?} name
     * @param {?} message
     * @return {?}
     */
    function (name, message) {
        this.createQueueIfNotExists(name);
        this.queues.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.name === name; }))[0].subject.next({ text: message });
    };
    /**
     * @param {?} name
     * @return {?}
     */
    MessageQueueService.prototype.clearMessages = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        this.createQueueIfNotExists(name);
        this.queues.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.name === name; }))[0].subject.next();
    };
    /**
     * @param {?} name
     * @return {?}
     */
    MessageQueueService.prototype.getMessage = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        this.createQueueIfNotExists(name);
        return this.queues.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x.name === name; }))[0].subject.asObservable();
    };
    /**
     * @param {?} name
     * @return {?}
     */
    MessageQueueService.prototype.createQueueIfNotExists = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        var e_1, _a;
        /** @type {?} */
        var queueExist = false;
        try {
            for (var _b = tslib_1.__values(this.queues), _c = _b.next(); !_c.done; _c = _b.next()) {
                var q = _c.value;
                if (q.name === name) {
                    // console.log('Queue Exist: ' + q.name);
                    queueExist = true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (!queueExist) {
            console.log('Creating Queue: ' + name);
            /** @type {?} */
            var tempQueue = { name: name, subject: new Subject() };
            this.queues.push(tempQueue);
        }
    };
    MessageQueueService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    MessageQueueService.ctorParameters = function () { return []; };
    /** @nocollapse */ MessageQueueService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function MessageQueueService_Factory() { return new MessageQueueService(); }, token: MessageQueueService, providedIn: "root" });
    return MessageQueueService;
}());
export { MessageQueueService };
if (false) {
    /** @type {?} */
    MessageQueueService.prototype.queues;
}
