// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/models/tc-case-processes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class LaProcessSelection {
    // Format of ref is <applicationName>.<applicationInternalName>.<processType>.<processName>
    /**
     * @param {?} type
     * @param {?} appSchema
     * @param {?} caseIdAttribute
     * @param {?} process
     * @param {?} ref
     * @param {?} caseReference
     */
    constructor(type, appSchema, caseIdAttribute, process, ref, caseReference) {
        this.type = type;
        this.appSchema = appSchema;
        this.caseIdAttribute = caseIdAttribute;
        this.process = process;
        this.ref = ref;
        this.caseReference = caseReference;
    }
}
if (false) {
    /** @type {?} */
    LaProcessSelection.prototype.type;
    /** @type {?} */
    LaProcessSelection.prototype.appSchema;
    /** @type {?} */
    LaProcessSelection.prototype.caseIdAttribute;
    /** @type {?} */
    LaProcessSelection.prototype.process;
    /** @type {?} */
    LaProcessSelection.prototype.ref;
    /** @type {?} */
    LaProcessSelection.prototype.caseReference;
}
