// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/models/tc-workitems.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DeserializableClass } from '@tibco-tcstk/tc-core-lib';
export class WiDistributionStrategy extends DeserializableClass {
}
if (false) {
    /** @type {?} */
    WiDistributionStrategy.prototype.stringKey;
}
export class WiScheduleStatus extends DeserializableClass {
}
if (false) {
    /** @type {?} */
    WiScheduleStatus.prototype.description;
    /** @type {?} */
    WiScheduleStatus.prototype.stringKey;
}
export class WiItemContext extends DeserializableClass {
}
if (false) {
    /** @type {?} */
    WiItemContext.prototype.activityId;
    /** @type {?} */
    WiItemContext.prototype.activityName;
    /** @type {?} */
    WiItemContext.prototype.appId;
    /** @type {?} */
    WiItemContext.prototype.appInstance;
    /** @type {?} */
    WiItemContext.prototype.appName;
    /** @type {?} */
    WiItemContext.prototype.caseRef;
}
export class WiFlags extends DeserializableClass {
}
if (false) {
    /** @type {?} */
    WiFlags.prototype.scheduleStatus;
}
export class WiHeader extends DeserializableClass {
}
if (false) {
    /** @type {?} */
    WiHeader.prototype.description;
    /** @type {?} */
    WiHeader.prototype.distributionStrategy;
    /** @type {?} */
    WiHeader.prototype.flags;
    /** @type {?} */
    WiHeader.prototype.itemContext;
    /** @type {?} */
    WiHeader.prototype.name;
    /** @type {?} */
    WiHeader.prototype.priority;
    /** @type {?} */
    WiHeader.prototype.startDate;
}
export class WiState extends DeserializableClass {
}
if (false) {
    /** @type {?} */
    WiState.prototype.description;
    /** @type {?} */
    WiState.prototype.stringKey;
}
export class Workitem extends DeserializableClass {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Workitem.prototype.header;
    /** @type {?} */
    Workitem.prototype.id;
    /** @type {?} */
    Workitem.prototype.state;
    /** @type {?} */
    Workitem.prototype.version;
    /** @type {?} */
    Workitem.prototype.visible;
}
