// @tibco-tcstk/tc-core-lib/esm2015/lib/interceptors/authentication-error-interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpErrorResponse, } from '@angular/common/http';
import { of } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { Router } from '@angular/router';
export class AuthErrorInterceptor {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
        this.TIBCO_CLOUD_DOMAIN = 'cloud.tibco.com';
        this.TIBCO_TEST_DOMAIN = 'tenant-integration.tcie.pro';
        this.TIBCO_DEV_DOMAIN = 'emea.tibco.com';
        this.TIBCO_CLOUD_LOGIN = 'https://account.cloud.tibco.com/idm/v1/login-saml?relayState=';
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        return next.handle(req).pipe(catchError((/**
         * @param {?} err
         * @return {?}
         */
        (err) => {
            if (err instanceof HttpErrorResponse) {
                if (err.status === 419) {
                    // timed out
                    // should only be used if not on tibco cloud (and hence using our login route)
                    // check if we are hosted on tibco.cloud.com
                    /** @type {?} */
                    const host = window.location.hostname.split('.');
                    /** @type {?} */
                    const hostDomain = host[host.length - 3] + '.' + host[host.length - 2] + '.' + host[host.length - 1];
                    if (hostDomain === this.TIBCO_CLOUD_DOMAIN || hostDomain === this.TIBCO_TEST_DOMAIN || hostDomain === this.TIBCO_DEV_DOMAIN) {
                        // redirect to Tibco Cloud Login
                        // supply base64 encoded redirect string
                        /** @type {?} */
                        const reauthUrl = btoa('{ "resumeURL":"' + window.location.href + '", "tenantId":"BPM" }');
                        console.log('redirecting to: ' + this.TIBCO_CLOUD_LOGIN + reauthUrl);
                        window.location.href = this.TIBCO_CLOUD_LOGIN + reauthUrl;
                    }
                    else {
                        this.router.navigate(['/login'], { queryParams: {} });
                    }
                }
                else {
                    throw err;
                }
            }
            return of(err);
        })));
    }
}
AuthErrorInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AuthErrorInterceptor.ctorParameters = () => [
    { type: Router }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_CLOUD_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_TEST_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_DEV_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_CLOUD_LOGIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.router;
}
