// @tibco-tcstk/tc-core-lib/esm5/lib/components/tibco-cloud-error/tibco-cloud-error.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
/**
 * Global Error Handling, contains generic Error Handling for
 *
 * - NO_ROLE :: You are not a Member of this Application, please contact the Application- or Subscription-Owner
 * - NO_ACCESS :: You are not entitled to access this Application, please contact the Application- or Subscription-Owner
 * - NO_ROUTE_ACCESS :: Sorry but you do not have the required role to access this page of the application, please contact the application owner
 * - NOT_ADMIN_INIT :: Before using this application a member of the TIBCO Live Apps Admin Group must login to set initial configuration
 */
/**
 * Exception Handling page
 *
 * \@example <tc-tibco-cloud-error></tc-tibco-cloud-error>
 */
var TibcoCloudErrorComponent = /** @class */ (function () {
    /**
     * todo: Add logger
     */
    function TibcoCloudErrorComponent(route) {
        var _this = this;
        this.route = route;
        this.knownErrorList = [
            {
                'errorCode': 'NO_ROLE',
                'errorTitle': 'Access Denied',
                'errorMessage': 'You are not a Member of this Application, please contact the Application- or Subscription-Owner.'
            },
            {
                'errorCode': 'NO_ACCESS',
                'errorTitle': 'Access Denied',
                'errorMessage': 'You are not entitled to access this Application, please contact the Application- or Subscription-Owner.'
            },
            {
                'errorCode': 'NO_ROUTE_ACCESS',
                'errorTitle': 'Access Denied',
                'errorMessage': 'Sorry but you do not have the required role to access this page of the application, please contact the application owner'
            },
            {
                'errorCode': 'NOT_ADMIN_INIT',
                'errorTitle': 'Admin configuration required',
                'errorMessage': 'Before using this application a member of the TIBCO Live Apps Admin Group must login to set initial configuration'
            }
        ];
        this.message = '';
        this.route.params.subscribe((/**
         * @param {?} params
         * @return {?}
         */
        function (params) {
            console.log(params); // log the entire params object
            console.log(params['errorCode']);
            console.log(params['errorData?']);
            if (params['errorCode'] != null) {
                _this.code = params['errorCode'];
                _this.title = _this.knownErrorList.find((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return x.errorCode === _this.code; })).errorTitle;
                _this.message = _this.knownErrorList.find((/**
                 * @param {?} x
                 * @return {?}
                 */
                function (x) { return x.errorCode === _this.code; })).errorMessage;
                _this.data = params['errorData'];
            }
            else {
                _this.code = 'Unknown Error Code';
            }
        }));
    }
    /**
     * @ignore
     */
    /**
     * @ignore
     * @return {?}
     */
    TibcoCloudErrorComponent.prototype.ngOnInit = /**
     * @ignore
     * @return {?}
     */
    function () {
    };
    TibcoCloudErrorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tc-tibco-cloud-error',
                    template: "<div fxFill fxLayout=\"row\" fxLayoutAlign=\"center center\">\n  <div fxFlex fxLayout=\"column\" fxLayoutAlign=\"start center\">\n    <mat-icon class=\"tcs-error-icon\" [svgIcon]=\"'ic-error-handler'\"></mat-icon>\n    <div style=\"margin-left: 30px; margin-right: 30px;\" fxLayout=\"column\" fxLayoutGap=\"10px\">\n      <span class=\"tcs-error-title\">{{title}}</span>\n      <span class=\"tcs-error-message\">{{message}}</span>\n      <span class=\"tcs-error-code\">[{{code}}] - {{data}}</span>\n    </div>\n  </div>\n</div>\n",
                    styles: [".tcs-error-icon{height:300px;width:300px}.tcs-error-icon svg{width:300px;height:300px}.tcs-error-title{font-family:\"Source Sans Pro\";font-size:40px}.tcs-error-message{font-family:\"Source Sans Pro\";font-size:16px;font-color:#212121}.tcs-error-code{font-family:\"Source Sans Pro\";font-size:16px;font-color:grey}"]
                }] }
    ];
    /** @nocollapse */
    TibcoCloudErrorComponent.ctorParameters = function () { return [
        { type: ActivatedRoute }
    ]; };
    return TibcoCloudErrorComponent;
}());
export { TibcoCloudErrorComponent };
if (false) {
    /** @type {?} */
    TibcoCloudErrorComponent.prototype.code;
    /** @type {?} */
    TibcoCloudErrorComponent.prototype.title;
    /** @type {?} */
    TibcoCloudErrorComponent.prototype.message;
    /** @type {?} */
    TibcoCloudErrorComponent.prototype.data;
    /** @type {?} */
    TibcoCloudErrorComponent.prototype.knownErrorList;
    /**
     * @type {?}
     * @protected
     */
    TibcoCloudErrorComponent.prototype.route;
}
