// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-case-data-display/live-apps-case-data-display.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { JsonSchema } from '../../models/liveappsdata';
import { CustomFormDefs } from '@tibco-tcstk/tc-forms-lib';
import { TcCaseProcessesService } from '../../services/tc-case-processes.service';
/**
 * Renders case data in a form
 *
 * ![alt-text](../live-apps-case-data-display.png "Image")
 *
 * \@example <tcla-live-apps-case-data-display></tcla-live-apps-case-data-display>
 */
var LiveAppsCaseDataDisplayComponent = /** @class */ (function () {
    function LiveAppsCaseDataDisplayComponent(caseProcessService) {
        this.caseProcessService = caseProcessService;
        this.DEFAULT_CASE_DATA_LAYOUT = [
            '*',
            { type: 'submit', title: 'Save', condition: '1===2' },
            { type: 'actions', title: 'hello', condition: '1===2' }
        ];
        /**
         * Allow override of forms framework
         * Options: bootstrap-4 or material-design
         */
        this.formsFramework = this.formsFramework ? this.formsFramework : 'material-design';
        /**
         * Layout object that can be passed to override default layout of the form renderer
         */
        this.layout = this.layout ? this.layout : this.DEFAULT_CASE_DATA_LAYOUT;
        /**
         * Emit event to cause refresh of page
         * *
         */
        this.refreshEvent = new EventEmitter();
    }
    /**
     * @param {?} data
     * @return {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.updateDataAction = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        var _a;
        data = (_a = {},
            _a[this.name] = data,
            _a);
        this.caseProcessService.caseDataFormUpdate(this.caseRef, this.appId, this.typeId, this.sandboxId, '$Update', data).subscribe((/**
         * @param {?} next
         * @return {?}
         */
        function (next) {
            _this.refreshEvent.emit();
        }), (/**
         * @param {?} error1
         * @return {?}
         */
        function (error1) {
            console.error('Unable to trigger update of case data');
        }));
    };
    /**
     * @return {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // since this is the 'display' of case data we set to readonly and no validation
        // legacy makes the labels appear above the data
        this.options = {
            defaultOptions: {
                'readonly': true,
                'appearance': 'legacy'
            },
            'validateOnRender': false
        };
        console.log(this.schema);
    };
    LiveAppsCaseDataDisplayComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-case-data-display',
                    template: "<tcfrm-rendered-form *ngIf=\"schema && caseData && options\" [schema]=\"schema\" [layout]=\"layout\" [data]=\"caseData\" [caseRef]=\"caseRef\" [appId]=\"appId\" [options]=\"options\" [customFormDefs]=\"customFormDefs\" [formsFramework]=\"formsFramework\" [formRef]=\"formRef\" (formSubmit)=\"updateDataAction($event)\"></tcfrm-rendered-form>\n",
                    styles: [":host ::ng-deep button{visibility:hidden}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCaseDataDisplayComponent.ctorParameters = function () { return [
        { type: TcCaseProcessesService }
    ]; };
    LiveAppsCaseDataDisplayComponent.propDecorators = {
        sandboxId: [{ type: Input }],
        caseData: [{ type: Input }],
        appId: [{ type: Input }],
        typeId: [{ type: Input }],
        caseRef: [{ type: Input }],
        name: [{ type: Input }],
        schema: [{ type: Input }],
        formsFramework: [{ type: Input }],
        layout: [{ type: Input }],
        customFormDefs: [{ type: Input }],
        formRef: [{ type: Input }],
        refreshEvent: [{ type: Output }]
    };
    return LiveAppsCaseDataDisplayComponent;
}());
export { LiveAppsCaseDataDisplayComponent };
if (false) {
    /** @type {?} */
    LiveAppsCaseDataDisplayComponent.prototype.DEFAULT_CASE_DATA_LAYOUT;
    /**
     * SandboxId
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.sandboxId;
    /**
     * The case data
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.caseData;
    /**
     * The case reference
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.appId;
    /**
     * The applicationId
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.typeId;
    /**
     * The typeId
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.caseRef;
    /**
     * The 'name' of the main case type
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.name;
    /**
     * The schema of the case type
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.schema;
    /**
     * Allow override of forms framework
     * Options: bootstrap-4 or material-design
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.formsFramework;
    /**
     * Layout object that can be passed to override default layout of the form renderer
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.layout;
    /**
     * Custom Form configuration file
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.customFormDefs;
    /**
     * Custom Form Reference
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.formRef;
    /**
     * Emit event to cause refresh of page
     * *
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.refreshEvent;
    /** @type {?} */
    LiveAppsCaseDataDisplayComponent.prototype.options;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsCaseDataDisplayComponent.prototype.caseProcessService;
}
