// @tibco-tcstk/tc-core-lib/esm2015/lib/services/tc-buttons-helper.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ToolbarButton } from '../models/tc-widget-header';
import * as i0 from "@angular/core";
export class TcButtonsHelperService {
    constructor() {
        this.createButton = (/**
         * @param {?} id
         * @param {?} icon
         * @param {?} filled
         * @param {?} tooltip
         * @param {?} visible
         * @param {?} enabled
         * @return {?}
         */
        (id, icon, filled, tooltip, visible, enabled) => {
            /** @type {?} */
            const newButton = new ToolbarButton().deserialize({
                id: id,
                icon: icon,
                filled: filled,
                tooltip: tooltip,
                visible: visible,
                enabled: enabled
            });
            return newButton;
        });
        this.updateButtons = (/**
         * @param {?} updatedToolbarButtons
         * @param {?} toolbarButtons
         * @return {?}
         */
        (updatedToolbarButtons, toolbarButtons) => {
            updatedToolbarButtons.forEach((/**
             * @param {?} updatedButton
             * @return {?}
             */
            updatedButton => {
                /** @type {?} */
                const idx = toolbarButtons.findIndex((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item.id === updatedButton.id));
                if (idx > -1) {
                    toolbarButtons[idx] = updatedButton;
                }
                else {
                    // no existing button so add it
                    toolbarButtons.push(updatedButton);
                }
            }));
            return toolbarButtons;
        });
        this.addButtons = (/**
         * @param {?} buttons
         * @param {?} toolbarButtons
         * @return {?}
         */
        (buttons, toolbarButtons) => {
            buttons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            button => {
                toolbarButtons.push(button);
            }));
            return toolbarButtons;
        });
        this.removeButtons = (/**
         * @param {?} buttons
         * @param {?} toolbarButtons
         * @return {?}
         */
        (buttons, toolbarButtons) => {
            buttons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            button => {
                /** @type {?} */
                const idx = toolbarButtons.findIndex((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item.id === button.id));
                toolbarButtons.splice(idx, 1);
            }));
            return toolbarButtons;
        });
    }
}
TcButtonsHelperService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TcButtonsHelperService.ctorParameters = () => [];
/** @nocollapse */ TcButtonsHelperService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcButtonsHelperService_Factory() { return new TcButtonsHelperService(); }, token: TcButtonsHelperService, providedIn: "root" });
if (false) {
    /** @type {?} */
    TcButtonsHelperService.prototype.createButton;
    /** @type {?} */
    TcButtonsHelperService.prototype.updateButtons;
    /** @type {?} */
    TcButtonsHelperService.prototype.addButtons;
    /** @type {?} */
    TcButtonsHelperService.prototype.removeButtons;
}
