// @tibco-tcstk/tc-core-lib/esm2015/lib/services/tc-general-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { SharedStateContent, SharedStateEntry, SharedStateList } from '../models/tc-shared-state';
import { map } from 'rxjs/operators';
import { Location } from '@angular/common';
import { TcCoreCommonFunctions } from '../common/tc-core-common-functions';
import { GeneralConfig } from '../models/tc-general-config';
import { TcSharedStateService } from './tc-shared-state.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "./tc-shared-state.service";
export class TcGeneralConfigService {
    /**
     * @param {?} location
     * @param {?} sharedStateService
     */
    constructor(location, sharedStateService) {
        this.location = location;
        this.sharedStateService = sharedStateService;
    }
    /**
     * @param {?} sandboxId
     * @param {?} uiAppId
     * @param {?} generalConfig
     * @return {?}
     */
    createGeneralConfig(sandboxId, uiAppId, generalConfig) {
        /** @type {?} */
        const ssName = uiAppId + '.general.config.tibcolabs.client.context.PUBLIC';
        /** @type {?} */
        const content = new SharedStateContent();
        content.json = TcCoreCommonFunctions.escapeString(JSON.stringify(generalConfig));
        return this.sharedStateService.createSharedState(ssName, 'PUBLIC', '', sandboxId, undefined, undefined, undefined, content)
            .pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        value => value)));
    }
    /**
     * @param {?} uiAppId
     * @param {?} useCache
     * @param {?} flushCache
     * @return {?}
     */
    getGeneralConfig(uiAppId, useCache, flushCache) {
        // if useCache is false this will trigger the service to update the cached version with latest
        /** @type {?} */
        const ssName = uiAppId + '.general.config.tibcolabs.client.context.PUBLIC';
        return this.sharedStateService.getSharedState(ssName, 'PUBLIC', useCache, flushCache)
            .pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            if (value.sharedStateEntries.length > 0) {
                /** @type {?} */
                const ssresult = new GeneralConfig().deserialize(JSON.parse(value.sharedStateEntries[0].content.json));
                ssresult.id = value.sharedStateEntries[0].id;
                return ssresult;
            }
            else {
                return undefined;
            }
        })));
    }
    /**
     * @param {?} sandboxId
     * @param {?} uiAppId
     * @param {?} generalConfig
     * @param {?} id
     * @return {?}
     */
    updateGeneralConfig(sandboxId, uiAppId, generalConfig, id) {
        /** @type {?} */
        const ssName = uiAppId + '.general.config.tibcolabs.client.context.PUBLIC';
        /** @type {?} */
        const content = new SharedStateContent();
        content.json = TcCoreCommonFunctions.escapeString(JSON.stringify(generalConfig));
        /** @type {?} */
        const entry = new SharedStateEntry();
        entry.content = content;
        entry.sandboxId = sandboxId;
        entry.name = ssName;
        entry.type = 'PUBLIC';
        entry.id = id;
        /** @type {?} */
        const ssList = new SharedStateList();
        ssList.sharedStateEntries = [];
        ssList.sharedStateEntries.push(entry);
        return this.sharedStateService.updateSharedState(ssList.sharedStateEntries)
            .pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            // flush the cache
            this.getGeneralConfig(uiAppId, true, true).subscribe();
            return new GeneralConfig().deserialize((JSON.parse(value.sharedStateEntries[0].content.json)));
        })));
    }
}
TcGeneralConfigService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TcGeneralConfigService.ctorParameters = () => [
    { type: Location },
    { type: TcSharedStateService }
];
/** @nocollapse */ TcGeneralConfigService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcGeneralConfigService_Factory() { return new TcGeneralConfigService(i0.ɵɵinject(i1.Location), i0.ɵɵinject(i2.TcSharedStateService)); }, token: TcGeneralConfigService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    TcGeneralConfigService.prototype.location;
    /**
     * @type {?}
     * @private
     */
    TcGeneralConfigService.prototype.sharedStateService;
}
