// @tibco-tcstk/tc-core-lib/esm2015/lib/services/tc-general-landing-page.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { TcGeneralLandingPageConfigService } from './tc-general-landing-page-config.service';
import * as i0 from "@angular/core";
import * as i1 from "./tc-general-landing-page-config.service";
export class TcGeneralLandingPageService {
    /**
     * @param {?} generalLandingPageConfigService
     */
    constructor(generalLandingPageConfigService) {
        this.generalLandingPageConfigService = generalLandingPageConfigService;
        this.getLandingPage = (/**
         * @param {?} key
         * @param {?} uiAppId
         * @return {?}
         */
        (key, uiAppId) => {
            /** @type {?} */
            const landingPage = this.generalLandingPageConfigService.getGeneralLandingPageConfig(uiAppId, true, true).pipe(map((/**
             * @param {?} landingPages
             * @return {?}
             */
            landingPages => {
                /** @type {?} */
                const page = landingPages.landingPage.filter((/**
                 * @param {?} element
                 * @return {?}
                 */
                element => element.key === key))[0];
                return page;
            })));
            return landingPage;
        });
        this.getLandingPageForRole = (/**
         * @param {?} roleId
         * @param {?} uiAppId
         * @return {?}
         */
        (roleId, uiAppId) => {
            return this.generalLandingPageConfigService.getGeneralLandingPageConfig(uiAppId, true, true).pipe(map((/**
             * @param {?} landingPages
             * @return {?}
             */
            landingPages => {
                /** @type {?} */
                const candidatePages = landingPages.landingPage.filter((/**
                 * @param {?} element
                 * @return {?}
                 */
                element => element.roles.indexOf(roleId) != -1));
                switch (candidatePages.length) {
                    case 0:
                        return undefined;
                    case 1:
                        return candidatePages[0];
                    default:
                        return candidatePages[0];
                }
            })));
        });
    }
}
TcGeneralLandingPageService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TcGeneralLandingPageService.ctorParameters = () => [
    { type: TcGeneralLandingPageConfigService }
];
/** @nocollapse */ TcGeneralLandingPageService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcGeneralLandingPageService_Factory() { return new TcGeneralLandingPageService(i0.ɵɵinject(i1.TcGeneralLandingPageConfigService)); }, token: TcGeneralLandingPageService, providedIn: "root" });
if (false) {
    /** @type {?} */
    TcGeneralLandingPageService.prototype.getLandingPage;
    /** @type {?} */
    TcGeneralLandingPageService.prototype.getLandingPageForRole;
    /**
     * @type {?}
     * @private
     */
    TcGeneralLandingPageService.prototype.generalLandingPageConfigService;
}
