// @tibco-tcstk/tc-core-lib/esm2015/lib/services/tc-login.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @ngdoc component
 * @name tcLoginService
 *
 * @description
 *
 * tcLoginService provides services for authenticating against Tibco Subscriber cloud and authorizing with the various
 * Tibco Cloud tenants such as liveapps.
 *
 * These services and related components are typically used when the UI is NOT running on Live Apps WRP/Tibco Cloud
 * When deployed to Live Apps WRP authentication is handled by Tibco Cloud when accessing the protected WRP URL
 *
 *
 */
import { Injectable } from '@angular/core';
import { AccessToken, AuthInfo } from '../models/tc-login';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { map, tap } from 'rxjs/operators';
import { Location } from '@angular/common';
import { EMAIL_ID_KEY, CLIENT_ID_KEY } from '../resolvers/login-prefill.resolver';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "@angular/common";
export class TcLoginService {
    /**
     * @param {?} http
     * @param {?} location
     */
    constructor(http, location) {
        this.http = http;
        this.location = location;
    }
    // Provide ability to login to Tibco Subscriber Cloud
    /**
     * @param {?} username
     * @param {?} password
     * @param {?} clientID
     * @return {?}
     */
    login(username, password, clientID) {
        localStorage.setItem(EMAIL_ID_KEY, username);
        localStorage.setItem(CLIENT_ID_KEY, clientID);
        /** @type {?} */
        const url = '/idm/v3/login-oauth';
        /** @type {?} */
        const body = new HttpParams()
            .set('Email', username)
            .set('Password', password)
            .set('TenantId', 'bpm')
            .set('ClientID', clientID);
        /** @type {?} */
        const headers = new HttpHeaders()
            .set('Content-Type', 'application/x-www-form-urlencoded');
        return this.http.post(url, body.toString(), { headers })
            .pipe(tap((/**
         * @param {?} val
         * @return {?}
         */
        val => sessionStorage.setItem('tcsTimestamp', Date.now().toString()))), map((/**
         * @param {?} authInfo
         * @return {?}
         */
        authInfo => new AuthInfo().deserialize(authInfo))));
    }
    /**
     * @param {?} username
     * @param {?} password
     * @return {?}
     */
    loginV2(username, password) {
        /** @type {?} */
        const url = '/as/token.oauth2';
        /** @type {?} */
        const body = new HttpParams()
            .set('username', username)
            .set('password', password)
            .set('client_id', 'ropc_ipass')
            .set('grant_type', 'password');
        /** @type {?} */
        const headers = new HttpHeaders()
            .set('Content-Type', 'application/x-www-form-urlencoded');
        return this.http.post(url, body.toString(), { headers })
            .pipe(tap((/**
         * @param {?} val
         * @return {?}
         */
        val => sessionStorage.setItem('tcsTimestamp', Date.now().toString()))), map((/**
         * @param {?} accessToken
         * @return {?}
         */
        accessToken => new AccessToken().deserialize(accessToken))));
    }
    // Provide ability to authorize against live apps (note tenantId: bpm)
    /**
     * @param {?} accessToken
     * @param {?} accountId
     * @return {?}
     */
    laAuthorize(accessToken, accountId) {
        /** @type {?} */
        const url = '/idm/v2/login-oauth';
        /** @type {?} */
        const body = new HttpParams()
            .set('AccessToken', accessToken.access_token)
            .set('TenantId', 'bpm')
            .set('AccountId', accountId);
        /** @type {?} */
        const headers = new HttpHeaders()
            .set('Content-Type', 'application/x-www-form-urlencoded');
        return this.http.post(url, body.toString(), { headers })
            .pipe(tap((/**
         * @param {?} val
         * @return {?}
         */
        val => sessionStorage.setItem('tcsTimestamp', Date.now().toString()))), map((/**
         * @param {?} authInfo
         * @return {?}
         */
        authInfo => new AuthInfo().deserialize(authInfo))));
    }
}
TcLoginService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TcLoginService.ctorParameters = () => [
    { type: HttpClient },
    { type: Location }
];
/** @nocollapse */ TcLoginService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcLoginService_Factory() { return new TcLoginService(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.Location)); }, token: TcLoginService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    TcLoginService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    TcLoginService.prototype.location;
}
