// @tibco-tcstk/tc-core-lib/esm5/lib/interceptors/authentication-error-interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpErrorResponse, } from '@angular/common/http';
import { of } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { Router } from '@angular/router';
var AuthErrorInterceptor = /** @class */ (function () {
    function AuthErrorInterceptor(router) {
        this.router = router;
        this.TIBCO_CLOUD_DOMAIN = 'cloud.tibco.com';
        this.TIBCO_TEST_DOMAIN = 'tenant-integration.tcie.pro';
        this.TIBCO_DEV_DOMAIN = 'emea.tibco.com';
        this.TIBCO_CLOUD_LOGIN = 'https://account.cloud.tibco.com/idm/v1/login-saml?relayState=';
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    AuthErrorInterceptor.prototype.intercept = /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    function (req, next) {
        var _this = this;
        return next.handle(req).pipe(catchError((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            if (err instanceof HttpErrorResponse) {
                if (err.status === 419) {
                    // timed out
                    // should only be used if not on tibco cloud (and hence using our login route)
                    // check if we are hosted on tibco.cloud.com
                    /** @type {?} */
                    var host = window.location.hostname.split('.');
                    /** @type {?} */
                    var hostDomain = host[host.length - 3] + '.' + host[host.length - 2] + '.' + host[host.length - 1];
                    if (hostDomain === _this.TIBCO_CLOUD_DOMAIN || hostDomain === _this.TIBCO_TEST_DOMAIN || hostDomain === _this.TIBCO_DEV_DOMAIN) {
                        // redirect to Tibco Cloud Login
                        // supply base64 encoded redirect string
                        /** @type {?} */
                        var reauthUrl = btoa('{ "resumeURL":"' + window.location.href + '", "tenantId":"BPM" }');
                        console.log('redirecting to: ' + _this.TIBCO_CLOUD_LOGIN + reauthUrl);
                        window.location.href = _this.TIBCO_CLOUD_LOGIN + reauthUrl;
                    }
                    else {
                        _this.router.navigate(['/login'], { queryParams: {} });
                    }
                }
                else {
                    throw err;
                }
            }
            return of(err);
        })));
    };
    AuthErrorInterceptor.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AuthErrorInterceptor.ctorParameters = function () { return [
        { type: Router }
    ]; };
    return AuthErrorInterceptor;
}());
export { AuthErrorInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_CLOUD_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_TEST_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_DEV_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.TIBCO_CLOUD_LOGIN;
    /**
     * @type {?}
     * @private
     */
    AuthErrorInterceptor.prototype.router;
}
