// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-settings-recent-cases/live-apps-settings-recent-cases.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { LiveAppsConfig } from '../../models/tc-liveapps-config';
import { LiveAppsSettingsComponent } from '../live-apps-settings/live-apps-settings.component';
/**
 * Allow Selection of live apps applications (part of config)
 *
 * ![alt-text](../live-apps-settings-recent-cases.png "")
 *
 * \@example <tcla-live-apps-settings-recent-cases></tcla-live-apps-settings-recent-cases>
 */
export class LiveAppsSettingsRecentCasesComponent extends LiveAppsSettingsComponent {
    constructor() {
        super(...arguments);
        this.LIVE_APPS_URL = '/apps/dt-app/index.html#/application-content/';
        this.runSaveFunction = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const liveAppsConfig = new LiveAppsConfig().deserialize({
                applicationIds: this.selectedAppIds,
                recentExcludedAppIds: this.excludeRecentIds,
                caseIconsFolderId: this.liveAppsConfig.caseIconsFolderId,
                documentAppId: this.liveAppsConfig.documentAppId,
                collaborationAppId: this.liveAppsConfig.documentAppId
            });
            // unable to use super.liveAppsConfigService
            super.getLiveAppsConfigService().updateLiveAppsConfig(this.sandboxId, this.generalConfig.uiAppId, liveAppsConfig, this.liveAppsConfig.id).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            result => {
                this.snackBar.open('Recent cases settings saved', 'OK', {
                    duration: 3000
                });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                this.snackBar.open('Error saving Recent cases settings', 'OK', {
                    duration: 3000
                });
            }));
        });
    }
    /**
     * @param {?} appIds
     * @return {?}
     */
    handleExcludeRecentAppIdSelection(appIds) {
        this.excludeRecentIds = appIds;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.excludeRecentIds = this.liveAppsConfig.recentExcludedAppIds;
    }
}
LiveAppsSettingsRecentCasesComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-settings-recent-cases',
                template: "<div fxLayout=\"column\" class=\"tcs-live-apps-settings-pane\" fxFill>\n  <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\" [headerText]=\"'Live Apps Recent Cases Configuration'\">\n</tc-tibco-cloud-widget-header>\n  <div fxFlex style=\"padding: 20px;\" fxLayout=\"column\">\n    <div fxLayout=\"row\" fxLayoutAlign=\"start start\" fxLayoutGap=\"5px\" style=\"height: 20px\">\n      <p>Select any TIBCO Cloud Live Apps Applications you wish to EXCLUDE from recent cases:</p>\n      <mat-icon class=\"tcs-designer-link\" [svgIcon]=\"'tcs-application-edit'\" (click)=\"handleEditLiveAppClick()\" matTooltip=\"Open Live Apps Designer\" matTooltipShowDelay=\"1000\" matTooltipPosition=\"below\"></mat-icon>\n    </div>\n    <tcla-live-apps-application-list class=\"tcs-app-selection-list-widget\" fxFlex=\"grow\" [sandboxId]=\"sandboxId\" [selectedAppIds]=\"excludeRecentIds\" (appIdsSelected)=\"handleExcludeRecentAppIdSelection($event)\"></tcla-live-apps-application-list>\n    <div fxFlex class=\"tcs-filler-panel\"></div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n      <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n    </div>\n  </div>\n</div>\n",
                styles: [".tcs-designer-link{margin-bottom:10px}.tcs-designer-link:hover{fill:#0081cb;cursor:pointer}.tcs-app-selection-list-widget,.tcs-live-apps-settings-pane{overflow:hidden}"]
            }] }
];
if (false) {
    /** @type {?} */
    LiveAppsSettingsRecentCasesComponent.prototype.LIVE_APPS_URL;
    /** @type {?} */
    LiveAppsSettingsRecentCasesComponent.prototype.excludeRecentIds;
    /** @type {?} */
    LiveAppsSettingsRecentCasesComponent.prototype.runSaveFunction;
}
