// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-milestone/live-apps-milestone.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { TcCaseStatesService } from '../../services/tc-case-states.service';
/**
 * Render a dynamic milestone SVG
 *
 * ![alt-text](../live-apps-milestone.png "Documents Component Image")
 *
 * \@example <tcla-live-apps-milestone></tcla-live-apps-milestone>
 */
var LiveAppsMilestoneComponent = /** @class */ (function () {
    function LiveAppsMilestoneComponent(caseStatesService) {
        this.caseStatesService = caseStatesService;
    }
    /**
     * @return {?}
     */
    LiveAppsMilestoneComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // work out which section the MS trailer is (first/middle/end)
        /** @type {?} */
        var section = this.isFirst ? 'first' : (this.isLast ? 'end' : 'middle');
        // construct the icon filename
        /** @type {?} */
        var svgFileName = section + '-section-' + this.status + '.svg';
        // work out correct class
        /** @type {?} */
        var bgClass;
        if (this.status === 'completed' && this.isTerminal) {
            bgClass = 'tc-ms-' + 'completed-terminal' + '-bg';
        }
        else {
            bgClass = 'tc-ms-' + this.status + '-bg';
        }
        this.milestoneSvg = 'tcs-milestone-' + this.status + (this.isTerminal ? '-terminal' : '');
        this.sectionSVG = this.caseStatesService.getMilestoneSectionSvg(this.label, 'tc-ms-label', bgClass, svgFileName);
    };
    LiveAppsMilestoneComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-milestone',
                    template: "<div class=\"tcs-milestone-block\" fxLayout=\"row\" fxLayoutAlign=\"center start\">\n  <mat-icon class=\"tc-ms-circle-icon\" [svgIcon]=\"milestoneSvg\"></mat-icon>\n  <div [innerHtml]=\"sectionSVG\" fxLayoutAlign=\"start center\"></div>\n</div>\n",
                    styles: [".tc-ms-circle-icon{height:36px;width:36px;margin-right:-8px;margin-left:-8px}.tcs-milestone-block{margin-bottom:1px;margin-top:1px}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsMilestoneComponent.ctorParameters = function () { return [
        { type: TcCaseStatesService }
    ]; };
    LiveAppsMilestoneComponent.propDecorators = {
        isFirst: [{ type: Input }],
        isLast: [{ type: Input }],
        status: [{ type: Input }],
        label: [{ type: Input }],
        isTerminal: [{ type: Input }],
        phase: [{ type: Input }],
        previousPhase: [{ type: Input }]
    };
    return LiveAppsMilestoneComponent;
}());
export { LiveAppsMilestoneComponent };
if (false) {
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.isFirst;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.isLast;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.status;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.label;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.isTerminal;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.phase;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.previousPhase;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.sectionSVG;
    /** @type {?} */
    LiveAppsMilestoneComponent.prototype.milestoneSvg;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsMilestoneComponent.prototype.caseStatesService;
}
