// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-case-list/live-apps-case-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { LiveAppsService } from '../../services/live-apps.service';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
/**
 * Renders list of cases for caserefs
 *
 * ![alt-text](../live-apps-case-list.png "Image")
 *
 * \@example <tcla-live-apps-case-list></tcla-live-apps-case-list>
 */
var LiveAppsCaseListComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsCaseListComponent, _super);
    function LiveAppsCaseListComponent(liveapps) {
        var _this = _super.call(this) || this;
        _this.liveapps = liveapps;
        /**
         * ~event clickCase : Case clicked
         * ~payload CaseRoute : CaseRoute object output when case is clicked so calling component can route accordingly - ie. route to case
         */
        _this.clickCase = new EventEmitter();
        /**
         * ~event clearMatches : Clear Matches button clicked
         *
         */
        _this.clearMatches = new EventEmitter();
        _this.clickCaseAction = (/**
         * @param {?} caseRoute
         * @return {?}
         */
        function (caseRoute) {
            _this.clickCase.emit(caseRoute);
        });
        _this.clearMatchingCases = (/**
         * @return {?}
         */
        function () {
            _this.clearMatches.emit();
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsCaseListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    LiveAppsCaseListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-case-list',
                    template: "<div class=\"tcs-case-list-box\" fxLayout=\"column\">\n  <div class=\"tcs-case-list-header\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n    <div fxFlex fxLayoutAlign=\"start center\">\n      <mat-icon class=\"tcs-icon tcs-caselist-icon\" svgIcon=\"tcs-caselist-icon\"></mat-icon>\n      <div class=\"tcs-case-list-header-text\">{{headerText}}</div>\n      <div fxFlex class=\"tcs-type-filter-text\" fxLayoutAlign=\"end center\">\n        <div *ngIf=\"headerMessage\" fxLayout=\"row\">\n          [\n          <div class=\"tcs-type-filter-text-highlight\">{{headerMessage}}</div>\n          ]\n        </div>\n        <mat-icon (click)=\"clearMatchingCases()\" class=\"tcs-icon tcs-icon-active tcs-clear-cases-icon\" svgIcon=\"tcs-clear-icon\" matTooltip=\"Clear Matching Cases\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\"></mat-icon>\n      </div>\n    </div>\n  </div>\n\n  <!--div *ngIf=\"displayType != 'list'\" class=\"tcs-case-list-flow-list\" fxLayout=\"column\">\n    <div class=\"tcs-case-list-item-box\" *ngFor=\"let caseRef of caseRefs\">\n      <tcla-live-apps-case-summary style=\"max-width: 297px;\" [uiAppId]=\"uiAppId\" [typeBar]=\"true\" [borderCard]=\"false\" [displayType]=\"displayType\" [sandboxId]=\"sandboxId\" [caseRef]=\"caseRef\" [highlight]=\"highlight\" (clickCase)=\"clickCaseAction($event)\"></tcla-live-apps-case-summary>\n      <div class=\"tcs-case-summary-line\"></div>\n    </div>\n  </div-->\n\n  <div *ngIf=\"displayType === 'list'\" fxFlex fxLayout=\"column\">\n    <cdk-virtual-scroll-viewport class=\"tcs-case-list-flow-list\" itemSize=\"56\">\n      <div class=\"tcs-case-list-item-box\" *cdkVirtualFor=\"let caseRef of caseRefs; templateCacheSize: 0; let index = index\">\n        <tcla-live-apps-case-summary style=\"max-width: 297px;\" [uiAppId]=\"uiAppId\" [typeBar]=\"true\" [borderCard]=\"false\" [displayType]=\"displayType\" [sandboxId]=\"sandboxId\" [caseRef]=\"caseRef\" [highlight]=\"highlight\" (clickCase)=\"clickCaseAction($event)\"></tcla-live-apps-case-summary>\n        <div class=\"tcs-case-summary-line\"></div>\n      </div>\n    </cdk-virtual-scroll-viewport>\n  </div>\n</div>\n",
                    styles: [".tcs-case-list-box{width:100%;min-height:366px;border-radius:3px;box-shadow:0 2px 8px 0 #dedede;background-color:#fff}.tcs-case-list-header{height:40px;border-radius:3px 3px 0 0;box-shadow:0 1px 2px 0 #dedede;padding-left:20px;padding-right:20px}.tcs-case-list-header-text{font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;text-align:left;color:#000;margin-left:10px}.tcs-case-list-flow-list{margin:24px;overflow-y:auto;height:278px;max-height:278px}.tcs-case-list-item-box{margin:3px 5px 10px}.tcs-icon.tcs-icon-active:hover{cursor:pointer}:host ::ng-deep .tcs-icon.tcs-icon-active:hover .svg-content{fill:#0081cb}.tcs-case-summary-line{padding:0;margin:4px 5px 0;border-bottom:1.1px solid #f4f4f4}.tcs-type-filter-text{font-family:Source Sans Pro;font-size:12px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;text-align:left;color:#000;margin-left:10px}.tcs-type-filter-text-highlight{color:#ff7800}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCaseListComponent.ctorParameters = function () { return [
        { type: LiveAppsService }
    ]; };
    LiveAppsCaseListComponent.propDecorators = {
        headerText: [{ type: Input }],
        displayType: [{ type: Input }],
        sandboxId: [{ type: Input }],
        uiAppId: [{ type: Input }],
        caseRefs: [{ type: Input }],
        highlight: [{ type: Input }],
        headerMessage: [{ type: Input }],
        clickCase: [{ type: Output }],
        clearMatches: [{ type: Output }]
    };
    return LiveAppsCaseListComponent;
}(LiveAppsComponent));
export { LiveAppsCaseListComponent };
if (false) {
    /**
     * Text shown in menu bar
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.headerText;
    /**
     * case card format - list, card, miniCard, staticList (no click event)
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.displayType;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.sandboxId;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.uiAppId;
    /**
     * List of case references to display in the list
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.caseRefs;
    /**
     * Text to highlight in the list of cases (normall text that was searched)
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.highlight;
    /**
     * Filter text displayed when listing cases after selecting case type and state via report widget
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.headerMessage;
    /**
     * ~event clickCase : Case clicked
     * ~payload CaseRoute : CaseRoute object output when case is clicked so calling component can route accordingly - ie. route to case
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.clickCase;
    /**
     * ~event clearMatches : Clear Matches button clicked
     *
     * @type {?}
     */
    LiveAppsCaseListComponent.prototype.clearMatches;
    /** @type {?} */
    LiveAppsCaseListComponent.prototype.errorMessage;
    /** @type {?} */
    LiveAppsCaseListComponent.prototype.clickCaseAction;
    /** @type {?} */
    LiveAppsCaseListComponent.prototype.clearMatchingCases;
    /**
     * @type {?}
     * @private
     */
    LiveAppsCaseListComponent.prototype.liveapps;
}
