// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-role-switcher/live-apps-role-switcher.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { RoleAttribute } from '@tibco-tcstk/tc-core-lib';
/**
 * Switch roles
 *
 * \@example <tcla-live-apps-role-switcher></tcla-live-apps-role-switcher>
 */
var LiveAppsRoleSwitcherComponent = /** @class */ (function () {
    function LiveAppsRoleSwitcherComponent() {
        var _this = this;
        /**
         * ~event changeRoleEvent : Role selected
         * ~payload RoleAttribute : RoleAttribute object emitted when role is c hanged by switcher
         */
        this.changeRoleEvent = new EventEmitter();
        this.selectRole = (/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            _this.changeRoleEvent.emit(role);
        });
        this.compareObjects = (/**
         * @param {?} o1
         * @param {?} o2
         * @return {?}
         */
        function (o1, o2) {
            return o1.id === o2.id;
        });
    }
    /**
     * @return {?}
     */
    LiveAppsRoleSwitcherComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    LiveAppsRoleSwitcherComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-role-switcher',
                    template: "<div>\n    <mat-form-field>\n        <mat-label>Viewing as</mat-label>\n        <mat-select [(ngModel)]=\"selectedRole\" (selectionChange)=\"selectRole(selectedRole)\" [compareWith]=\"compareObjects\"> \n            <mat-option *ngFor=\"let role of roles\" [value]=\"role\">{{role.display}}</mat-option>\n        </mat-select>\n    </mat-form-field>\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsRoleSwitcherComponent.ctorParameters = function () { return []; };
    LiveAppsRoleSwitcherComponent.propDecorators = {
        roles: [{ type: Input }],
        selectedRole: [{ type: Input }],
        changeRoleEvent: [{ type: Output }]
    };
    return LiveAppsRoleSwitcherComponent;
}());
export { LiveAppsRoleSwitcherComponent };
if (false) {
    /**
     * Roles - The users current roles
     * @type {?}
     */
    LiveAppsRoleSwitcherComponent.prototype.roles;
    /**
     * RoleAttribute - selected Role - check with MC
     * @type {?}
     */
    LiveAppsRoleSwitcherComponent.prototype.selectedRole;
    /**
     * ~event changeRoleEvent : Role selected
     * ~payload RoleAttribute : RoleAttribute object emitted when role is c hanged by switcher
     * @type {?}
     */
    LiveAppsRoleSwitcherComponent.prototype.changeRoleEvent;
    /** @type {?} */
    LiveAppsRoleSwitcherComponent.prototype.selectRole;
    /** @type {?} */
    LiveAppsRoleSwitcherComponent.prototype.compareObjects;
}
