// @tibco-tcstk/tc-core-lib/esm5/lib/components/tibco-cloud-login/tibco-cloud-login.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { LoginPrefill } from '../../models/tc-login';
import { map } from 'rxjs/operators';
import { TcLoginService } from '../../services/tc-login.service';
/**
 * This component will attempt to log the user in.
 *
 * ![alt-text](../Cloud-Login.png "Image")
 *
 * \@example <tc-tibco-cloud-login *ngIf="!loggedIn (loggedIn)="handleLoggedIn($event)"></tc-tibco-cloud-login>
 */
var TibcoCloudLoginComponent = /** @class */ (function () {
    /**
    * The Constructor creates the Login Dialog
    */
    function TibcoCloudLoginComponent(tcLogin) {
        this.tcLogin = tcLogin;
        /**
         * Notify parent that user is logged in ok.
         */
        this.loggedIn = new EventEmitter();
        this.loading = false;
    }
    /**
    * @ignore
    */
    /**
     * @ignore
     * @return {?}
     */
    TibcoCloudLoginComponent.prototype.ngOnInit = /**
     * @ignore
     * @return {?}
     */
    function () {
        console.log('Login Init');
        if (this.loginPrefill) {
            this.name = this.loginPrefill.emailId;
            this.clientId = this.loginPrefill.clientId;
        }
    };
    /**
     * @return {?}
     */
    TibcoCloudLoginComponent.prototype.doLogin = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.loading = true;
        this.loginError = undefined;
        this.auth = this.tcLogin.login(this.name, this.password, this.clientId).pipe(map((/**
         * @param {?} authInfo
         * @return {?}
         */
        function (authInfo) {
            _this.authInfo = authInfo;
            return authInfo;
        })));
        this.auth.subscribe((/**
         * @param {?} authorize
         * @return {?}
         */
        function (authorize) {
            _this.loading = false;
            // ok logged in
            console.log('User logged in...');
            _this.loggedIn.emit({ authInfo: authorize, accessToken: _this.token });
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.loading = false;
            _this.loginError = error.error.errorMsg;
            console.error('Login Failed: ');
            console.error(error);
        }));
    };
    TibcoCloudLoginComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tc-tibco-cloud-login',
                    template: "<div fxFill fxLayoutAlign=\"center start\">\n  <div fxFlex.gt-sm=\"grow 75\" class=\"la-login-div\" ngStyle.lt-md=\"height: 100%; width:100%\" ngStyle.gt-sm=\"max-width: 500px; margin-top: 10px;\">\n    <mat-card fxFlex class=\"la-login-card\" fxLayout=\"column\">\n        <div fxflex class=\"la-outer-form-div\" fxLayoutAlign=\"start stretch\">\n          <form #laLoginForm=\"ngForm\" class=\"la-login-form\" (submit)=\"doLogin()\" fxLayout=\"column\" fxFlex disabled=\"loading\">\n            <mat-form-field>\n              <input matInput placeholder=\"Email Address\" [(ngModel)]=\"name\" name=\"email\" required>\n            </mat-form-field>\n            <mat-form-field>\n              <input matInput placeholder=\"Password\" [(ngModel)]=\"password\" type=\"password\" name=\"password\" required>\n            </mat-form-field>\n\n            <mat-accordion>\n              <mat-expansion-panel [expanded]=\"!clientId\">\n                <mat-expansion-panel-header>\n                  <mat-panel-title>API Access Keys</mat-panel-title>\n                </mat-expansion-panel-header>\n                <mat-form-field fxFlex>\n                  <input matInput placeholder=\"Client Id\" [(ngModel)]=\"clientId\"  name=\"ClientID\" required>\n                </mat-form-field>\n              </mat-expansion-panel>\n            </mat-accordion>\n\n            <div class=\"la-login-button-div\" fxFlex fxLayoutAlign=\"center start\">\n              <button ngStyle.lt-md=\"width: 100%;\" type=\"submit\" mat-raised-button color=\"primary\" [disabled]=\"!laLoginForm.form.valid || loading\">Login</button>\n            </div>\n\n          </form>\n        </div>\n\n        <div fxFlex class=\"spacer\"></div>\n\n        <div class=\"la-login-info-pane\">\n          <div *ngIf=\"loading\">\n            Logging in...\n          </div>\n          <div *ngIf=\"loginError\" style=\"color: red\">{{loginError}}</div>\n        </div>\n    </mat-card>\n  </div>\n</div>\n",
                    styles: [".la-login-div{min-width:300px}.la-login-button{width:50px}.la-login-info-pane{min-height:50px;height:50px;max-height:50px;margin-top:10px}.spacer{min-height:0}.la-login-button-div{margin-top:25px}"]
                }] }
    ];
    /** @nocollapse */
    TibcoCloudLoginComponent.ctorParameters = function () { return [
        { type: TcLoginService }
    ]; };
    TibcoCloudLoginComponent.propDecorators = {
        loggedIn: [{ type: Output }],
        loginPrefill: [{ type: Input }]
    };
    return TibcoCloudLoginComponent;
}());
export { TibcoCloudLoginComponent };
if (false) {
    /**
     * Notify parent that user is logged in ok.
     * @type {?}
     */
    TibcoCloudLoginComponent.prototype.loggedIn;
    /**
     * Login Data
     * @type {?}
     */
    TibcoCloudLoginComponent.prototype.loginPrefill;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.name;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.password;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.clientId;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.loading;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.accountsInfo;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.loginError;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.token;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.authInfo;
    /** @type {?} */
    TibcoCloudLoginComponent.prototype.auth;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudLoginComponent.prototype.tcLogin;
}
