// @tibco-tcstk/tc-core-lib/esm5/lib/components/tibco-cloud-select-table/tibco-cloud-select-table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { TibcoCloudTableComponent } from '../tibco-cloud-table/tibco-cloud-table.component';
import { LogService } from '../../services/tc-logging.service';
/**
 * A table that allows rows to be selected
 *
 * \@example <tc-tibco-cloud-select-table></tc-tibco-cloud-select-table>
 */
var TibcoCloudSelectTableComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TibcoCloudSelectTableComponent, _super);
    function TibcoCloudSelectTableComponent(logger) {
        var _this = _super.call(this) || this;
        _this.logger = logger;
        /**
         * ~event selectedlines : the lines selected
         * ~payload any : The lines that are selected in the table by the user
         */
        _this.selectedlines = new EventEmitter();
        _this.selected = {};
        logger.info('Select Table Started... ');
        return _this;
    }
    /**
     * @param {?} element
     * @return {?}
     */
    TibcoCloudSelectTableComponent.prototype.highlight = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        element.highlighted = !element.highlighted;
    };
    /**
     * @param {?} row
     * @return {?}
     */
    TibcoCloudSelectTableComponent.prototype.clicked = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        var e_1, _a;
        // console.log(row);
        /** @type {?} */
        var myData = JSON.parse(JSON.stringify(this.dataSource.data));
        // console.log(this.dataSource);
        /** @type {?} */
        var selectedArray = new Array();
        try {
            for (var myData_1 = tslib_1.__values(myData), myData_1_1 = myData_1.next(); !myData_1_1.done; myData_1_1 = myData_1.next()) {
                var line = myData_1_1.value;
                // console.log('s:' , line);
                if (line.highlighted) {
                    delete line['hovered'];
                    delete line['highlighted'];
                    selectedArray.push(line);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (myData_1_1 && !myData_1_1.done && (_a = myData_1.return)) _a.call(myData_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        // console.log('Selected Array: ' , selectedArray);
        this.selected = JSON.stringify(selectedArray);
        this.selectedlines.emit(this.selected);
    };
    TibcoCloudSelectTableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tc-tibco-cloud-select-table',
                    template: "<div class=\"mat-elevation-z8\">\n  <table id=\"{{id}}\" mat-table class=\"tcst-full-width-table\" [dataSource]=\"dataSource\" matSort aria-label=\"Elements\">\n    <ng-container *ngFor=\"let column of columns\" [matColumnDef]='column.columnDef'>\n      <mat-header-cell *matHeaderCellDef class=\"tcst-header\" mat-sort-header>{{camelCaseTW(column.header)}}</mat-header-cell>\n      <mat-cell *matCellDef=\"let row\" >{{ column.cell(row) }}</mat-cell>\n    </ng-container>\n    <mat-header-row *matHeaderRowDef=\"displayedColumns\"></mat-header-row>\n    <mat-row *matRowDef=\"let row; columns: displayedColumns;  let even = even;\"\n             [ngClass]=\"{hovered: row.hovered, highlighted: row.highlighted, tcstgray: even}\"\n             (click)=\"row.highlighted = !row.highlighted; clicked(row);\" (mouseover)=\"row.hovered = true\" (mouseout)=\"row.hovered = false\"></mat-row>\n\n\n  </table> <!-- *ngIf=\"dataSource.data.length > 10\" -->\n<!-- <div class='' ngClass=\"{'tcst-hide': dataSource.data.length < 10}\"> -->\n  <mat-paginator [ngClass]=\"{'tcst-hide': dataSource.data.length < 10}\"  #paginator\n                 [length]=\"dataSource.data.length\"\n                 [pageIndex]=\"0\"\n                 [pageSize]=\"10\"\n                 [pageSizeOptions]=\"[5, 10, 20, 25]\">\n  </mat-paginator>\n</div>\n<!--</div>-->\n",
                    styles: [".tcst-full-width-table{width:100%}.mat-row.hovered{background:#eee;cursor:pointer}.mat-row.highlighted{background:#999}.tcst-header{background:#0081cb;color:#fff;padding:5px 30px;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center}.tcst-header:hover{background-color:#229be0}.tcstgray{background-color:#f5f5f5}.tcst-hide{display:none}"]
                }] }
    ];
    /** @nocollapse */
    TibcoCloudSelectTableComponent.ctorParameters = function () { return [
        { type: LogService }
    ]; };
    TibcoCloudSelectTableComponent.propDecorators = {
        selectedlines: [{ type: Output }]
    };
    return TibcoCloudSelectTableComponent;
}(TibcoCloudTableComponent));
export { TibcoCloudSelectTableComponent };
if (false) {
    /**
     * ~event selectedlines : the lines selected
     * ~payload any : The lines that are selected in the table by the user
     * @type {?}
     */
    TibcoCloudSelectTableComponent.prototype.selectedlines;
    /** @type {?} */
    TibcoCloudSelectTableComponent.prototype.selected;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudSelectTableComponent.prototype.logger;
}
