// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-settings/live-apps-settings.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { LiveAppsConfig } from '../../models/tc-liveapps-config';
import { ActivatedRoute } from '@angular/router';
import { TcLiveAppsConfigService } from '../../services/tc-live-apps-config.service';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { MatSnackBar } from '@angular/material';
/**
 * Allow Selection of live apps applications (part of config)
 *
 * ![alt-text](../live-apps-settings.png "")
 *
 * \@example <tcla-live-apps-settings></tcla-live-apps-settings>
 */
export class LiveAppsSettingsComponent extends LiveAppsComponent {
    /**
     * @param {?} route
     * @param {?} liveAppsConfigService
     * @param {?} snackBar
     */
    constructor(route, liveAppsConfigService, snackBar) {
        super();
        this.route = route;
        this.liveAppsConfigService = liveAppsConfigService;
        this.snackBar = snackBar;
        this.LIVE_APPS_URL = '/apps/dt-app/index.html#/application-content/';
        this.handleEditLiveAppClick = (/**
         * @return {?}
         */
        () => {
            // window.open(this.LIVE_APPS_URL + applicationId);
            // currently no public API to get the appId version for design time to populate the URL
            window.open(this.LIVE_APPS_URL);
        });
        this.runSaveFunction = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const liveAppsConfig = new LiveAppsConfig().deserialize({
                applicationIds: this.selectedAppIds,
                caseIconsFolderId: this.liveAppsConfig.caseIconsFolderId,
                documentAppId: this.liveAppsConfig.documentAppId,
                collaborationAppId: this.liveAppsConfig.documentAppId
            });
            this.liveAppsConfigService.updateLiveAppsConfig(this.sandboxId, this.generalConfig.uiAppId, liveAppsConfig, this.liveAppsConfig.id).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            result => {
                this.snackBar.open('Live Apps App Selection Configuration settings saved', 'OK', {
                    duration: 3000
                });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                this.snackBar.open('Error saving Live Apps App Selection Configuration', 'OK', {
                    duration: 3000
                });
            }));
        });
    }
    /**
     * @protected
     * @return {?}
     */
    getLiveAppsConfigService() {
        return this.liveAppsConfigService;
    }
    /**
     * @param {?} appIds
     * @return {?}
     */
    handleAppIdSelection(appIds) {
        console.log('Selected: ', appIds);
        this.selectedAppIds = appIds;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.generalConfig = this.route.snapshot.data.laConfigHolder.generalConfig;
        this.liveAppsConfig = this.route.snapshot.data.laConfigHolder.liveAppsConfig;
        this.claims = this.route.snapshot.data.claims;
        this.sandboxId = Number(this.claims.primaryProductionSandbox.id).valueOf();
        this.selectedAppIds = this.liveAppsConfig.applicationIds;
    }
}
LiveAppsSettingsComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-settings',
                template: "<div fxLayout=\"column\" class=\"tcs-live-apps-settings-pane\" fxFill>\n  <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\" [headerText]=\"'Live Apps App Selection Configuration'\">\n</tc-tibco-cloud-widget-header>\n  <div fxFlex style=\"padding: 20px; height: 100%\" fxLayout=\"column\" fxFill>\n    <div fxLayout=\"row\" fxLayoutAlign=\"start start\" fxLayoutGap=\"5px\" style=\"height: 20px\">\n      <p>Choose the TIBCO Cloud Live Apps Applications to be used within this application:</p>\n      <mat-icon class=\"tcs-designer-link\" [svgIcon]=\"'tcs-application-edit'\" (click)=\"handleEditLiveAppClick()\" matTooltip=\"Open Live Apps Designer\" matTooltipShowDelay=\"1000\" matTooltipPosition=\"below\"></mat-icon>\n    </div>\n    <tcla-live-apps-application-list class=\"tcs-app-selection-list-widget\" fxFlex=\"grow\" [sandboxId]=\"sandboxId\" [selectedAppIds]=\"selectedAppIds\" (appIdsSelected)=\"handleAppIdSelection($event)\"></tcla-live-apps-application-list>\n    <div fxFlex class=\"tcs-filler-panel\"></div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n      <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n    </div>\n  </div>\n</div>\n",
                styles: [".tcs-designer-link{margin-bottom:10px}.tcs-designer-link:hover{fill:#0081cb;cursor:pointer}.tcs-app-selection-list-widget,.tcs-live-apps-settings-pane{overflow:hidden}"]
            }] }
];
/** @nocollapse */
LiveAppsSettingsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: TcLiveAppsConfigService },
    { type: MatSnackBar }
];
if (false) {
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.LIVE_APPS_URL;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.sandboxId;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.selectedAppIds;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.liveAppsConfig;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.generalConfig;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.claims;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.handleEditLiveAppClick;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.runSaveFunction;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsComponent.prototype.route;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsComponent.prototype.liveAppsConfigService;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsComponent.prototype.snackBar;
}
