// @tibco-tcstk/tc-liveapps-lib/esm5/lib/models/liveappsdata.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var NoteThread = /** @class */ (function () {
    function NoteThread(threadId, showReplies, showNewReply, editMode, newReply, thread, note) {
        this.threadId = threadId;
        this.showReplies = showReplies;
        this.showNewReply = showNewReply;
        this.editMode = editMode;
        this.newReply = newReply;
        this.thread = thread;
        this.note = note;
    }
    return NoteThread;
}());
export { NoteThread };
if (false) {
    /** @type {?} */
    NoteThread.prototype.threadId;
    /** @type {?} */
    NoteThread.prototype.showReplies;
    /** @type {?} */
    NoteThread.prototype.showNewReply;
    /** @type {?} */
    NoteThread.prototype.editMode;
    /** @type {?} */
    NoteThread.prototype.newReply;
    /** @type {?} */
    NoteThread.prototype.thread;
    /** @type {?} */
    NoteThread.prototype.note;
}
var CaseRoute = /** @class */ (function () {
    function CaseRoute() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseRoute.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseRoute;
}());
export { CaseRoute };
if (false) {
    /** @type {?} */
    CaseRoute.prototype.caseRef;
    /** @type {?} */
    CaseRoute.prototype.appId;
    /** @type {?} */
    CaseRoute.prototype.typeId;
}
var NotificationCollection = /** @class */ (function () {
    function NotificationCollection(collectionName, lifecycledWithType, lifecycledWithId) {
        this.collectionName = collectionName;
        this.lifecycledWithType = lifecycledWithType;
        this.lifecycledWithId = lifecycledWithId;
    }
    return NotificationCollection;
}());
export { NotificationCollection };
if (false) {
    /** @type {?} */
    NotificationCollection.prototype.collectionName;
    /** @type {?} */
    NotificationCollection.prototype.lifecycledWithType;
    /** @type {?} */
    NotificationCollection.prototype.lifecycledWithId;
}
var Notification = /** @class */ (function () {
    function Notification() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Notification.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Notification;
}());
export { Notification };
if (false) {
    /** @type {?} */
    Notification.prototype.entityId;
    /** @type {?} */
    Notification.prototype.entityName;
    /** @type {?} */
    Notification.prototype.entityType;
    /** @type {?} */
    Notification.prototype.id;
    /** @type {?} */
    Notification.prototype.notifyCollection;
    /** @type {?} */
    Notification.prototype.notifyOnAllReplies;
    /** @type {?} */
    Notification.prototype.notifyOnUpdate;
    /** @type {?} */
    Notification.prototype.threadId;
    /** @type {?} */
    Notification.prototype.topicId;
    /** @type {?} */
    Notification.prototype.type;
}
var NotificationList = /** @class */ (function () {
    function NotificationList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    NotificationList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).notifications = [];
        Object.assign((/** @type {?} */ (this)).notifications, input);
        return (/** @type {?} */ (this));
    };
    return NotificationList;
}());
export { NotificationList };
if (false) {
    /** @type {?} */
    NotificationList.prototype.notifications;
}
var ThreadList = /** @class */ (function () {
    function ThreadList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    ThreadList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).threads = [];
        Object.assign((/** @type {?} */ (this)).threads, input);
        return (/** @type {?} */ (this));
    };
    return ThreadList;
}());
export { ThreadList };
if (false) {
    /** @type {?} */
    ThreadList.prototype.threads;
}
var Metadata = /** @class */ (function () {
    function Metadata(createdBy, creationTimestamp, createdByDetails, modifiedBy, modificationTimestamp, modifiedByDetails, lock, lockType, lockedBy, msLockExpiry, msSystemTime, markedForPurge, applicationId, applicationLabel, typeId, stateColor, stateIcon, caseTypeColor, caseTypeIcon, useCaseTypeColor) {
        this.createdBy = createdBy;
        this.creationTimestamp = creationTimestamp;
        this.createdByDetails = createdByDetails;
        this.modifiedBy = modifiedBy;
        this.modificationTimestamp = modificationTimestamp;
        this.modifiedByDetails = modifiedByDetails;
        this.lock = lock;
        this.lockType = lockType;
        this.lockedBy = lockedBy;
        this.msLockExpiry = msLockExpiry;
        this.msSystemTime = msSystemTime;
        this.markedForPurge = markedForPurge;
        this.applicationId = applicationId;
        this.applicationLabel = applicationLabel;
        this.typeId = typeId;
        this.stateColor = stateColor;
        this.stateIcon = stateIcon;
        this.caseTypeColor = caseTypeColor;
        this.caseTypeIcon = caseTypeIcon;
        this.useCaseTypeColor = useCaseTypeColor;
    }
    return Metadata;
}());
export { Metadata };
if (false) {
    /** @type {?} */
    Metadata.prototype.createdBy;
    /** @type {?} */
    Metadata.prototype.creationTimestamp;
    /** @type {?} */
    Metadata.prototype.createdByDetails;
    /** @type {?} */
    Metadata.prototype.modifiedBy;
    /** @type {?} */
    Metadata.prototype.modificationTimestamp;
    /** @type {?} */
    Metadata.prototype.modifiedByDetails;
    /** @type {?} */
    Metadata.prototype.lock;
    /** @type {?} */
    Metadata.prototype.lockType;
    /** @type {?} */
    Metadata.prototype.lockedBy;
    /** @type {?} */
    Metadata.prototype.msLockExpiry;
    /** @type {?} */
    Metadata.prototype.msSystemTime;
    /** @type {?} */
    Metadata.prototype.markedForPurge;
    /** @type {?} */
    Metadata.prototype.applicationId;
    /** @type {?} */
    Metadata.prototype.applicationLabel;
    /** @type {?} */
    Metadata.prototype.typeId;
    /** @type {?} */
    Metadata.prototype.stateColor;
    /** @type {?} */
    Metadata.prototype.stateIcon;
    /** @type {?} */
    Metadata.prototype.caseTypeColor;
    /** @type {?} */
    Metadata.prototype.caseTypeIcon;
    /** @type {?} */
    Metadata.prototype.useCaseTypeColor;
}
var CaseTypeState = /** @class */ (function () {
    function CaseTypeState() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeState.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeState;
}());
export { CaseTypeState };
if (false) {
    /** @type {?} */
    CaseTypeState.prototype.id;
    /** @type {?} */
    CaseTypeState.prototype.label;
    /** @type {?} */
    CaseTypeState.prototype.value;
    /** @type {?} */
    CaseTypeState.prototype.isTerminal;
}
var CaseTypeStatesListList = /** @class */ (function () {
    function CaseTypeStatesListList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeStatesListList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).casetypes = [];
        Object.assign((/** @type {?} */ (this)).casetypes, input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeStatesListList;
}());
export { CaseTypeStatesListList };
if (false) {
    /** @type {?} */
    CaseTypeStatesListList.prototype.casetypes;
}
var CaseTypeStatesList = /** @class */ (function () {
    function CaseTypeStatesList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeStatesList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).states = [];
        Object.assign((/** @type {?} */ (this)).states, input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeStatesList;
}());
export { CaseTypeStatesList };
if (false) {
    /** @type {?} */
    CaseTypeStatesList.prototype.states;
}
var CaseInfo = /** @class */ (function () {
    function CaseInfo() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseInfo.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        (/** @type {?} */ (this)).summaryObj = (/** @type {?} */ (this)).summary ? JSON.parse((/** @type {?} */ (this)).summary) : undefined;
        (/** @type {?} */ (this)).casedataObj = (/** @type {?} */ (this)).casedata ? JSON.parse((/** @type {?} */ (this)).casedata) : undefined;
        (/** @type {?} */ (this)).untaggedCasedataObj = (/** @type {?} */ (this)).untaggedCasedata ? JSON.parse((/** @type {?} */ (this)).untaggedCasedata) : undefined;
        return (/** @type {?} */ (this));
    };
    return CaseInfo;
}());
export { CaseInfo };
if (false) {
    /** @type {?} */
    CaseInfo.prototype.deleted;
    /** @type {?} */
    CaseInfo.prototype.caseReference;
    /** @type {?} */
    CaseInfo.prototype.untaggedCasedata;
    /** @type {?} */
    CaseInfo.prototype.untaggedCasedataObj;
    /** @type {?} */
    CaseInfo.prototype.casedata;
    /** @type {?} */
    CaseInfo.prototype.casedataObj;
    /** @type {?} */
    CaseInfo.prototype.summary;
    /** @type {?} */
    CaseInfo.prototype.summaryObj;
    /** @type {?} */
    CaseInfo.prototype.metadata;
}
var CaseInfoList = /** @class */ (function () {
    function CaseInfoList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseInfoList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).caseinfos = [];
        Object.assign((/** @type {?} */ (this)).caseinfos, input);
        (/** @type {?} */ (this)).caseinfos.forEach((/**
         * @param {?} caseinfo
         * @return {?}
         */
        function (caseinfo) {
            if (caseinfo.summary) {
                caseinfo.summaryObj = JSON.parse(caseinfo.summary);
            }
            if (caseinfo.casedata) {
                caseinfo.casedataObj = JSON.parse(caseinfo.casedata);
            }
        }));
        return (/** @type {?} */ (this));
    };
    return CaseInfoList;
}());
export { CaseInfoList };
if (false) {
    /** @type {?} */
    CaseInfoList.prototype.caseinfos;
    /** @type {?} */
    CaseInfoList.prototype.casecount;
}
var CaseAttribute = /** @class */ (function () {
    function CaseAttribute() {
    }
    return CaseAttribute;
}());
export { CaseAttribute };
if (false) {
    /** @type {?} */
    CaseAttribute.prototype.isIdentifier;
    /** @type {?} */
    CaseAttribute.prototype.isMandatory;
    /** @type {?} */
    CaseAttribute.prototype.isStructuredType;
    /** @type {?} */
    CaseAttribute.prototype.label;
    /** @type {?} */
    CaseAttribute.prototype.name;
    /** @type {?} */
    CaseAttribute.prototype.type;
    /** @type {?} */
    CaseAttribute.prototype.maximum;
    /** @type {?} */
    CaseAttribute.prototype.minimum;
}
var JsonSchema = /** @class */ (function () {
    function JsonSchema() {
    }
    return JsonSchema;
}());
export { JsonSchema };
if (false) {
    /** @type {?} */
    JsonSchema.prototype.$schema;
    /** @type {?} */
    JsonSchema.prototype.definitions;
    /** @type {?} */
    JsonSchema.prototype.properties;
    /** @type {?} */
    JsonSchema.prototype.type;
    /** @type {?} */
    JsonSchema.prototype.required;
}
var Process = /** @class */ (function () {
    function Process() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Process.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Process;
}());
export { Process };
if (false) {
    /** @type {?} */
    Process.prototype.jsonSchema;
    /** @type {?} */
    Process.prototype.name;
    /** @type {?} */
    Process.prototype.id;
    /** @type {?} */
    Process.prototype.unsupportedForm;
}
var ProcessId = /** @class */ (function () {
    function ProcessId() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    ProcessId.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return ProcessId;
}());
export { ProcessId };
if (false) {
    /** @type {?} */
    ProcessId.prototype.caseIdentifier;
    /** @type {?} */
    ProcessId.prototype.caseReference;
}
var CaseType = /** @class */ (function () {
    function CaseType() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseType.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseType;
}());
export { CaseType };
if (false) {
    /** @type {?} */
    CaseType.prototype.actions;
    /** @type {?} */
    CaseType.prototype.applicationId;
    /** @type {?} */
    CaseType.prototype.applicationInternalName;
    /** @type {?} */
    CaseType.prototype.applicationName;
    /** @type {?} */
    CaseType.prototype.attributes;
    /** @type {?} */
    CaseType.prototype.creators;
    /** @type {?} */
    CaseType.prototype.id;
    /** @type {?} */
    CaseType.prototype.isCase;
    /** @type {?} */
    CaseType.prototype.label;
    /** @type {?} */
    CaseType.prototype.name;
    /** @type {?} */
    CaseType.prototype.jsonSchema;
}
var CaseTypesList = /** @class */ (function () {
    function CaseTypesList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypesList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).casetypes = [];
        Object.assign((/** @type {?} */ (this)).casetypes, input);
        return (/** @type {?} */ (this));
    };
    return CaseTypesList;
}());
export { CaseTypesList };
if (false) {
    /** @type {?} */
    CaseTypesList.prototype.casetypes;
}
var CaseAction = /** @class */ (function () {
    function CaseAction() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseAction.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseAction;
}());
export { CaseAction };
if (false) {
    /** @type {?} */
    CaseAction.prototype.activityName;
    /** @type {?} */
    CaseAction.prototype.applicationId;
    /** @type {?} */
    CaseAction.prototype.applicationName;
    /** @type {?} */
    CaseAction.prototype.availableInStates;
    /** @type {?} */
    CaseAction.prototype.chatBotEnabled;
    /** @type {?} */
    CaseAction.prototype.id;
    /** @type {?} */
    CaseAction.prototype.label;
    /** @type {?} */
    CaseAction.prototype.name;
    /** @type {?} */
    CaseAction.prototype.noData;
    /** @type {?} */
    CaseAction.prototype.performerPaths;
    /** @type {?} */
    CaseAction.prototype.roles;
    /** @type {?} */
    CaseAction.prototype.version;
    /** @type {?} */
    CaseAction.prototype.process;
}
var CaseActionsList = /** @class */ (function () {
    function CaseActionsList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseActionsList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).actions = [];
        Object.assign((/** @type {?} */ (this)).actions, input);
        return (/** @type {?} */ (this));
    };
    return CaseActionsList;
}());
export { CaseActionsList };
if (false) {
    /** @type {?} */
    CaseActionsList.prototype.actions;
}
var CaseCreator = /** @class */ (function () {
    function CaseCreator() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseCreator.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseCreator;
}());
export { CaseCreator };
if (false) {
    /** @type {?} */
    CaseCreator.prototype.activityName;
    /** @type {?} */
    CaseCreator.prototype.applicationId;
    /** @type {?} */
    CaseCreator.prototype.applicationName;
    /** @type {?} */
    CaseCreator.prototype.chatBotEnabled;
    /** @type {?} */
    CaseCreator.prototype.id;
    /** @type {?} */
    CaseCreator.prototype.label;
    /** @type {?} */
    CaseCreator.prototype.name;
    /** @type {?} */
    CaseCreator.prototype.roles;
    /** @type {?} */
    CaseCreator.prototype.version;
    /** @type {?} */
    CaseCreator.prototype.process;
}
var CaseCreatorsList = /** @class */ (function () {
    function CaseCreatorsList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseCreatorsList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).creators = [];
        Object.assign((/** @type {?} */ (this)).creators, input);
        return (/** @type {?} */ (this));
    };
    return CaseCreatorsList;
}());
export { CaseCreatorsList };
if (false) {
    /** @type {?} */
    CaseCreatorsList.prototype.creators;
}
var CaseList = /** @class */ (function () {
    function CaseList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).uniqueKey = input.uniqueKey;
        (/** @type {?} */ (this)).maxSize = input.maxSize;
        (/** @type {?} */ (this)).caseRefs = [];
        Object.assign((/** @type {?} */ (this)).caseRefs, input.caseRefs);
        return (/** @type {?} */ (this));
    };
    return CaseList;
}());
export { CaseList };
if (false) {
    /** @type {?} */
    CaseList.prototype.uniqueKey;
    /** @type {?} */
    CaseList.prototype.maxSize;
    /** @type {?} */
    CaseList.prototype.caseRefs;
}
var ApiResponseText = /** @class */ (function () {
    function ApiResponseText() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    ApiResponseText.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return ApiResponseText;
}());
export { ApiResponseText };
if (false) {
    /** @type {?} */
    ApiResponseText.prototype.message;
}
var ApiResponseError = /** @class */ (function () {
    function ApiResponseError() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    ApiResponseError.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return ApiResponseError;
}());
export { ApiResponseError };
if (false) {
    /** @type {?} */
    ApiResponseError.prototype.errorMsg;
    /** @type {?} */
    ApiResponseError.prototype.errorCode;
    /** @type {?} */
    ApiResponseError.prototype.contextAttributes;
}
var UserInfo = /** @class */ (function () {
    function UserInfo() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    UserInfo.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return UserInfo;
}());
export { UserInfo };
if (false) {
    /** @type {?} */
    UserInfo.prototype.externalId;
    /** @type {?} */
    UserInfo.prototype.firstName;
    /** @type {?} */
    UserInfo.prototype.lastName;
    /** @type {?} */
    UserInfo.prototype.username;
    /** @type {?} */
    UserInfo.prototype.email;
    /** @type {?} */
    UserInfo.prototype.type;
    /** @type {?} */
    UserInfo.prototype.id;
}
var NotesRole = /** @class */ (function () {
    function NotesRole() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    NotesRole.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return NotesRole;
}());
export { NotesRole };
if (false) {
    /** @type {?} */
    NotesRole.prototype.entityId;
    /** @type {?} */
    NotesRole.prototype.entityName;
    /** @type {?} */
    NotesRole.prototype.entityType;
    /** @type {?} */
    NotesRole.prototype.id;
    /** @type {?} */
    NotesRole.prototype.role;
    /** @type {?} */
    NotesRole.prototype.threadId;
}
var Thread = /** @class */ (function () {
    function Thread() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Thread.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Thread;
}());
export { Thread };
if (false) {
    /** @type {?} */
    Thread.prototype.acceptedStatusCount;
    /** @type {?} */
    Thread.prototype.actionDate;
    /** @type {?} */
    Thread.prototype.actionId;
    /** @type {?} */
    Thread.prototype.actionType;
    /** @type {?} */
    Thread.prototype.answerNoteCount;
    /** @type {?} */
    Thread.prototype.answeredStatusCount;
    /** @type {?} */
    Thread.prototype.createdById;
    /** @type {?} */
    Thread.prototype.createdByName;
    /** @type {?} */
    Thread.prototype.createdDate;
    /** @type {?} */
    Thread.prototype.id;
    /** @type {?} */
    Thread.prototype.informationalNoteCount;
    /** @type {?} */
    Thread.prototype.isAbandoned;
    /** @type {?} */
    Thread.prototype.isLocked;
    /** @type {?} */
    Thread.prototype.isOrphaned;
    /** @type {?} */
    Thread.prototype.issueNoteCount;
    /** @type {?} */
    Thread.prototype.lifecycledWithId;
    /** @type {?} */
    Thread.prototype.lifecycledWithType;
    /** @type {?} */
    Thread.prototype.modifiedById;
    /** @type {?} */
    Thread.prototype.modifiedByName;
    /** @type {?} */
    Thread.prototype.modifiedDate;
    /** @type {?} */
    Thread.prototype.questionNoteCount;
    /** @type {?} */
    Thread.prototype.relatedItemCollection;
    /** @type {?} */
    Thread.prototype.relatedItemId;
    /** @type {?} */
    Thread.prototype.relatedItemType;
    /** @type {?} */
    Thread.prototype.resolutionNoteCount;
    /** @type {?} */
    Thread.prototype.resolvedStatusCount;
    /** @type {?} */
    Thread.prototype.roles;
    /** @type {?} */
    Thread.prototype.status;
    /** @type {?} */
    Thread.prototype.topLevelNoteId;
    /** @type {?} */
    Thread.prototype.topicId;
    /** @type {?} */
    Thread.prototype.totalCount;
    /** @type {?} */
    Thread.prototype.unacceptedStatusCount;
    /** @type {?} */
    Thread.prototype.unansweredStatusCount;
    /** @type {?} */
    Thread.prototype.unclassifiedStatusCount;
    /** @type {?} */
    Thread.prototype.unresolvedStatusCount;
}
var Note = /** @class */ (function () {
    function Note() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Note.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Note;
}());
export { Note };
if (false) {
    /** @type {?} */
    Note.prototype.attributes;
    /** @type {?} */
    Note.prototype.createdById;
    /** @type {?} */
    Note.prototype.createdByName;
    /** @type {?} */
    Note.prototype.createdDate;
    /** @type {?} */
    Note.prototype.descendantModifiedDate;
    /** @type {?} */
    Note.prototype.id;
    /** @type {?} */
    Note.prototype.isDescendantNewOrModified;
    /** @type {?} */
    Note.prototype.isHidden;
    /** @type {?} */
    Note.prototype.isModified;
    /** @type {?} */
    Note.prototype.isNew;
    /** @type {?} */
    Note.prototype.level;
    /** @type {?} */
    Note.prototype.modifiedByName;
    /** @type {?} */
    Note.prototype.modifiedDate;
    /** @type {?} */
    Note.prototype.notificationLabel;
    /** @type {?} */
    Note.prototype.notificationUrl;
    /** @type {?} */
    Note.prototype.parentId;
    /** @type {?} */
    Note.prototype.relatedItemCollection;
    /** @type {?} */
    Note.prototype.relatedItemId;
    /** @type {?} */
    Note.prototype.relatedItemType;
    /** @type {?} */
    Note.prototype.replyAcceptedId;
    /** @type {?} */
    Note.prototype.replyCount;
    /** @type {?} */
    Note.prototype.replyDate;
    /** @type {?} */
    Note.prototype.status;
    /** @type {?} */
    Note.prototype.text;
    /** @type {?} */
    Note.prototype.thread;
    /** @type {?} */
    Note.prototype.threadId;
    /** @type {?} */
    Note.prototype.title;
    /** @type {?} */
    Note.prototype.topicId;
    /** @type {?} */
    Note.prototype.type;
    /** @type {?} */
    Note.prototype.editMode;
}
var NotesList = /** @class */ (function () {
    function NotesList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    NotesList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).notes = [];
        Object.assign((/** @type {?} */ (this)).notes, input);
        return (/** @type {?} */ (this));
    };
    return NotesList;
}());
export { NotesList };
if (false) {
    /** @type {?} */
    NotesList.prototype.notes;
}
var IconMap = /** @class */ (function () {
    function IconMap(isCaseType, state, fill, icon) {
        this.isCaseType = isCaseType;
        this.state = state;
        this.fill = fill;
        this.icon = icon;
    }
    return IconMap;
}());
export { IconMap };
if (false) {
    /** @type {?} */
    IconMap.prototype.isCaseType;
    /** @type {?} */
    IconMap.prototype.state;
    /** @type {?} */
    IconMap.prototype.fill;
    /** @type {?} */
    IconMap.prototype.icon;
}
var CardConfig = /** @class */ (function () {
    function CardConfig() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CardConfig.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        // this.stateMap = [];
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CardConfig;
}());
export { CardConfig };
if (false) {
    /** @type {?} */
    CardConfig.prototype.id;
    /** @type {?} */
    CardConfig.prototype.useCaseTypeColor;
    /** @type {?} */
    CardConfig.prototype.stateMap;
}
var LoginContext = /** @class */ (function () {
    function LoginContext() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    LoginContext.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return LoginContext;
}());
export { LoginContext };
if (false) {
    /** @type {?} */
    LoginContext.prototype.authInfo;
    /** @type {?} */
    LoginContext.prototype.accessToken;
}
var CaseSearchResults = /** @class */ (function () {
    function CaseSearchResults() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseSearchResults.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseSearchResults;
}());
export { CaseSearchResults };
if (false) {
    /** @type {?} */
    CaseSearchResults.prototype.caserefs;
    /** @type {?} */
    CaseSearchResults.prototype.searchString;
}
