// @tibco-tcstk/tc-liveapps-lib/esm5/lib/models/tc-case-processes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var LaProcessSelection = /** @class */ (function () {
    // Format of ref is <applicationName>.<applicationInternalName>.<processType>.<processName>
    function LaProcessSelection(type, appSchema, caseIdAttribute, process, ref, caseReference) {
        this.type = type;
        this.appSchema = appSchema;
        this.caseIdAttribute = caseIdAttribute;
        this.process = process;
        this.ref = ref;
        this.caseReference = caseReference;
    }
    return LaProcessSelection;
}());
export { LaProcessSelection };
if (false) {
    /** @type {?} */
    LaProcessSelection.prototype.type;
    /** @type {?} */
    LaProcessSelection.prototype.appSchema;
    /** @type {?} */
    LaProcessSelection.prototype.caseIdAttribute;
    /** @type {?} */
    LaProcessSelection.prototype.process;
    /** @type {?} */
    LaProcessSelection.prototype.ref;
    /** @type {?} */
    LaProcessSelection.prototype.caseReference;
}
