// @tibco-tcstk/tc-liveapps-lib/esm5/lib/models/tc-case-states.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var StateAuditEvent = /** @class */ (function () {
    function StateAuditEvent() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    StateAuditEvent.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return StateAuditEvent;
}());
export { StateAuditEvent };
if (false) {
    /** @type {?} */
    StateAuditEvent.prototype.caseState;
    /** @type {?} */
    StateAuditEvent.prototype.type;
    /** @type {?} */
    StateAuditEvent.prototype.phaseLabel;
    /** @type {?} */
    StateAuditEvent.prototype.previousPhaseLabel;
    /** @type {?} */
    StateAuditEvent.prototype.messageId;
    /** @type {?} */
    StateAuditEvent.prototype.principalName;
    /** @type {?} */
    StateAuditEvent.prototype.creationTime;
    /** @type {?} */
    StateAuditEvent.prototype.isTerminal;
}
var StateAuditEventList = /** @class */ (function () {
    function StateAuditEventList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    StateAuditEventList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).auditEvents = [];
        Object.assign((/** @type {?} */ (this)).auditEvents, input);
        return (/** @type {?} */ (this));
    };
    return StateAuditEventList;
}());
export { StateAuditEventList };
if (false) {
    /** @type {?} */
    StateAuditEventList.prototype.auditEvents;
}
var StateTrackerData = /** @class */ (function () {
    function StateTrackerData() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    StateTrackerData.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return StateTrackerData;
}());
export { StateTrackerData };
if (false) {
    /** @type {?} */
    StateTrackerData.prototype.possibleStates;
    /** @type {?} */
    StateTrackerData.prototype.currentState;
    /** @type {?} */
    StateTrackerData.prototype.phase;
    /** @type {?} */
    StateTrackerData.prototype.caseAudit;
}
var TrackerState = /** @class */ (function () {
    function TrackerState() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    TrackerState.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return TrackerState;
}());
export { TrackerState };
if (false) {
    /** @type {?} */
    TrackerState.prototype.name;
    /** @type {?} */
    TrackerState.prototype.label;
    /** @type {?} */
    TrackerState.prototype.phase;
    /** @type {?} */
    TrackerState.prototype.previousPhase;
    /** @type {?} */
    TrackerState.prototype.isTerminal;
    /** @type {?} */
    TrackerState.prototype.status;
    /** @type {?} */
    TrackerState.prototype.user;
    /** @type {?} */
    TrackerState.prototype.changed;
}
var StateTracker = /** @class */ (function () {
    function StateTracker() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    StateTracker.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).states = [];
        Object.assign((/** @type {?} */ (this)).states, input);
        return (/** @type {?} */ (this));
    };
    return StateTracker;
}());
export { StateTracker };
if (false) {
    /** @type {?} */
    StateTracker.prototype.states;
    /** @type {?} */
    StateTracker.prototype.valid;
}
