// @tibco-tcstk/tc-liveapps-lib/esm5/lib/models/tc-document.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var DocumentAction = /** @class */ (function () {
    function DocumentAction() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    DocumentAction.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return DocumentAction;
}());
export { DocumentAction };
if (false) {
    /** @type {?} */
    DocumentAction.prototype.action;
    /** @type {?} */
    DocumentAction.prototype.document;
}
var Document = /** @class */ (function () {
    function Document() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Document.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Document;
}());
export { Document };
if (false) {
    /** @type {?} */
    Document.prototype.artifactCheckSum;
    /** @type {?} */
    Document.prototype.artifactRef;
    /** @type {?} */
    Document.prototype.artifactVersion;
    /** @type {?} */
    Document.prototype.author;
    /** @type {?} */
    Document.prototype.authorDetails;
    /** @type {?} */
    Document.prototype.creationDate;
    /** @type {?} */
    Document.prototype.description;
    /** @type {?} */
    Document.prototype.id;
    /** @type {?} */
    Document.prototype.lastModifiedBy;
    /** @type {?} */
    Document.prototype.lastModifiedByDetails;
    /** @type {?} */
    Document.prototype.lastModifiedDate;
    /** @type {?} */
    Document.prototype.mimeType;
    /** @type {?} */
    Document.prototype.name;
    /** @type {?} */
    Document.prototype.ownerApp;
    /** @type {?} */
    Document.prototype.size;
    /** @type {?} */
    Document.prototype.fileSize;
    /** @type {?} */
    Document.prototype.extension;
    /** @type {?} */
    Document.prototype.fileIcon;
}
var DocumentList = /** @class */ (function () {
    function DocumentList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    DocumentList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).documents = [];
        Object.assign((/** @type {?} */ (this)).documents, input);
        return (/** @type {?} */ (this));
    };
    return DocumentList;
}());
export { DocumentList };
if (false) {
    /** @type {?} */
    DocumentList.prototype.documents;
}
var OrgFolder = /** @class */ (function () {
    function OrgFolder() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    OrgFolder.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return OrgFolder;
}());
export { OrgFolder };
if (false) {
    /** @type {?} */
    OrgFolder.prototype.id;
    /** @type {?} */
    OrgFolder.prototype.name;
    /** @type {?} */
    OrgFolder.prototype.owner;
    /** @type {?} */
    OrgFolder.prototype.ownerSub;
    /** @type {?} */
    OrgFolder.prototype.ownerSandbox;
    /** @type {?} */
    OrgFolder.prototype.creationDate;
    /** @type {?} */
    OrgFolder.prototype.lastModifiedDate;
    /** @type {?} */
    OrgFolder.prototype.lastModifiedBy;
    /** @type {?} */
    OrgFolder.prototype.publishedVersion;
    /** @type {?} */
    OrgFolder.prototype.latestVersion;
    /** @type {?} */
    OrgFolder.prototype.publisedVersionId;
    /** @type {?} */
    OrgFolder.prototype.latestVersionId;
    /** @type {?} */
    OrgFolder.prototype.checksum;
    /** @type {?} */
    OrgFolder.prototype.extRef;
    /** @type {?} */
    OrgFolder.prototype.tags;
    /** @type {?} */
    OrgFolder.prototype.descriptor;
}
var DocTags = /** @class */ (function () {
    function DocTags() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    DocTags.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        return (/** @type {?} */ (this));
    };
    return DocTags;
}());
export { DocTags };
if (false) {
    /** @type {?} */
    DocTags.prototype.id;
    /** @type {?} */
    DocTags.prototype.name;
    /** @type {?} */
    DocTags.prototype.ownerSub;
}
