// @tibco-tcstk/tc-liveapps-lib/esm5/lib/models/tc-groups-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var Group = /** @class */ (function () {
    function Group() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Group.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Group;
}());
export { Group };
if (false) {
    /** @type {?} */
    Group.prototype.name;
    /** @type {?} */
    Group.prototype.description;
    /** @type {?} */
    Group.prototype.type;
    /** @type {?} */
    Group.prototype.id;
}
var Groups = /** @class */ (function () {
    function Groups() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Groups.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Groups;
}());
export { Groups };
if (false) {
    /** @type {?} */
    Groups.prototype.groups;
}
var Roles = /** @class */ (function () {
    function Roles() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Roles.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Roles;
}());
export { Roles };
if (false) {
    /** @type {?} */
    Roles.prototype.roles;
}
var RouteAccessDef = /** @class */ (function () {
    function RouteAccessDef() {
    }
    return RouteAccessDef;
}());
export { RouteAccessDef };
if (false) {
    /** @type {?} */
    RouteAccessDef.prototype.routeUrl;
    /** @type {?} */
    RouteAccessDef.prototype.requiredRoleId;
}
var ButtonAccessDef = /** @class */ (function () {
    function ButtonAccessDef() {
    }
    return ButtonAccessDef;
}());
export { ButtonAccessDef };
if (false) {
    /** @type {?} */
    ButtonAccessDef.prototype.buttonId;
    /** @type {?} */
    ButtonAccessDef.prototype.requiredRoleId;
}
var RouteAccessControlConfig = /** @class */ (function () {
    function RouteAccessControlConfig() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    RouteAccessControlConfig.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return RouteAccessControlConfig;
}());
export { RouteAccessControlConfig };
if (false) {
    /** @type {?} */
    RouteAccessControlConfig.prototype.routes;
    /** @type {?} */
    RouteAccessControlConfig.prototype.buttons;
}
