// @tibco-tcstk/tc-liveapps-lib/esm5/lib/models/tc-live-apps-reporting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CaseTypeReportRecordInfo = /** @class */ (function () {
    function CaseTypeReportRecordInfo() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeReportRecordInfo.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeReportRecordInfo;
}());
export { CaseTypeReportRecordInfo };
if (false) {
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.id;
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.label;
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.name;
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.color;
}
var CaseTypeReportRecord = /** @class */ (function () {
    function CaseTypeReportRecord() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeReportRecord.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeReportRecord;
}());
export { CaseTypeReportRecord };
if (false) {
    /** @type {?} */
    CaseTypeReportRecord.prototype.activeStateCaseCount;
    /** @type {?} */
    CaseTypeReportRecord.prototype.applicationId;
    /** @type {?} */
    CaseTypeReportRecord.prototype.terminalStateCaseCount;
    /** @type {?} */
    CaseTypeReportRecord.prototype.incTerminal;
    /** @type {?} */
    CaseTypeReportRecord.prototype.caseTypeInfo;
}
var CaseTypesReport = /** @class */ (function () {
    function CaseTypesReport() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypesReport.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseTypesReport;
}());
export { CaseTypesReport };
if (false) {
    /** @type {?} */
    CaseTypesReport.prototype.caseTypes;
}
var CaseTypeStateReportStateInfo = /** @class */ (function () {
    function CaseTypeStateReportStateInfo() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeStateReportStateInfo.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeStateReportStateInfo;
}());
export { CaseTypeStateReportStateInfo };
if (false) {
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.id;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.label;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.value;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.color;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.isTerminal;
}
var CaseTypeStateReportRecord = /** @class */ (function () {
    function CaseTypeStateReportRecord() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeStateReportRecord.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeStateReportRecord;
}());
export { CaseTypeStateReportRecord };
if (false) {
    /** @type {?} */
    CaseTypeStateReportRecord.prototype.caseCount;
    /** @type {?} */
    CaseTypeStateReportRecord.prototype.stateInfo;
}
var CaseTypeStateReport = /** @class */ (function () {
    function CaseTypeStateReport() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    CaseTypeStateReport.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return CaseTypeStateReport;
}());
export { CaseTypeStateReport };
if (false) {
    /** @type {?} */
    CaseTypeStateReport.prototype.caseStates;
}
