// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/models/tc-case-creator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class CaseCreatorSelectionContext {
    /**
     * @param {?} application
     * @param {?} initialData
     * @param {?} sandboxId
     * @param {?} customFormDefs
     * @param {?} legacyCreators
     * @param {?} formsFramework
     */
    constructor(application, initialData, sandboxId, customFormDefs, legacyCreators, formsFramework) {
        this.application = application;
        this.initialData = initialData;
        this.sandboxId = sandboxId;
        this.customFormDefs = customFormDefs;
        this.legacyCreators = legacyCreators;
        this.formsFramework = formsFramework;
    }
}
if (false) {
    /** @type {?} */
    CaseCreatorSelectionContext.prototype.application;
    /** @type {?} */
    CaseCreatorSelectionContext.prototype.initialData;
    /** @type {?} */
    CaseCreatorSelectionContext.prototype.sandboxId;
    /** @type {?} */
    CaseCreatorSelectionContext.prototype.customFormDefs;
    /** @type {?} */
    CaseCreatorSelectionContext.prototype.legacyCreators;
    /** @type {?} */
    CaseCreatorSelectionContext.prototype.formsFramework;
}
