// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/models/tc-case-states.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class StateAuditEvent {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    StateAuditEvent.prototype.caseState;
    /** @type {?} */
    StateAuditEvent.prototype.type;
    /** @type {?} */
    StateAuditEvent.prototype.phaseLabel;
    /** @type {?} */
    StateAuditEvent.prototype.previousPhaseLabel;
    /** @type {?} */
    StateAuditEvent.prototype.messageId;
    /** @type {?} */
    StateAuditEvent.prototype.principalName;
    /** @type {?} */
    StateAuditEvent.prototype.creationTime;
    /** @type {?} */
    StateAuditEvent.prototype.isTerminal;
}
export class StateAuditEventList {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        (/** @type {?} */ (this)).auditEvents = [];
        Object.assign((/** @type {?} */ (this)).auditEvents, input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    StateAuditEventList.prototype.auditEvents;
}
export class StateTrackerData {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    StateTrackerData.prototype.possibleStates;
    /** @type {?} */
    StateTrackerData.prototype.currentState;
    /** @type {?} */
    StateTrackerData.prototype.phase;
    /** @type {?} */
    StateTrackerData.prototype.caseAudit;
}
export class TrackerState {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    TrackerState.prototype.name;
    /** @type {?} */
    TrackerState.prototype.label;
    /** @type {?} */
    TrackerState.prototype.phase;
    /** @type {?} */
    TrackerState.prototype.previousPhase;
    /** @type {?} */
    TrackerState.prototype.isTerminal;
    /** @type {?} */
    TrackerState.prototype.status;
    /** @type {?} */
    TrackerState.prototype.user;
    /** @type {?} */
    TrackerState.prototype.changed;
}
export class StateTracker {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        (/** @type {?} */ (this)).states = [];
        Object.assign((/** @type {?} */ (this)).states, input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    StateTracker.prototype.states;
    /** @type {?} */
    StateTracker.prototype.valid;
}
