// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/models/tc-document.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class DocumentAction {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    DocumentAction.prototype.action;
    /** @type {?} */
    DocumentAction.prototype.document;
}
export class Document {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Document.prototype.artifactCheckSum;
    /** @type {?} */
    Document.prototype.artifactRef;
    /** @type {?} */
    Document.prototype.artifactVersion;
    /** @type {?} */
    Document.prototype.author;
    /** @type {?} */
    Document.prototype.authorDetails;
    /** @type {?} */
    Document.prototype.creationDate;
    /** @type {?} */
    Document.prototype.description;
    /** @type {?} */
    Document.prototype.id;
    /** @type {?} */
    Document.prototype.lastModifiedBy;
    /** @type {?} */
    Document.prototype.lastModifiedByDetails;
    /** @type {?} */
    Document.prototype.lastModifiedDate;
    /** @type {?} */
    Document.prototype.mimeType;
    /** @type {?} */
    Document.prototype.name;
    /** @type {?} */
    Document.prototype.ownerApp;
    /** @type {?} */
    Document.prototype.size;
    /** @type {?} */
    Document.prototype.fileSize;
    /** @type {?} */
    Document.prototype.extension;
    /** @type {?} */
    Document.prototype.fileIcon;
}
export class DocumentList {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        (/** @type {?} */ (this)).documents = [];
        Object.assign((/** @type {?} */ (this)).documents, input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    DocumentList.prototype.documents;
}
export class OrgFolder {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    OrgFolder.prototype.id;
    /** @type {?} */
    OrgFolder.prototype.name;
    /** @type {?} */
    OrgFolder.prototype.owner;
    /** @type {?} */
    OrgFolder.prototype.ownerSub;
    /** @type {?} */
    OrgFolder.prototype.ownerSandbox;
    /** @type {?} */
    OrgFolder.prototype.creationDate;
    /** @type {?} */
    OrgFolder.prototype.lastModifiedDate;
    /** @type {?} */
    OrgFolder.prototype.lastModifiedBy;
    /** @type {?} */
    OrgFolder.prototype.publishedVersion;
    /** @type {?} */
    OrgFolder.prototype.latestVersion;
    /** @type {?} */
    OrgFolder.prototype.publisedVersionId;
    /** @type {?} */
    OrgFolder.prototype.latestVersionId;
    /** @type {?} */
    OrgFolder.prototype.checksum;
    /** @type {?} */
    OrgFolder.prototype.extRef;
    /** @type {?} */
    OrgFolder.prototype.tags;
    /** @type {?} */
    OrgFolder.prototype.descriptor;
}
export class DocTags {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    DocTags.prototype.id;
    /** @type {?} */
    DocTags.prototype.name;
    /** @type {?} */
    DocTags.prototype.ownerSub;
}
