// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/models/tc-groups-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class Group {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Group.prototype.name;
    /** @type {?} */
    Group.prototype.description;
    /** @type {?} */
    Group.prototype.type;
    /** @type {?} */
    Group.prototype.id;
}
export class Groups {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Groups.prototype.groups;
}
export class Roles {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    Roles.prototype.roles;
}
export class RouteAccessDef {
}
if (false) {
    /** @type {?} */
    RouteAccessDef.prototype.routeUrl;
    /** @type {?} */
    RouteAccessDef.prototype.requiredRoleId;
}
export class ButtonAccessDef {
}
if (false) {
    /** @type {?} */
    ButtonAccessDef.prototype.buttonId;
    /** @type {?} */
    ButtonAccessDef.prototype.requiredRoleId;
}
export class RouteAccessControlConfig {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    RouteAccessControlConfig.prototype.routes;
    /** @type {?} */
    RouteAccessControlConfig.prototype.buttons;
}
