// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/models/tc-live-apps-reporting.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class CaseTypeReportRecordInfo {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.id;
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.label;
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.name;
    /** @type {?} */
    CaseTypeReportRecordInfo.prototype.color;
}
export class CaseTypeReportRecord {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    CaseTypeReportRecord.prototype.activeStateCaseCount;
    /** @type {?} */
    CaseTypeReportRecord.prototype.applicationId;
    /** @type {?} */
    CaseTypeReportRecord.prototype.terminalStateCaseCount;
    /** @type {?} */
    CaseTypeReportRecord.prototype.incTerminal;
    /** @type {?} */
    CaseTypeReportRecord.prototype.caseTypeInfo;
}
export class CaseTypesReport {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    CaseTypesReport.prototype.caseTypes;
}
export class CaseTypeStateReportStateInfo {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.id;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.label;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.value;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.color;
    /** @type {?} */
    CaseTypeStateReportStateInfo.prototype.isTerminal;
}
export class CaseTypeStateReportRecord {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    CaseTypeStateReportRecord.prototype.caseCount;
    /** @type {?} */
    CaseTypeStateReportRecord.prototype.stateInfo;
}
export class CaseTypeStateReport {
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    deserialize(input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    CaseTypeStateReport.prototype.caseStates;
}
