// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-login/live-apps-login.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Note: This may move to the core library
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { LoginContext } from '../../models/liveappsdata';
import { LoginPrefill } from '@tibco-tcstk/tc-core-lib';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
/**
 * Component perform a Login in case there is no valid Session yet.
 *
 * If the user is not logged in the login component will be displayed automatically.
 *
 * ![alt-text](../Login.png "Image")
 */
export class LiveAppsLoginComponent extends LiveAppsComponent {
    constructor() {
        super();
        this.loginContext = new EventEmitter();
        // run when logged in
        this.handleLoggedIn = (/**
         * @param {?} loginInfo
         * @return {?}
         */
        (loginInfo) => {
            sessionStorage.setItem('loggedIn', Date.now().toString());
            // emit useful details about the login and session/claims
            this.loginContext.emit(new LoginContext().deserialize({
                authInfo: loginInfo.authInfo,
                accessToken: loginInfo.accessToken
            }));
        });
    }
}
LiveAppsLoginComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-login',
                template: "<tc-tibco-cloud-login fxFill (loggedIn)=\"handleLoggedIn($event)\" [loginPrefill]=\"loginPrefill\"></tc-tibco-cloud-login>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
LiveAppsLoginComponent.ctorParameters = () => [];
LiveAppsLoginComponent.propDecorators = {
    loginContext: [{ type: Output }],
    loginPrefill: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    LiveAppsLoginComponent.prototype.loginContext;
    /** @type {?} */
    LiveAppsLoginComponent.prototype.loginPrefill;
    /** @type {?} */
    LiveAppsLoginComponent.prototype.handleLoggedIn;
}
