// @tibco/spotfire-wrapper/esm2015/lib/lazy-loading-library.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Copyright (c) 2018-2018. TIBCO Software Inc. All Rights Reserved. Confidential & Proprietary.
import { Injectable, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { ReplaySubject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
// https://stackoverflow.com/questions/46240293/how-to-lazyload-library-in-angular-4-module
export class LazyLoadingLibraryService {
    /**
     * @param {?} document
     */
    constructor(document) {
        this.document = document;
        this.loadedLibraries = {};
    }
    /**
     * @param {?} url
     * @return {?}
     */
    loadJs(url) {
        if (this.loadedLibraries[url]) {
            return this.loadedLibraries[url].asObservable();
        }
        this.loadedLibraries[url] = new ReplaySubject();
        /** @type {?} */
        const script = this.document.createElement('script');
        script.type = 'text/javascript';
        script.src = url;
        script.onload = (/**
         * @return {?}
         */
        () => {
            this.loadedLibraries[url].next('');
            this.loadedLibraries[url].complete();
        });
        script.onerror = (/**
         * @return {?}
         */
        () => {
            console.error(`Library ${url} is not loaded !`);
            this.loadedLibraries[url].error(`Cannot load Spotfire JS library from ${url}. Check the URL !`);
        });
        this.document.body.appendChild(script);
        return this.loadedLibraries[url].asObservable();
    }
}
LazyLoadingLibraryService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
LazyLoadingLibraryService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
/** @nocollapse */ LazyLoadingLibraryService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LazyLoadingLibraryService_Factory() { return new LazyLoadingLibraryService(i0.ɵɵinject(i1.DOCUMENT)); }, token: LazyLoadingLibraryService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    LazyLoadingLibraryService.prototype.loadedLibraries;
    /**
     * @type {?}
     * @private
     */
    LazyLoadingLibraryService.prototype.document;
}
