// @tibco/spotfire-wrapper/esm2015/lib/spotfire-customization.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Copyright (c) 2018-2018. TIBCO Software Inc. All Rights Reserved. Confidential & Proprietary.
export class SpotfireCustomization {
    /**
     * @param {?=} vars
     */
    constructor(vars) {
        this.showAbout = false;
        this.showAnalysisInformationTool = false;
        this.showAuthor = false; // this enable 'edit' button.
        // this enable 'edit' button.
        this.showClose = false;
        this.showCustomizableHeader = false;
        this.showDodPanel = false; // Details-on-Demand panel
        // Details-on-Demand panel
        this.showExportFile = false;
        // showExportVisualization = false; marked as obsolete since Spotfire X
        this.showFilterPanel = false;
        this.showHelp = false;
        this.showLogout = false;
        this.showPageNavigation = false;
        // showAnalysisInfo = false;  marked as obsolete since Spotfire X
        this.showReloadAnalysis = false;
        this.showStatusBar = false;
        this.showToolBar = false;
        this.showUndoRedo = false;
        if (vars) {
            Object.keys(vars).forEach((/**
             * @param {?} key
             * @return {?}
             */
            key => this[key] = vars[key]));
        }
    }
}
if (false) {
    /** @type {?} */
    SpotfireCustomization.prototype.showAbout;
    /** @type {?} */
    SpotfireCustomization.prototype.showAnalysisInformationTool;
    /** @type {?} */
    SpotfireCustomization.prototype.showAuthor;
    /** @type {?} */
    SpotfireCustomization.prototype.showClose;
    /** @type {?} */
    SpotfireCustomization.prototype.showCustomizableHeader;
    /** @type {?} */
    SpotfireCustomization.prototype.showDodPanel;
    /** @type {?} */
    SpotfireCustomization.prototype.showExportFile;
    /** @type {?} */
    SpotfireCustomization.prototype.showFilterPanel;
    /** @type {?} */
    SpotfireCustomization.prototype.showHelp;
    /** @type {?} */
    SpotfireCustomization.prototype.showLogout;
    /** @type {?} */
    SpotfireCustomization.prototype.showPageNavigation;
    /** @type {?} */
    SpotfireCustomization.prototype.showReloadAnalysis;
    /** @type {?} */
    SpotfireCustomization.prototype.showStatusBar;
    /** @type {?} */
    SpotfireCustomization.prototype.showToolBar;
    /** @type {?} */
    SpotfireCustomization.prototype.showUndoRedo;
}
class SpotfireFilterSetting {
    /**
     * @param {?=} vars
     */
    constructor(vars) {
        this.values = [];
        this.lowValue = null;
        this.highValue = null;
        this.includeEmpty = true;
        this.searchPattern = null;
        this.hierarchyPaths = [];
        if (vars) {
            Object.keys(vars).forEach((/**
             * @param {?} key
             * @return {?}
             */
            key => this[key] = vars[key]));
        }
    }
}
if (false) {
    /** @type {?} */
    SpotfireFilterSetting.prototype.values;
    /** @type {?} */
    SpotfireFilterSetting.prototype.lowValue;
    /** @type {?} */
    SpotfireFilterSetting.prototype.highValue;
    /** @type {?} */
    SpotfireFilterSetting.prototype.includeEmpty;
    /** @type {?} */
    SpotfireFilterSetting.prototype.searchPattern;
    /** @type {?} */
    SpotfireFilterSetting.prototype.hierarchyPaths;
}
export class SpotfireFilter {
    /**
     * @param {?=} vars
     */
    constructor(vars) {
        if (vars) {
            Object.keys(vars).forEach((/**
             * @param {?} key
             * @return {?}
             */
            key => this[key] = vars[key]));
        }
    }
}
if (false) {
    /** @type {?} */
    SpotfireFilter.prototype.filteringSchemeName;
    /** @type {?} */
    SpotfireFilter.prototype.dataTableName;
    /** @type {?} */
    SpotfireFilter.prototype.dataColumnName;
    /** @type {?} */
    SpotfireFilter.prototype.filterType;
    /** @type {?} */
    SpotfireFilter.prototype.filterSettings;
}
