// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-app-configuration-widget/live-apps-app-configuration-widget.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * Manages liveapps settings for Config page
 *
 * \@example <tcla-live-apps-app-configuration-widget></tcla-live-apps-app-configuration-widget>
 */
export class LiveAppsAppConfigurationWidgetComponent {
    constructor() {
        /**
         * ~event configChanged : Configuration values changed
         * ~payload CaseCardConfig : CaseCardConfig object when configuration is changed (so called can do a save with data)
         */
        this.configChanged = new EventEmitter();
        /**
         * ~event appSelected : Case Type selected
         * ~payload CaseType : CaseType object of app selected
         */
        this.appSelected = new EventEmitter();
        this.handleConfigAppSelection = (/**
         * @param {?} application
         * @return {?}
         */
        (application) => {
            // handle selection of app to config
            this.selectedAppConfig = application;
            this.appSelected.emit(application);
        });
        this.handleConfigChanged = (/**
         * @param {?} caseCardConfig
         * @return {?}
         */
        (caseCardConfig) => {
            this.configChanged.emit(caseCardConfig);
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
LiveAppsAppConfigurationWidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-app-configuration-widget',
                template: "<div fxFill style=\"margin: 0px 20px 20px 20px\" fxLayout=\"column\">\n  <div fxLayout=\"row\" fxLayoutAlign=\"start center\">\n    <span>Configure Live App Summary Cards for:</span>\n    <tcla-live-apps-applications style=\"margin-left: 5px; margin-top: -1px\" [sandboxId]=\"sandboxId\" [appIds]=\"appIds\" [selectFirstApp]=\"true\" (selection)=\"handleConfigAppSelection($event)\"></tcla-live-apps-applications>\n  </div>\n  <div fxLayout=\"column\" fxFlex style=\"overflow: auto\">\n    <tcla-live-apps-application-configuration *ngIf=\"selectedAppConfig\" [sandboxId]=\"sandboxId\" [uiAppId]=\"uiAppId\" [appId]=\"selectedAppConfig.applicationId\" [folderId]=\"folderId\" [appTypeLabel]=\"selectedAppConfig.label\" (configChanged)=\"handleConfigChanged($event)\"></tcla-live-apps-application-configuration>\n  </div>\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
LiveAppsAppConfigurationWidgetComponent.ctorParameters = () => [];
LiveAppsAppConfigurationWidgetComponent.propDecorators = {
    appIds: [{ type: Input }],
    sandboxId: [{ type: Input }],
    uiAppId: [{ type: Input }],
    folderId: [{ type: Input }],
    configChanged: [{ type: Output }],
    appSelected: [{ type: Output }]
};
if (false) {
    /**
     * The list of LA Application IDs you want to handle
     * @type {?}
     */
    LiveAppsAppConfigurationWidgetComponent.prototype.appIds;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsAppConfigurationWidgetComponent.prototype.sandboxId;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsAppConfigurationWidgetComponent.prototype.uiAppId;
    /**
     * The organisation folder to store/retrieve documents
     * @type {?}
     */
    LiveAppsAppConfigurationWidgetComponent.prototype.folderId;
    /**
     * ~event configChanged : Configuration values changed
     * ~payload CaseCardConfig : CaseCardConfig object when configuration is changed (so called can do a save with data)
     * @type {?}
     */
    LiveAppsAppConfigurationWidgetComponent.prototype.configChanged;
    /**
     * ~event appSelected : Case Type selected
     * ~payload CaseType : CaseType object of app selected
     * @type {?}
     */
    LiveAppsAppConfigurationWidgetComponent.prototype.appSelected;
    /** @type {?} */
    LiveAppsAppConfigurationWidgetComponent.prototype.selectedAppConfig;
    /** @type {?} */
    LiveAppsAppConfigurationWidgetComponent.prototype.handleConfigAppSelection;
    /** @type {?} */
    LiveAppsAppConfigurationWidgetComponent.prototype.handleConfigChanged;
}
