// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/resolvers/role-active.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { of } from 'rxjs';
import { LiveAppsService } from '../services/live-apps.service';
import { GeneralConfigResolver, TcGeneralConfigService, TcSharedStateService } from '@tibco-tcstk/tc-core-lib';
import { flatMap, map } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { Location } from '@angular/common';
import { ClaimsResolver } from './claims.resolver';
import { TcRolesService } from '../services/tc-roles-service.ts.service';
export class RoleActiveResolver {
    /**
     * @param {?} rolesService
     * @param {?} liveAppsService
     * @param {?} route
     * @param {?} tcSharedStateService
     * @param {?} generalConfigService
     * @param {?} httpClient
     * @param {?} location
     */
    constructor(rolesService, liveAppsService, route, tcSharedStateService, generalConfigService, httpClient, location) {
        this.rolesService = rolesService;
        this.liveAppsService = liveAppsService;
        this.route = route;
        this.tcSharedStateService = tcSharedStateService;
        this.generalConfigService = generalConfigService;
        this.httpClient = httpClient;
        this.location = location;
    }
    /**
     * @return {?}
     */
    resolve() {
        /** @type {?} */
        const currentRole = this.rolesService.getCurrentRole();
        if (currentRole == undefined) {
            /** @type {?} */
            let generalConfigResolver = new GeneralConfigResolver(this.tcSharedStateService, this.generalConfigService, this.httpClient, this.location);
            /** @type {?} */
            const claimResolver$ = new ClaimsResolver(this.liveAppsService);
            return claimResolver$.resolve().pipe(flatMap((/**
             * @param {?} claim
             * @return {?}
             */
            claim => {
                /** @type {?} */
                const sandboxId = claim.primaryProductionSandbox.id;
                generalConfigResolver.setSandbox(Number(sandboxId));
                return generalConfigResolver.resolve().pipe(map((/**
                 * @param {?} generalConfig
                 * @return {?}
                 */
                generalConfig => {
                    /** @type {?} */
                    const highestRole = generalConfig.roles.reduce((/**
                     * @param {?} prev
                     * @param {?} current
                     * @return {?}
                     */
                    function (prev, current) {
                        return (prev.priority > current.priority) ? prev : current;
                    }));
                    this.rolesService.setCurrentRole(highestRole);
                    return highestRole;
                })));
            })));
        }
        else {
            return of(currentRole);
        }
    }
}
RoleActiveResolver.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RoleActiveResolver.ctorParameters = () => [
    { type: TcRolesService },
    { type: LiveAppsService },
    { type: ActivatedRoute },
    { type: TcSharedStateService },
    { type: TcGeneralConfigService },
    { type: HttpClient },
    { type: Location }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.rolesService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.liveAppsService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.route;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.tcSharedStateService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.generalConfigService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.httpClient;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.location;
}
