// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/resolvers/role-active.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { of } from 'rxjs';
import { LiveAppsService } from '../services/live-apps.service';
import { TcGeneralConfigService, TcSharedStateService } from '@tibco-tcstk/tc-core-lib';
import { map } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { Location } from '@angular/common';
import { TcRolesService } from '../services/tc-roles-service.ts.service';
import { RolesResolver } from './roles.resolver';
export class RoleActiveResolver {
    /**
     * @param {?} rolesService
     * @param {?} liveAppsService
     * @param {?} tcSharedStateService
     * @param {?} generalConfigService
     * @param {?} httpClient
     * @param {?} location
     * @param {?} router
     */
    constructor(rolesService, liveAppsService, tcSharedStateService, generalConfigService, httpClient, location, router) {
        this.rolesService = rolesService;
        this.liveAppsService = liveAppsService;
        this.tcSharedStateService = tcSharedStateService;
        this.generalConfigService = generalConfigService;
        this.httpClient = httpClient;
        this.location = location;
        this.router = router;
    }
    /**
     * @return {?}
     */
    resolve() {
        /** @type {?} */
        const currentRole = this.rolesService.getCurrentRole();
        if (currentRole === undefined) {
            /** @type {?} */
            const roleResolver$ = new RolesResolver(this.tcSharedStateService, this.generalConfigService, this.httpClient, this.liveAppsService, this.location, this.router);
            return roleResolver$.resolve().pipe(map((/**
             * @param {?} roles
             * @return {?}
             */
            roles => {
                /** @type {?} */
                const highestRole = roles.roles.reduce((/**
                 * @param {?} prev
                 * @param {?} current
                 * @return {?}
                 */
                function (prev, current) {
                    return (prev.priority > current.priority) ? prev : current;
                }));
                this.rolesService.setCurrentRole(highestRole);
                return highestRole;
            })));
        }
        else {
            return of(currentRole);
        }
    }
}
RoleActiveResolver.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RoleActiveResolver.ctorParameters = () => [
    { type: TcRolesService },
    { type: LiveAppsService },
    { type: TcSharedStateService },
    { type: TcGeneralConfigService },
    { type: HttpClient },
    { type: Location },
    { type: Router }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.rolesService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.liveAppsService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.tcSharedStateService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.generalConfigService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.httpClient;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.location;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.router;
}
