// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-case-data/live-apps-case-data.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { map, take, takeUntil } from 'rxjs/operators';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { TcCaseDataService } from '../../services/tc-case-data.service';
import { CustomFormDefs } from '@tibco-tcstk/tc-forms-lib';
/**
 * Displays data for a case in a widget (high level)
 *
 * ![alt-text](../live-apps-case-data.png "Image")
 *
 * \@example <tcla-live-apps-case-data></tcla-live-apps-case-data>
 */
var LiveAppsCaseDataComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsCaseDataComponent, _super);
    function LiveAppsCaseDataComponent(caseDataService) {
        var _this = _super.call(this) || this;
        _this.caseDataService = caseDataService;
        /**
         * not used
         */
        _this.customDataId = _this.customDataId ? _this.customDataId : 'default';
        _this.refresh = (/**
         * @return {?}
         */
        function () {
            _this.caseDataService.getCaseWithSchema(_this.caseRef, _this.sandboxId, _this.appId, _this.typeId, _this.uiAppId)
                .pipe(take(1), takeUntil(_this._destroyed$), map((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                _this.casedata = result.caseInfo.untaggedCasedataObj;
                _this.metadata = result.caseInfo.metadata;
                _this.summary = result.caseInfo.summaryObj;
                _this.schema = result.caseSchema;
                _this.formRef = result.applicationName + '.' + result.applicationInternalName + '.casedata.' + _this.customDataId;
            }))).subscribe(null, (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { _this.errorMessage = 'Error retrieving case data: ' + error.error.errorMsg; }));
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsCaseDataComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.refresh();
    };
    LiveAppsCaseDataComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-case-data',
                    template: "<div class=\"tcs-case-data-pane\" fxLayout=\"column\" fxFill style=\"overflow: hidden;\">\n  <tc-tibco-cloud-widget-header *ngIf=\"showHeader\" fxFlex=\"nogrow\" [headerText]=\"'Case Data'\" [icon]=\"'tcs-case-data-icon'\"></tc-tibco-cloud-widget-header>\n  <tcla-live-apps-case-data-display fxFlex=\"grow\" id=\"readOnlyDiv\" style=\"overflow:auto\" *ngIf=\"casedata && schema\" [schema]=\"schema\" [caseRef]=\"caseRef\" [appId]=\"appId\" [caseData]=\"casedata\" [layout]=\"layout\" [customFormDefs]=\"customFormDefs\" [formRef]=\"formRef\"></tcla-live-apps-case-data-display>\n</div>\n",
                    styles: [".tcs-case-data-pane{border-radius:3px;box-shadow:0 2px 4px 0 rgba(0,0,0,.33);background-color:#fff}.tcs-case-data-header{height:40px;border-radius:3px 3px 0 0;box-shadow:0 1px 2px 0 #dedede;padding-left:20px;padding-right:20px}.tcs-case-data-header-text{margin-left:10px;font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCaseDataComponent.ctorParameters = function () { return [
        { type: TcCaseDataService }
    ]; };
    LiveAppsCaseDataComponent.propDecorators = {
        caseRef: [{ type: Input }],
        appId: [{ type: Input }],
        typeId: [{ type: Input }],
        sandboxId: [{ type: Input }],
        uiAppId: [{ type: Input }],
        showHeader: [{ type: Input }],
        layout: [{ type: Input }],
        customFormDefs: [{ type: Input }],
        customDataId: [{ type: Input }]
    };
    return LiveAppsCaseDataComponent;
}(LiveAppsComponent));
export { LiveAppsCaseDataComponent };
if (false) {
    /**
     * The case reference
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.caseRef;
    /**
     * The LA Application Id
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.appId;
    /**
     * The LA Application Type Id (generally 1)
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.typeId;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.sandboxId;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.uiAppId;
    /**
     * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.showHeader;
    /**
     * Layout object that can be passed to override default layout of the form renderer
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.layout;
    /**
     * Custom Form configuration file
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.customFormDefs;
    /**
     * not used
     * @type {?}
     */
    LiveAppsCaseDataComponent.prototype.customDataId;
    /** @type {?} */
    LiveAppsCaseDataComponent.prototype.casedata;
    /** @type {?} */
    LiveAppsCaseDataComponent.prototype.summary;
    /** @type {?} */
    LiveAppsCaseDataComponent.prototype.metadata;
    /** @type {?} */
    LiveAppsCaseDataComponent.prototype.errorMessage;
    /** @type {?} */
    LiveAppsCaseDataComponent.prototype.schema;
    /** @type {?} */
    LiveAppsCaseDataComponent.prototype.formRef;
    /** @type {?} */
    LiveAppsCaseDataComponent.prototype.refresh;
    /**
     * @type {?}
     * @private
     */
    LiveAppsCaseDataComponent.prototype.caseDataService;
}
