// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-case-states/live-apps-case-states.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { take, takeUntil } from 'rxjs/operators';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { TcCaseStatesService } from '../../services/tc-case-states.service';
import { DurationSincePipe } from '@tibco-tcstk/tc-core-lib';
/**
 * Case States/Milestone view
 *
 * ![alt-text](../live-apps-case-states.png "Image")
 *
 * \@example <tcla-live-apps-case-states></tcla-live-apps-case-states>
 */
var LiveAppsCaseStatesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsCaseStatesComponent, _super);
    function LiveAppsCaseStatesComponent(caseStatesService, durationSince) {
        var _this = _super.call(this) || this;
        _this.caseStatesService = caseStatesService;
        _this.durationSince = durationSince;
        _this.getToolTipText = (/**
         * @param {?} trackerState
         * @return {?}
         */
        function (trackerState) {
            /** @type {?} */
            var toolTipText = '';
            if (trackerState.status != 'pending') {
                toolTipText = toolTipText + trackerState.user + ' ' + _this.durationSince.transform(trackerState.changed);
            }
            return toolTipText;
        });
        _this.refresh = (/**
         * @return {?}
         */
        function () {
            _this.caseStatesService.getTracker(_this.caseRef, _this.sandboxId, _this.appId).pipe(take(1), takeUntil(_this._destroyed$)).subscribe((/**
             * @param {?} tracker
             * @return {?}
             */
            function (tracker) {
                _this.tracker = tracker;
                if (!tracker.valid) {
                    console.error('Unable to create milestone trailer. Case Audit likely removed due to subscription retention period.');
                }
                return tracker;
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { _this.errorMessage = 'Error constructing state tracker: ' + error.error.errorMsg; }));
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsCaseStatesComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.refresh();
    };
    LiveAppsCaseStatesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-case-states',
                    template: "<div fxFlex class=\"starters-case-states-container\" fxLayout=\"row\" fxLayoutAlign=\"start center\">\n  <div *ngIf=\"tracker && tracker.valid\">\n    <span class=\"tc-milestone-title\" fxLayoutAlign=\"start center\">Milestones</span>\n    <div fxFlex class=\"starters-milestone-trailer\" fxLayout=\"row wrap\">\n      <div class=\"starters-ms-wrapper\" *ngFor=\"let trackerState of tracker.states; first as isFirst; last as isLast\" ngclass=\"{'starters-milestone-first-section': isFirst, 'starters-milestone-mid-section': (!isFirst && !isLast), 'starters-milestone-end-section': isLast}\">\n        <tcla-live-apps-milestone [isFirst]=\"isFirst\" [isLast]=\"isLast\" [isTerminal]=\"trackerState.isTerminal\" [label]=\"trackerState.label\" [phase]=\"trackerState.phase\" [previousPhase]=\"trackerState.previousPhase\" [status]=\"trackerState.status\" [matTooltip]=\"getToolTipText(trackerState)\" matTooltipShowDelay=\"1000\" matTooltipPosition=\"below\"></tcla-live-apps-milestone>\n      </div>\n    </div>\n  </div>\n  <div *ngIf=\"tracker && !tracker.valid\">\n    <span class=\"tc-milestone-title\">Milestone data not available</span>\n  </div>\n</div>\n\n\n",
                    styles: [".tc-milestone-title{min-width:150px;font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;color:#000}:host ::ng-deep tcla-live-apps-milestone .tc-ms-pending-bg{fill:#dedede}:host ::ng-deep tcla-live-apps-milestone .tc-ms-inprogress-bg{fill:#0081cb}:host ::ng-deep tcla-live-apps-milestone .tc-ms-completed-bg{fill:#062e79}:host ::ng-deep tcla-live-apps-milestone .tc-ms-completed-terminal-bg{fill:#04be5b}:host ::ng-deep tcla-live-apps-milestone .tc-ms-label{color:#000}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsCaseStatesComponent.ctorParameters = function () { return [
        { type: TcCaseStatesService },
        { type: DurationSincePipe }
    ]; };
    LiveAppsCaseStatesComponent.propDecorators = {
        appId: [{ type: Input }],
        sandboxId: [{ type: Input }],
        caseRef: [{ type: Input }]
    };
    return LiveAppsCaseStatesComponent;
}(LiveAppsComponent));
export { LiveAppsCaseStatesComponent };
if (false) {
    /**
     * The LA Application Id
     * @type {?}
     */
    LiveAppsCaseStatesComponent.prototype.appId;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsCaseStatesComponent.prototype.sandboxId;
    /**
     * The case reference
     * @type {?}
     */
    LiveAppsCaseStatesComponent.prototype.caseRef;
    /** @type {?} */
    LiveAppsCaseStatesComponent.prototype.states;
    /** @type {?} */
    LiveAppsCaseStatesComponent.prototype.tracker;
    /** @type {?} */
    LiveAppsCaseStatesComponent.prototype.errorMessage;
    /** @type {?} */
    LiveAppsCaseStatesComponent.prototype.getToolTipText;
    /** @type {?} */
    LiveAppsCaseStatesComponent.prototype.refresh;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsCaseStatesComponent.prototype.caseStatesService;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsCaseStatesComponent.prototype.durationSince;
}
