// @tibco-tcstk/tc-liveapps-lib/esm5/lib/services/tc-live-apps-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { Location } from '@angular/common';
import { TcSharedStateService, TcCoreCommonFunctions, SharedStateContent, SharedStateEntry, SharedStateList } from '@tibco-tcstk/tc-core-lib';
import { LiveAppsConfig } from '../models/tc-liveapps-config';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@tibco-tcstk/tc-core-lib";
var TcLiveAppsConfigService = /** @class */ (function () {
    function TcLiveAppsConfigService(location, sharedStateService) {
        this.location = location;
        this.sharedStateService = sharedStateService;
    }
    /**
     * @param {?} sandboxId
     * @param {?} uiAppId
     * @param {?} liveappsConfig
     * @return {?}
     */
    TcLiveAppsConfigService.prototype.createLiveAppsConfig = /**
     * @param {?} sandboxId
     * @param {?} uiAppId
     * @param {?} liveappsConfig
     * @return {?}
     */
    function (sandboxId, uiAppId, liveappsConfig) {
        /** @type {?} */
        var ssName = uiAppId + '.liveapps.config.tibcolabs.client.context.PUBLIC';
        /** @type {?} */
        var content = new SharedStateContent();
        content.json = TcCoreCommonFunctions.escapeString(JSON.stringify(liveappsConfig));
        return this.sharedStateService.createSharedState(ssName, 'PUBLIC', '', sandboxId, undefined, undefined, undefined, content)
            .pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return value; })));
    };
    /**
     * @param {?} uiAppId
     * @param {?} useCache
     * @param {?} flushCache
     * @return {?}
     */
    TcLiveAppsConfigService.prototype.getLiveAppsConfig = /**
     * @param {?} uiAppId
     * @param {?} useCache
     * @param {?} flushCache
     * @return {?}
     */
    function (uiAppId, useCache, flushCache) {
        // if useCache is false this will trigger the service to update the cached version with latest
        /** @type {?} */
        var ssName = uiAppId + '.liveapps.config.tibcolabs.client.context.PUBLIC';
        return this.sharedStateService.getSharedState(ssName, 'PUBLIC', useCache, flushCache)
            .pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value.sharedStateEntries.length > 0) {
                /** @type {?} */
                var ssresult = new LiveAppsConfig().deserialize(JSON.parse(value.sharedStateEntries[0].content.json));
                ssresult.id = value.sharedStateEntries[0].id;
                return ssresult;
            }
            else {
                return undefined;
            }
        })));
    };
    /**
     * @param {?} sandboxId
     * @param {?} uiAppId
     * @param {?} liveappsConfig
     * @param {?} id
     * @return {?}
     */
    TcLiveAppsConfigService.prototype.updateLiveAppsConfig = /**
     * @param {?} sandboxId
     * @param {?} uiAppId
     * @param {?} liveappsConfig
     * @param {?} id
     * @return {?}
     */
    function (sandboxId, uiAppId, liveappsConfig, id) {
        var _this = this;
        /** @type {?} */
        var ssName = uiAppId + '.liveapps.config.tibcolabs.client.context.PUBLIC';
        /** @type {?} */
        var content = new SharedStateContent();
        content.json = TcCoreCommonFunctions.escapeString(JSON.stringify(liveappsConfig));
        /** @type {?} */
        var entry = new SharedStateEntry();
        entry.content = content;
        entry.sandboxId = sandboxId;
        entry.name = ssName;
        entry.type = 'PUBLIC';
        entry.id = id;
        /** @type {?} */
        var ssList = new SharedStateList();
        ssList.sharedStateEntries = [];
        ssList.sharedStateEntries.push(entry);
        return this.sharedStateService.updateSharedState(ssList.sharedStateEntries)
            .pipe(map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            // flush the cache
            _this.getLiveAppsConfig(uiAppId, true, true).subscribe();
            return new LiveAppsConfig().deserialize((JSON.parse(value.sharedStateEntries[0].content.json)));
        })));
    };
    TcLiveAppsConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TcLiveAppsConfigService.ctorParameters = function () { return [
        { type: Location },
        { type: TcSharedStateService }
    ]; };
    /** @nocollapse */ TcLiveAppsConfigService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcLiveAppsConfigService_Factory() { return new TcLiveAppsConfigService(i0.ɵɵinject(i1.Location), i0.ɵɵinject(i2.TcSharedStateService)); }, token: TcLiveAppsConfigService, providedIn: "root" });
    return TcLiveAppsConfigService;
}());
export { TcLiveAppsConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TcLiveAppsConfigService.prototype.location;
    /**
     * @type {?}
     * @private
     */
    TcLiveAppsConfigService.prototype.sharedStateService;
}
