// @tibco-tcstk/tc-core-lib/esm5/lib/directives/on-create.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// simple directive that can be used to fire an event when a component is created
import { Directive, EventEmitter, Input, Output } from '@angular/core';
var OnCreateDirective = /** @class */ (function () {
    function OnCreateDirective() {
        this.fireCreated = new EventEmitter();
    }
    /**
     * @return {?}
     */
    OnCreateDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.fireCreated.emit();
    };
    OnCreateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[tcOnCreate]',
                    exportAs: 'onCreate'
                },] }
    ];
    /** @nocollapse */
    OnCreateDirective.ctorParameters = function () { return []; };
    OnCreateDirective.propDecorators = {
        idx: [{ type: Input }],
        fireCreated: [{ type: Output }]
    };
    return OnCreateDirective;
}());
export { OnCreateDirective };
if (false) {
    /** @type {?} */
    OnCreateDirective.prototype.idx;
    /** @type {?} */
    OnCreateDirective.prototype.fireCreated;
}
