// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-search-widget/live-apps-search-widget.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { CaseType } from '../../models/liveappsdata';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { LiveAppsCaseSearchComponent } from '../live-apps-case-search/live-apps-case-search.component';
import { LiveAppsService } from '../../services/live-apps.service';
/**
 * High Level search widget component (wraps others)
 * This Component allows to search for existing Cases and list Case Cards.
 *
 * ![alt-text](../live-apps-search-widget.png "")
 *
 * \@example <tcla-live-apps-search-widget></tcla-live-apps-search-widget>
 */
var LiveAppsSearchWidgetComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsSearchWidgetComponent, _super);
    function LiveAppsSearchWidgetComponent(liveapps) {
        var _this = _super.call(this) || this;
        _this.liveapps = liveapps;
        /**
         * ~event caseSelected : Case Clicked
         * ~payload string : emits case reference when a case is clicked (so parent can navigate to case)
         */
        _this.caseSelected = new EventEmitter();
        // case search
        _this.matchedRefs = [];
        // handle case search results
        _this.handleSearchResults = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.message = undefined;
            _this.matchedRefs = data.caserefs;
            _this.searchString = data.searchString;
        });
        _this.handleClearMatches = (/**
         * @return {?}
         */
        function () {
            _this.caseSearchComponent.clearResults();
            _this.matchedRefs = [];
            _this.message = undefined;
        });
        // case clicked
        _this.clickCaseAction = (/**
         * @param {?} caseReference
         * @return {?}
         */
        function (caseReference) {
            _this.caseSelected.emit(caseReference);
        });
        _this.refresh = (/**
         * @return {?}
         */
        function () {
            if (_this.caseSearchComponent) {
                _this.caseSearchComponent.refresh();
            }
        });
        _this.searchCasesByState = (/**
         * @param {?} stateId
         * @param {?} stateLabel
         * @param {?} appId
         * @param {?} typeId
         * @param {?} message
         * @return {?}
         */
        function (stateId, stateLabel, appId, typeId, message) {
            _this.caseSearchComponent.setCaseType(new CaseType().deserialize({ applicationId: appId, id: typeId }));
            _this.caseSearchComponent.setSelectedStateId(stateId, stateLabel);
            _this.message = message;
            _this.liveapps.caseSearchEntries('', _this.sandboxId, appId, typeId, true, 0, 1000, stateId).subscribe((/**
             * @param {?} results
             * @return {?}
             */
            function (results) {
                _this.matchedRefs = results.caserefs;
            }));
        });
        return _this;
    }
    LiveAppsSearchWidgetComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-search-widget',
                    template: "<div>\n  <div fxLayout=\"row\" style=\"margin-left: 50px; margin-right: 50px; margin-top: 10px; margin-bottom: 10px;\">\n    <tcla-live-apps-case-search fxFlex fxLayoutAlign=\"center center\" [sandboxId]=\"sandboxId\" [appIds]=\"appIds\" (foundRefs)=\"handleSearchResults($event)\"></tcla-live-apps-case-search>\n  </div>\n  <div *ngIf=\"matchedRefs.length>0\" fxLayout=\"row\" fxLayoutAlign=\"space-around start\" style=\"margin-left: 50px; margin-right: 50px; margin-top: 20px; margin-bottom: 20px\">\n      <tcla-live-apps-case-list fxFlex [sandboxId]=\"sandboxId\" [uiAppId]=\"uiAppId\" [caseRefs]=\"matchedRefs\" [displayType]=\"'list'\" [headerText]=\"'Matching Cases'\" [headerMessage]=\"message\" [highlight]=\"searchString\" (clickCase)=\"clickCaseAction($event)\" (clearMatches)=\"handleClearMatches()\"></tcla-live-apps-case-list>\n  </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsSearchWidgetComponent.ctorParameters = function () { return [
        { type: LiveAppsService }
    ]; };
    LiveAppsSearchWidgetComponent.propDecorators = {
        sandboxId: [{ type: Input }],
        uiAppId: [{ type: Input }],
        appIds: [{ type: Input }],
        caseSelected: [{ type: Output }],
        caseSearchComponent: [{ type: ViewChild, args: [LiveAppsCaseSearchComponent, { static: false },] }]
    };
    return LiveAppsSearchWidgetComponent;
}(LiveAppsComponent));
export { LiveAppsSearchWidgetComponent };
if (false) {
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsSearchWidgetComponent.prototype.sandboxId;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsSearchWidgetComponent.prototype.uiAppId;
    /**
     * The list of LA Application IDs you want to handle
     * @type {?}
     */
    LiveAppsSearchWidgetComponent.prototype.appIds;
    /**
     * ~event caseSelected : Case Clicked
     * ~payload string : emits case reference when a case is clicked (so parent can navigate to case)
     * @type {?}
     */
    LiveAppsSearchWidgetComponent.prototype.caseSelected;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.caseSearchComponent;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.matchedRefs;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.searchString;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.message;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.handleSearchResults;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.handleClearMatches;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.clickCaseAction;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.refresh;
    /** @type {?} */
    LiveAppsSearchWidgetComponent.prototype.searchCasesByState;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSearchWidgetComponent.prototype.liveapps;
}
