// @tibco-tcstk/tc-core-lib/esm2015/lib/components/tibco-cloud-settings-general/tibco-cloud-settings-general.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { TcGeneralConfigService } from '../../services/tc-general-config.service';
import { MatSnackBar } from '@angular/material';
/**
 * Manages general settings for Config page
 *
 * ![alt-text](../tibco-cloud-settings-general.png "Image")
 *
 * \@example <tc-tibco-cloud-settings-general></tc-tibco-cloud-settings-general>
 */
export class TibcoCloudSettingsGeneralComponent {
    /**
     * @param {?} route
     * @param {?} generalConfigService
     * @param {?} snackBar
     */
    constructor(route, generalConfigService, snackBar) {
        this.route = route;
        this.generalConfigService = generalConfigService;
        this.snackBar = snackBar;
        this.panelOpenState = false;
        this.runSaveFunction = (/**
         * @return {?}
         */
        () => {
            this.generalConfigService.updateGeneralConfig(this.sandboxId, this.generalConfig.uiAppId, this.generalConfig, this.generalConfig.id).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            result => {
                this.snackBar.open('General configuration saved', 'OK', {
                    duration: 3000
                });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                this.snackBar.open('Error saving general configuration saved', 'OK', {
                    duration: 3000
                });
            }));
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.generalConfig = this.route.snapshot.data.generalConfigHolder;
        this.claims = this.route.snapshot.data.claims;
        this.sandboxId = Number(this.claims.primaryProductionSandbox.id).valueOf();
        this.applicationTitle = this.generalConfig.applicationTitle;
        // this.roles = this.generalConfig.roles;
        this.displayName = this.generalConfig.displayName;
        this.documentationURL = this.generalConfig.documentationUrl;
    }
    /**
     * @protected
     * @return {?}
     */
    getRoute() {
        return this.route;
    }
}
TibcoCloudSettingsGeneralComponent.decorators = [
    { type: Component, args: [{
                selector: 'tc-tibco-cloud-settings-general',
                template: "<div fxLayout=\"column\" fxFill>\n  <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\"\n                                [headerText]=\"'General Configuration'\"></tc-tibco-cloud-widget-header>\n  <div fxFlex style=\"padding: 20px; overflow: hidden\" fxLayout=\"column\">\n    <div style=\"overflow: auto;\">\n      <div fxLayout=\"column\" fxFlex>\n        <p>You can some of the default Application Settings here.</p>\n        <br>\n        <!-- Application specifics -->\n        <div class=\"tcs-app-form-field\" fxLayout=\"row\">\n          <mat-form-field style=\"min-width: 400px\">\n            <input matInput placeholder=\"Browser Tab Title\" [(ngModel)]=\"generalConfig.browserTitle\">\n          </mat-form-field>\n        </div>\n        <div class=\"tcs-app-form-field\" fxLayout=\"row\">\n          <mat-form-field style=\"min-width: 400px\">\n            <input matInput placeholder=\"Application Title\" [(ngModel)]=\"generalConfig.applicationTitle\">\n          </mat-form-field>\n        </div>\n        <div class=\"tcs-app-form-field\" fxLayout=\"row\">\n          <mat-form-field style=\"min-width: 400px\">\n            <input matInput placeholder=\"Welcome Message\" [(ngModel)]=\"generalConfig.welcomeMessage\">\n          </mat-form-field>\n        </div>\n\n        <!-- Advanced Settings (all Optional from Shared State) -->\n        <mat-accordion>\n          <mat-expansion-panel>\n            <mat-expansion-panel-header>\n              <mat-panel-title>Advanced Settings</mat-panel-title>\n            </mat-expansion-panel-header>\n            <mat-form-field style=\"min-width: 400px\">\n              <input matInput placeholder=\"Application Documentation URL\" [(ngModel)]=\"generalConfig.documentationUrl\">\n            </mat-form-field>\n          </mat-expansion-panel>\n        </mat-accordion>\n      </div>\n\n    </div>\n    <div fxFlex class=\"tcs-filler-panel\"></div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n      <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n    </div>\n  </div>\n</div>\n",
                styles: [".tcs-app-form-field{min-height:57px}"]
            }] }
];
/** @nocollapse */
TibcoCloudSettingsGeneralComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: TcGeneralConfigService },
    { type: MatSnackBar }
];
if (false) {
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.applicationTitle;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.roles;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.displayName;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.documentationURL;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.panelOpenState;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.generalConfig;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.sandboxId;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.claims;
    /** @type {?} */
    TibcoCloudSettingsGeneralComponent.prototype.runSaveFunction;
    /**
     * @type {?}
     * @protected
     */
    TibcoCloudSettingsGeneralComponent.prototype.route;
    /**
     * @type {?}
     * @protected
     */
    TibcoCloudSettingsGeneralComponent.prototype.generalConfigService;
    /**
     * @type {?}
     * @protected
     */
    TibcoCloudSettingsGeneralComponent.prototype.snackBar;
}
