// @tibco-tcstk/tc-liveapps-lib/esm5/lib/resolvers/accessControlConfiguration.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import { TcCoreCommonFunctions, UiAppIdConfig, } from '@tibco-tcstk/tc-core-lib';
import { map, switchMap, mergeMap, flatMap } from 'rxjs/operators';
import { RouteAccessControlConfig } from '../models/tc-groups-data';
import { HttpClient } from '@angular/common/http';
import { Location } from '@angular/common';
import { TcAccessControlService } from '../services/tc-access-control.service';
var AccessControlConfigurationResolver = /** @class */ (function () {
    function AccessControlConfigurationResolver(location, http, accessControlService) {
        var _this = this;
        this.location = location;
        this.http = http;
        this.accessControlService = accessControlService;
        this.DEFAULT_CONFIG_URL = 'assets/config/routeAccessControl.json';
        this.APP_ID_URL = TcCoreCommonFunctions.prepareUrlForStaticResource(this.location, 'assets/config/uiAppId.json');
        // can be used to load defaultAppConfig from a JSON config
        this.getDefaultAppConfig = (/**
         * @return {?}
         */
        function () {
            return _this.http.get(TcCoreCommonFunctions.prepareUrlForStaticResource(_this.location, _this.DEFAULT_CONFIG_URL));
        });
        // loads uiAppId from json file in assets (appId.json)
        this.getAppId = (/**
         * @return {?}
         */
        function () {
            return _this.http.get(_this.APP_ID_URL).pipe(map((/**
             * @param {?} uiAppId
             * @return {?}
             */
            function (uiAppId) {
                /** @type {?} */
                var uiAppIdConfig = new UiAppIdConfig().deserialize(uiAppId);
                _this.uiAppId = uiAppIdConfig.uiAppId;
                return uiAppIdConfig;
            })));
        });
    }
    /**
     * @return {?}
     */
    AccessControlConfigurationResolver.prototype.resolve = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var accessConfig$ = this.getAppId().pipe(switchMap((/**
         * @param {?} uiAppId
         * @return {?}
         */
        function (uiAppId) { return _this.accessControlService.getAccessControlConfig(uiAppId.uiAppId, true, false)
            .pipe(mergeMap((/**
         * @param {?} accessControlConfig
         * @return {?}
         */
        function (accessControlConfig) {
            if (accessControlConfig === undefined) {
                return _this.getDefaultAppConfig().pipe(flatMap((/**
                 * @param {?} config
                 * @return {?}
                 */
                function (config) {
                    _this.defaultAccessControlConfig = new RouteAccessControlConfig().deserialize(config);
                    _this.defaultAccessControlConfig.uiAppId = _this.uiAppId;
                    return _this.accessControlService.createLAccessControlConfig(_this.sandboxId, _this.uiAppId, _this.defaultAccessControlConfig)
                        .pipe(flatMap((/**
                     * @param {?} result
                     * @return {?}
                     */
                    function (result) {
                        /** @type {?} */
                        var newAppConfig = _this.defaultAccessControlConfig;
                        newAppConfig.id = result;
                        return _this.accessControlService.updateAccessControlConfig(_this.sandboxId, _this.uiAppId, newAppConfig, result).pipe(flatMap((
                        // trigger a read to flush the cache since we changed it
                        // trigger a read to flush the cache since we changed it
                        /**
                         * @param {?} updatedConf
                         * @return {?}
                         */
                        function (updatedConf) {
                            return _this.accessControlService.getAccessControlConfig(_this.uiAppId, true, true).pipe(map((/**
                             * @param {?} cachedConfig
                             * @return {?}
                             */
                            function (cachedConfig) {
                                return cachedConfig;
                            })));
                        })));
                        // return newAppConfig;
                    })));
                })));
            }
            else {
                return of(accessControlConfig);
            }
        }))); })));
        return accessConfig$;
    };
    AccessControlConfigurationResolver.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AccessControlConfigurationResolver.ctorParameters = function () { return [
        { type: Location },
        { type: HttpClient },
        { type: TcAccessControlService }
    ]; };
    return AccessControlConfigurationResolver;
}());
export { AccessControlConfigurationResolver };
if (false) {
    /** @type {?} */
    AccessControlConfigurationResolver.prototype.DEFAULT_CONFIG_URL;
    /** @type {?} */
    AccessControlConfigurationResolver.prototype.APP_ID_URL;
    /**
     * @type {?}
     * @private
     */
    AccessControlConfigurationResolver.prototype.sandboxId;
    /** @type {?} */
    AccessControlConfigurationResolver.prototype.defaultAccessControlConfig;
    /**
     * @type {?}
     * @private
     */
    AccessControlConfigurationResolver.prototype.uiAppId;
    /**
     * @type {?}
     * @private
     */
    AccessControlConfigurationResolver.prototype.getDefaultAppConfig;
    /**
     * @type {?}
     * @private
     */
    AccessControlConfigurationResolver.prototype.getAppId;
    /**
     * @type {?}
     * @private
     */
    AccessControlConfigurationResolver.prototype.location;
    /**
     * @type {?}
     * @private
     */
    AccessControlConfigurationResolver.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AccessControlConfigurationResolver.prototype.accessControlService;
}
