// @tibco-tcstk/tc-liveapps-lib/esm5/lib/resolvers/roles.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { forkJoin } from 'rxjs';
import { LiveAppsService } from '../services/live-apps.service';
import { GeneralConfigResolver, TcGeneralConfigService, TcSharedStateService } from '@tibco-tcstk/tc-core-lib';
import { flatMap, map } from 'rxjs/operators';
import { Roles } from '../models/tc-groups-data';
import { HttpClient } from '@angular/common/http';
import { Location } from '@angular/common';
import { ClaimsResolver } from './claims.resolver';
var RolesResolver = /** @class */ (function () {
    function RolesResolver(sharedStateService, generalConfigService, http, liveapps, location, router) {
        this.sharedStateService = sharedStateService;
        this.generalConfigService = generalConfigService;
        this.http = http;
        this.liveapps = liveapps;
        this.location = location;
        this.router = router;
        this.createRoles = (/**
         * @param {?} groups
         * @param {?} roles
         * @return {?}
         */
        function (groups, roles) {
            /** @type {?} */
            var calcRoles = [];
            roles.forEach((/**
             * @param {?} role
             * @return {?}
             */
            function (role) {
                /** @type {?} */
                var targetGroup = groups.find((/**
                 * @param {?} grp
                 * @return {?}
                 */
                function (grp) {
                    return (grp.name === role.group);
                }));
                if (targetGroup) {
                    calcRoles.push(role);
                }
            }));
            return new Roles().deserialize(new Roles().deserialize({ roles: calcRoles }));
        });
    }
    /**
     * @return {?}
     */
    RolesResolver.prototype.resolve = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // we will need the general config to understand the roles definition
        /** @type {?} */
        var generalConfigResolver = new GeneralConfigResolver(this.sharedStateService, this.generalConfigService, this.http, this.location, this.router);
        // aim is to return an object that only contains roles where the user is a member of the matching group
        /** @type {?} */
        var claimResolver$ = new ClaimsResolver(this.liveapps).resolve().pipe(flatMap((/**
         * @param {?} claiminfo
         * @return {?}
         */
        function (claiminfo) {
            /** @type {?} */
            var sandboxId = claiminfo.primaryProductionSandbox.id;
            generalConfigResolver.setSandbox(Number(sandboxId));
            /** @type {?} */
            var generalConfig$ = generalConfigResolver.resolve();
            /** @type {?} */
            var groups$ = _this.liveapps.getGroupMemberships(+claiminfo.primaryProductionSandbox.id, claiminfo.id, 1000, true);
            return forkJoin(generalConfig$, groups$).pipe(map((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), configData = _b[0], groupData = _b[1];
                return _this.createRoles(groupData.groups, configData.roles);
            })));
        })));
        return claimResolver$;
    };
    RolesResolver.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RolesResolver.ctorParameters = function () { return [
        { type: TcSharedStateService },
        { type: TcGeneralConfigService },
        { type: HttpClient },
        { type: LiveAppsService },
        { type: Location },
        { type: Router }
    ]; };
    return RolesResolver;
}());
export { RolesResolver };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RolesResolver.prototype.createRoles;
    /**
     * @type {?}
     * @private
     */
    RolesResolver.prototype.sharedStateService;
    /**
     * @type {?}
     * @private
     */
    RolesResolver.prototype.generalConfigService;
    /**
     * @type {?}
     * @private
     */
    RolesResolver.prototype.http;
    /**
     * @type {?}
     * @private
     */
    RolesResolver.prototype.liveapps;
    /**
     * @type {?}
     * @private
     */
    RolesResolver.prototype.location;
    /**
     * @type {?}
     * @private
     */
    RolesResolver.prototype.router;
}
