// @tibco-tcstk/tc-liveapps-lib/esm5/lib/resolvers/role-active.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { of } from 'rxjs';
import { LiveAppsService } from '../services/live-apps.service';
import { GeneralConfigResolver, TcGeneralConfigService, TcSharedStateService } from '@tibco-tcstk/tc-core-lib';
import { flatMap, map } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { Location } from '@angular/common';
import { ClaimsResolver } from './claims.resolver';
import { TcRolesService } from '../services/tc-roles-service.ts.service';
var RoleActiveResolver = /** @class */ (function () {
    function RoleActiveResolver(rolesService, liveAppsService, route, tcSharedStateService, generalConfigService, httpClient, location) {
        this.rolesService = rolesService;
        this.liveAppsService = liveAppsService;
        this.route = route;
        this.tcSharedStateService = tcSharedStateService;
        this.generalConfigService = generalConfigService;
        this.httpClient = httpClient;
        this.location = location;
    }
    /**
     * @return {?}
     */
    RoleActiveResolver.prototype.resolve = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var currentRole = this.rolesService.getCurrentRole();
        if (currentRole == undefined) {
            /** @type {?} */
            var generalConfigResolver_1 = new GeneralConfigResolver(this.tcSharedStateService, this.generalConfigService, this.httpClient, this.location);
            /** @type {?} */
            var claimResolver$ = new ClaimsResolver(this.liveAppsService);
            return claimResolver$.resolve().pipe(flatMap((/**
             * @param {?} claim
             * @return {?}
             */
            function (claim) {
                /** @type {?} */
                var sandboxId = claim.primaryProductionSandbox.id;
                generalConfigResolver_1.setSandbox(Number(sandboxId));
                return generalConfigResolver_1.resolve().pipe(map((/**
                 * @param {?} generalConfig
                 * @return {?}
                 */
                function (generalConfig) {
                    /** @type {?} */
                    var highestRole = generalConfig.roles.reduce((/**
                     * @param {?} prev
                     * @param {?} current
                     * @return {?}
                     */
                    function (prev, current) {
                        return (prev.priority > current.priority) ? prev : current;
                    }));
                    _this.rolesService.setCurrentRole(highestRole);
                    return highestRole;
                })));
            })));
        }
        else {
            return of(currentRole);
        }
    };
    RoleActiveResolver.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RoleActiveResolver.ctorParameters = function () { return [
        { type: TcRolesService },
        { type: LiveAppsService },
        { type: ActivatedRoute },
        { type: TcSharedStateService },
        { type: TcGeneralConfigService },
        { type: HttpClient },
        { type: Location }
    ]; };
    return RoleActiveResolver;
}());
export { RoleActiveResolver };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.rolesService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.liveAppsService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.route;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.tcSharedStateService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.generalConfigService;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.httpClient;
    /**
     * @type {?}
     * @private
     */
    RoleActiveResolver.prototype.location;
}
