// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/resolvers/access.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { forkJoin } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { Location } from '@angular/common';
import { TcAccessControlService } from '../services/tc-access-control.service';
import { AccessControlConfigurationResolver } from './accessControlConfiguration.resolver';
import { map } from 'rxjs/operators';
import { TcRolesService } from '../services/tc-roles-service.ts.service';
import { TcSharedStateService, TcGeneralConfigService } from '@tibco-tcstk/tc-core-lib';
import { LiveAppsService } from '../services/live-apps.service';
import { RoleActiveResolver } from './role-active.resolver';
export class AccessResolver {
    /**
     * @param {?} location
     * @param {?} http
     * @param {?} accessControlService
     * @param {?} rolesService
     * @param {?} liveapps
     * @param {?} route
     * @param {?} sharedStateService
     * @param {?} generalConfigService
     * @param {?} router
     */
    constructor(location, http, accessControlService, rolesService, liveapps, route, sharedStateService, generalConfigService, router) {
        this.location = location;
        this.http = http;
        this.accessControlService = accessControlService;
        this.rolesService = rolesService;
        this.liveapps = liveapps;
        this.route = route;
        this.sharedStateService = sharedStateService;
        this.generalConfigService = generalConfigService;
        this.router = router;
    }
    /**
     * @return {?}
     */
    resolve() {
        /** @type {?} */
        const accessControlConfiguration$ = new AccessControlConfigurationResolver(this.location, this.http, this.accessControlService).resolve();
        // we will need the active user role
        /** @type {?} */
        const activeRoleRes = new RoleActiveResolver(this.rolesService, this.liveapps, this.sharedStateService, this.generalConfigService, this.http, this.location, this.router);
        /** @type {?} */
        const activeRole$ = activeRoleRes.resolve();
        // run both in parallel then check access
        /** @type {?} */
        const accessControl$ = forkJoin(accessControlConfiguration$, activeRole$).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        ([accessConfig, activeRole]) => {
            return accessConfig.configuration.find((/**
             * @param {?} element
             * @return {?}
             */
            element => element.roleId === activeRole.id));
        })));
        return accessControl$;
    }
}
AccessResolver.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AccessResolver.ctorParameters = () => [
    { type: Location },
    { type: HttpClient },
    { type: TcAccessControlService },
    { type: TcRolesService },
    { type: LiveAppsService },
    { type: ActivatedRoute },
    { type: TcSharedStateService },
    { type: TcGeneralConfigService },
    { type: Router }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.location;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.accessControlService;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.rolesService;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.liveapps;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.route;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.sharedStateService;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.generalConfigService;
    /**
     * @type {?}
     * @private
     */
    AccessResolver.prototype.router;
}
