// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-recent-cases/live-apps-recent-cases.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { LiveAppsService } from '../../services/live-apps.service';
import { take, takeUntil } from 'rxjs/operators';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { TcCoreCommonFunctions } from '@tibco-tcstk/tc-core-lib';
/**
 * Recent cases widget, this Component list recent visited Cases.
 *
 * ![alt-text](../live-apps-recent-cases.png "")
 *
 * \@example <tcla-live-apps-recent-cases></tcla-live-apps-recent-cases>
 */
var LiveAppsRecentCasesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsRecentCasesComponent, _super);
    function LiveAppsRecentCasesComponent(liveapps) {
        var _this = _super.call(this) || this;
        _this.liveapps = liveapps;
        /**
         * case card format - list, card, miniCard, staticList (no click event)
         */
        _this.displayType = _this.displayType ? _this.displayType : 'miniCard'; // miniCard, card, list
        // miniCard, card, list
        /**
         * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
         */
        _this.showHeader = _this.showHeader ? _this.showHeader : true;
        /**
         * ~event clickCase : Case clicked
         * ~payload CaseRoute : CaseRoute object output when case is clicked so calling component can route accordingly - ie. route to case
         */
        _this.clickCase = new EventEmitter();
        _this.clickCaseAction = (/**
         * @param {?} caseRoute
         * @return {?}
         */
        function (caseRoute) {
            _this.clickCase.emit(caseRoute);
        });
        _this.refresh = (/**
         * @return {?}
         */
        function () {
            _this.recentCases = [];
            _this.liveapps.getRecentCases(_this.uiAppId, _this.sandboxId)
                .pipe(take(1), takeUntil(_this._destroyed$)).subscribe((/**
             * @param {?} next
             * @return {?}
             */
            function (next) {
                _this.recentCases = next.caseRefs || [];
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { _this.errorMessage = 'Error retrieving recent cases: ' + error.error.errorMsg; }));
        });
        _this.clearRecentCases = (/**
         * @return {?}
         */
        function () {
            // -1 will clear recent cases
            _this.liveapps.setRecentCase('-1', _this.uiAppId, _this.sandboxId);
            _this.recentCases = [];
        });
        _this.handleDeleted = (/**
         * @param {?} caseRef
         * @return {?}
         */
        function (caseRef) {
            _this.recentCases.splice(_this.recentCases.indexOf(caseRef), 1);
            _this.liveapps.unsetRecentCase(caseRef, _this.uiAppId, _this.sandboxId);
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsRecentCasesComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngAfterViewInit.call(this);
        this.cardWidthPct = TcCoreCommonFunctions.calcSummaryCardPct(this.widget);
        this.containerChanges$.subscribe((/**
         * @param {?} widget
         * @return {?}
         */
        function (widget) {
            _this.cardWidthPct = TcCoreCommonFunctions.calcSummaryCardPct(widget);
        }));
    };
    /**
     * @return {?}
     */
    LiveAppsRecentCasesComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.refresh();
    };
    LiveAppsRecentCasesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-recent-cases',
                    template: "<div #componentDiv class=\"tcs-case-recent-box\" fxLayout=\"column\" fxFill>\n  <div *ngIf=\"showHeader\" class=\"tcs-case-recent-header\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n    <div fxLayoutAlign=\"start center\">\n      <mat-icon class=\"tcs-icon tcs-recent-icon\" svgIcon=\"tcs-recent-icon\"></mat-icon>\n      <div class=\"tcs-case-recent-header-text\">Recent</div>\n    </div>\n    <mat-icon (click)=\"clearRecentCases()\" class=\"tcs-icon tcs-icon-active tcs-clear-recent-icon\" svgIcon=\"tcs-clear-icon\" matTooltip=\"Clear Recent Cases\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\"></mat-icon>\n  </div>\n  <div *ngIf=\"!showHeader\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n    <mat-icon style=\"margin: 5px\" (click)=\"clearRecentCases()\" class=\"tcs-icon tcs-icon-active tcs-clear-recent-icon\" svgIcon=\"tcs-clear-icon\" matTooltip=\"Clear Recent Cases\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\"></mat-icon>\n  </div>\n  <div *ngIf=\"recentCases.length > 0\"class=\"tcs-case-recent-flow-list\" fxLayout=\"column\" fxLayout=\"row wrap\">\n    <div class=\"tcs-case-recent-item-box\" *ngFor=\"let case of recentCases\" [fxFlex]=\"cardWidthPct\">\n      <tcla-live-apps-case-summary [uiAppId]=\"uiAppId\" [typeBar]=\"true\" [borderCard]=\"false\" [displayType]=\"displayType\" [sandboxId]=\"sandboxId\" [caseRef]=\"case\" (clickCase)=\"clickCaseAction($event)\" (deleted)=\"handleDeleted($event)\"></tcla-live-apps-case-summary>\n      <div class=\"tcs-case-recent-line\"></div>\n    </div>\n  </div>\n  <div *ngIf=\"!(recentCases.length > 0)\" fxLayout=\"row\" fxLayoutAlign=\"center start\" fxLayoutGap=\"10px\" style=\"margin: 20px;\">\n    <mat-icon [svgIcon]=\"'ic-no-cases-icon'\" style=\"height: 48px; width: 48px;\"></mat-icon>\n    <div style=\"height: 100%\" fxLayoutAlign=\"start center\">\n      <span class=\"tcs-no-item-text\">No cases found</span>\n    </div>\n  </div>\n</div>\n",
                    styles: [".tcs-case-recent-box{border-radius:3px;box-shadow:0 2px 8px 0 #dedede;background-color:#fff}.tcs-case-recent-header{min-height:40px;border-radius:3px 3px 0 0;box-shadow:0 1px 2px 0 #dedede;padding-left:20px;padding-right:20px}.tcs-case-recent-header-text{font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;text-align:left;color:#000;margin-left:10px}.tcs-case-recent-flow-list{margin:10px;overflow-y:auto;padding:4px}.tcs-case-recent-item-box{padding:5px}.tcs-icon.tcs-icon-active:hover{cursor:pointer}:host ::ng-deep .tcs-icon.tcs-icon-active:hover .svg-content{fill:#0081cb}.tcs-case-recent-line{padding:0;margin:4px 5px 0;border-bottom:1.1px solid #f4f4f4}.tcs-no-item-text{font-family:Source Sans Pro;font-size:16px;font-weight:400;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:normal;color:#b6b6b6}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsRecentCasesComponent.ctorParameters = function () { return [
        { type: LiveAppsService }
    ]; };
    LiveAppsRecentCasesComponent.propDecorators = {
        sandboxId: [{ type: Input }],
        uiAppId: [{ type: Input }],
        displayType: [{ type: Input }],
        showHeader: [{ type: Input }],
        clickCase: [{ type: Output }],
        componentDiv: [{ type: ViewChild, args: ['componentDiv', { static: false },] }]
    };
    return LiveAppsRecentCasesComponent;
}(LiveAppsComponent));
export { LiveAppsRecentCasesComponent };
if (false) {
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsRecentCasesComponent.prototype.sandboxId;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsRecentCasesComponent.prototype.uiAppId;
    /**
     * case card format - list, card, miniCard, staticList (no click event)
     * @type {?}
     */
    LiveAppsRecentCasesComponent.prototype.displayType;
    /**
     * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
     * @type {?}
     */
    LiveAppsRecentCasesComponent.prototype.showHeader;
    /**
     * ~event clickCase : Case clicked
     * ~payload CaseRoute : CaseRoute object output when case is clicked so calling component can route accordingly - ie. route to case
     * @type {?}
     */
    LiveAppsRecentCasesComponent.prototype.clickCase;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.componentDiv;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.recentCases;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.errorMessage;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.widget;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.cardWidthPct;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.clickCaseAction;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.refresh;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.clearRecentCases;
    /** @type {?} */
    LiveAppsRecentCasesComponent.prototype.handleDeleted;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsRecentCasesComponent.prototype.liveapps;
}
