// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-component/live-apps-component.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, HostListener, ViewChild, ViewChildren } from '@angular/core';
import { Observable, Subject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/internal/operators/distinctUntilChanged';
import { map } from 'rxjs/operators';
import { TcComponent } from '@tibco-tcstk/tc-core-lib';
/**
 * Generic component extended by others
 *
 * \@example <tcla-live-apps-component></tcla-live-apps-component>
 */
/**
 * Class Description of the Component
 */
export class LiveAppsComponent {
    constructor() {
        // use the _destroyed$/takeUntil pattern to avoid memory leaks if a response was never received
        this._destroyed$ = new Subject();
        this.resize = (/**
         * @return {?}
         */
        () => {
            // explicit size changes
            if (this.observer) {
                setTimeout((/**
                 * @param {?} handler
                 * @return {?}
                 */
                handler => {
                    this.observer.next(this.componentDiv.nativeElement.offsetWidth);
                }));
            }
            this.componentChildDivs.forEach((/**
             * @param {?} child
             * @return {?}
             */
            child => {
                if (child.resize) {
                    child.resize();
                }
            }));
        });
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onResize(event) {
        // window resize changes
        if (this.observer) {
            setTimeout((/**
             * @param {?} handler
             * @return {?}
             */
            handler => {
                this.observer.next(this.componentDiv.nativeElement.offsetWidth);
            }));
        }
    }
    /**
     * @return {?}
     */
    setupWidthObserver() {
        this.containerChanges$ = new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        observer => {
            if (!this.observer) {
                this.observer = observer;
            }
            // send initial value - need async so any flex items have resized
            setTimeout((/**
             * @param {?} handler
             * @return {?}
             */
            handler => {
                observer.next(this.componentDiv.nativeElement.offsetWidth);
            }));
        })).pipe(distinctUntilChanged(), map((/**
         * @param {?} width
         * @return {?}
         */
        width => {
            if (this.widget && width) {
                this.widget.width = width;
            }
            return this.widget;
        })));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // components that extend this can use the widget.width object to detect size of container by subscribing to containerChanges$
        // we detect div size changes using two methods:
        // 1: if window resizes
        // 2: if a manual resize of the element is triggered - this.resize() - (example - from a button in UI)
        this.setupWidthObserver();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        // on destroy
        this._destroyed$.next();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.widget = new TcComponent();
    }
}
LiveAppsComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-component',
                template: "<p>\n  New Live Apps Component\n</p>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
LiveAppsComponent.ctorParameters = () => [];
LiveAppsComponent.propDecorators = {
    componentDiv: [{ type: ViewChild, args: ['componentDiv', { static: false },] }],
    componentChildDivs: [{ type: ViewChildren, args: ['componentChildDiv',] }],
    onResize: [{ type: HostListener, args: ['window:resize', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    LiveAppsComponent.prototype.componentDiv;
    /** @type {?} */
    LiveAppsComponent.prototype.componentChildDivs;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsComponent.prototype._destroyed$;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsComponent.prototype.containerChanges$;
    /** @type {?} */
    LiveAppsComponent.prototype.widget;
    /**
     * @type {?}
     * @private
     */
    LiveAppsComponent.prototype.observer;
    /** @type {?} */
    LiveAppsComponent.prototype.resize;
}
