// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-actions/live-apps-actions.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { CustomFormDefs } from '@tibco-tcstk/tc-forms-lib';
import { LiveAppsCaseActionsComponent } from '../live-apps-case-actions/live-apps-case-actions.component';
/**
 * Not used by app but wraps action list and action execution.
 *
 * \@example <tcla-live-apps-actions></tcla-live-apps-actions>
 */
var LiveAppsActionsComponent = /** @class */ (function () {
    function LiveAppsActionsComponent() {
        var _this = this;
        /**
         * Allow override of forms framework
         * Options: bootstrap-4 or material-design
         */
        this.formsFramework = this.formsFramework ? this.formsFramework : 'material-design';
        /**
         * ~event caseActioned : Case action starterd (process started)
         * ~payload ProcessId : processId of started process in live apps (action)
         */
        this.caseActioned = new EventEmitter();
        // action clicked
        this.handleActionCompleted = (/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            _this.caseActioned.emit(result);
            _this.actionSelector.toggleEnable();
        });
        this.handleActionClicked = (/**
         * @param {?} action
         * @return {?}
         */
        function (action) {
            _this.actionSelector.toggleEnable();
            _this.selectedAction = action;
        });
    }
    /**
     * @return {?}
     */
    LiveAppsActionsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    LiveAppsActionsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-actions',
                    template: "<div fxLayout=\"column\" fxFlexFill>\n  <tcla-live-apps-case-actions [sandboxId]=\"sandboxId\" [typeId]=\"typeId\" [appId]=\"applicationId\" [caseRef]=\"caseRef\" [maxActions]=\"5\" [actionFilter]=\"actionFilter\" (actionClicked)=\"handleActionClicked($event)\"></tcla-live-apps-case-actions>\n  <tcla-live-apps-case-action *ngIf=\"selectedAction\" [process]=\"selectedAction\" [caseRef]=\"caseRef\" [typeId]=\"typeId\" [applicationId]=\"applicationId\" [sandboxId]=\"sandboxId\" [formsFramework]=\"formsFramework\"></tcla-live-apps-case-action>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsActionsComponent.ctorParameters = function () { return []; };
    LiveAppsActionsComponent.propDecorators = {
        actionSelector: [{ type: ViewChild, args: [LiveAppsCaseActionsComponent, { static: false },] }],
        caseRef: [{ type: Input }],
        caseState: [{ type: Input }],
        sandboxId: [{ type: Input }],
        applicationId: [{ type: Input }],
        typeId: [{ type: Input }],
        customFormDefs: [{ type: Input }],
        formsFramework: [{ type: Input }],
        actionFilter: [{ type: Input }],
        caseActioned: [{ type: Output }]
    };
    return LiveAppsActionsComponent;
}());
export { LiveAppsActionsComponent };
if (false) {
    /** @type {?} */
    LiveAppsActionsComponent.prototype.actionSelector;
    /**
     * The case reference
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.caseRef;
    /**
     * The state of the case
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.caseState;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.sandboxId;
    /**
     * LA Application ID
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.applicationId;
    /**
     * The LA Application Type Id (generally 1)
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.typeId;
    /**
     * Custom Form configuration file
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.customFormDefs;
    /**
     * Allow override of forms framework
     * Options: bootstrap-4 or material-design
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.formsFramework;
    /**
     * Dont show buttons for any actions that start with this string.
     * eg: '$' will remove the action $Update
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.actionFilter;
    /**
     * ~event caseActioned : Case action starterd (process started)
     * ~payload ProcessId : processId of started process in live apps (action)
     * @type {?}
     */
    LiveAppsActionsComponent.prototype.caseActioned;
    /** @type {?} */
    LiveAppsActionsComponent.prototype.selectedAction;
    /** @type {?} */
    LiveAppsActionsComponent.prototype.handleActionCompleted;
    /** @type {?} */
    LiveAppsActionsComponent.prototype.handleActionClicked;
}
