// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-application-list/live-apps-application-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { LiveAppsApplicationsComponent } from '../live-apps-applications/live-apps-applications.component';
import { map, take, takeUntil } from 'rxjs/operators';
import { MatSelectionList } from '@angular/material';
/**
 * Special rendering of LiveAppsApplicationsComponent
 *
 * \@example <tcla-live-apps-application-list></tcla-live-apps-application-list>
 */
export class LiveAppsApplicationListComponent extends LiveAppsApplicationsComponent {
    constructor() {
        super(...arguments);
        /**
         * The app Ids selected (output)
         */
        this.selectedAppIds = [];
        /**
         * ~event appsSelected : Applications selected in component (CaseType objects)
         * ~payload CaseType[] : Array of CaseType objects of what was selected
         */
        this.appsSelected = new EventEmitter();
        /**
         * ~event appIdsSelected : Applications selected in component (appIds)
         * ~payload string[] : selected App Ids from the application list (used on config)
         */
        this.appIdsSelected = new EventEmitter();
        this.handleAppSelection = (/**
         * @param {?} selectionEvent
         * @return {?}
         */
        (selectionEvent) => {
            this.appsSelected.emit(selectionEvent);
            /** @type {?} */
            const selIds = [];
            selectionEvent.forEach((/**
             * @param {?} selected
             * @return {?}
             */
            (selected) => {
                selIds.push(selected.applicationId);
            }));
            this.appIdsSelected.emit(selIds);
        });
        this.handleSelectAll = (/**
         * @return {?}
         */
        () => {
            this.mySelectedOptions = this.applications.casetypes;
            this.appIdsSelected.emit(this.getAppIds(this.mySelectedOptions));
            this.appsSelected.emit(this.applications.casetypes);
        });
        this.handleDeselectAll = (/**
         * @return {?}
         */
        () => {
            this.mySelectedOptions = [];
            this.appIdsSelected.emit([]);
            this.appsSelected.emit([]);
        });
        this.getAppIds = (/**
         * @param {?} casetypes
         * @return {?}
         */
        (casetypes) => {
            /** @type {?} */
            const selIds = [];
            casetypes.forEach((/**
             * @param {?} selected
             * @return {?}
             */
            (selected) => {
                selIds.push(selected.applicationId);
            }));
            return selIds;
        });
        this.refresh = (/**
         * @param {?} bypassCache
         * @return {?}
         */
        (bypassCache) => {
            this.liveapps.getApplications(this.sandboxId, this.appIds, 100, bypassCache)
                .pipe(take(1), takeUntil(this._destroyed$), map((/**
             * @param {?} applicationList
             * @return {?}
             */
            applicationList => {
                this.applications = applicationList;
                // handle single app selection input
                if (this.selectedApp && this.selectedApp.applicationId) {
                    this.selectedApp = applicationList.casetypes.find((/**
                     * @param {?} casetype
                     * @return {?}
                     */
                    (casetype) => {
                        return casetype.applicationId === this.selectedApp.applicationId;
                    }));
                    // this.selection.emit(this.selectedApp);
                }
                else if (this.selectedAppIds && this.selectedAppIds.length > 0) {
                    // pre select any casetypes that were passed in the selectedAppIds input attribute
                    this.mySelectedOptions = this.applications.casetypes.filter((/**
                     * @param {?} casetype
                     * @return {?}
                     */
                    casetype => {
                        return (this.selectedAppIds.findIndex((/**
                         * @param {?} ct
                         * @return {?}
                         */
                        (ct) => {
                            return ct === casetype.applicationId;
                        })) !== -1);
                    }));
                }
            })))
                .subscribe(null, (/**
             * @param {?} error
             * @return {?}
             */
            error => { this.errorMessage = 'Error retrieving applications: ' + error.error.errorMsg; }));
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        // only interested in initial selection - catching every change would create an infinite loop!
        if (changes.selectedAppIds && (changes.selectedAppIds.firstChange)) {
            this.refresh(false);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // leave this blank so superclass ngOnInit not called
    }
}
LiveAppsApplicationListComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-application-list',
                template: "<div fxFlex style=\"padding: 10px;\" fxLayout=\"column\">\n  <div fxLayout=\"row\" class=\"tcs-application-list-buttons\">\n    <button mat-button (click)=\"handleSelectAll()\">\n      <span>Select All</span>\n    </button>\n    <button mat-button (click)=\"handleDeselectAll()\">\n      <span>Deselect All</span>\n    </button>\n  </div>\n  <mat-selection-list class=\"tcs-application-list\" #selectedApps [(ngModel)]=\"mySelectedOptions\" (ngModelChange)=\"handleAppSelection($event)\"\n                      fxLayout=\"column\" fxLayout=\"row wrap\">\n    <mat-list-option class=\"tcs-application-option\" checkboxPosition=\"before\"\n                     *ngFor=\"let casetype of applications.casetypes\"\n                     [value]=\"casetype\">\n      <span>{{casetype.label}}</span>\n    </mat-list-option>\n  </mat-selection-list>\n</div>\n",
                styles: [".tcs-application-option{max-width:320px}.tcs-application-list{overflow:auto}.tcs-application-list-buttons{min-height:40px}"]
            }] }
];
LiveAppsApplicationListComponent.propDecorators = {
    appSelectionList: [{ type: ViewChild, args: [MatSelectionList, { static: false },] }],
    selectedAppIds: [{ type: Input }],
    appsSelected: [{ type: Output }],
    appIdsSelected: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LiveAppsApplicationListComponent.prototype.appSelectionList;
    /**
     * The app Ids selected (output)
     * @type {?}
     */
    LiveAppsApplicationListComponent.prototype.selectedAppIds;
    /**
     * ~event appsSelected : Applications selected in component (CaseType objects)
     * ~payload CaseType[] : Array of CaseType objects of what was selected
     * @type {?}
     */
    LiveAppsApplicationListComponent.prototype.appsSelected;
    /**
     * ~event appIdsSelected : Applications selected in component (appIds)
     * ~payload string[] : selected App Ids from the application list (used on config)
     * @type {?}
     */
    LiveAppsApplicationListComponent.prototype.appIdsSelected;
    /** @type {?} */
    LiveAppsApplicationListComponent.prototype.mySelectedOptions;
    /** @type {?} */
    LiveAppsApplicationListComponent.prototype.handleAppSelection;
    /** @type {?} */
    LiveAppsApplicationListComponent.prototype.handleSelectAll;
    /** @type {?} */
    LiveAppsApplicationListComponent.prototype.handleDeselectAll;
    /**
     * @type {?}
     * @private
     */
    LiveAppsApplicationListComponent.prototype.getAppIds;
    /** @type {?} */
    LiveAppsApplicationListComponent.prototype.refresh;
}
