// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-application-create-button/live-apps-application-create-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { LiveAppsApplicationsComponent } from '../live-apps-applications/live-apps-applications.component';
/**
 * Special rendering of LiveAppsApplicationsComponent
 *
 * ![alt-text](../live-apps-application-create-button.png "Image")
 *
 * \@example <tcla-live-apps-application-create-button></tcla-live-apps-application-create-button>
 */
export class LiveAppsApplicationCreateButtonComponent extends LiveAppsApplicationsComponent {
    constructor() {
        super(...arguments);
        this.handleCreatorAppSelection = (/**
         * @param {?} application
         * @return {?}
         */
        (application) => {
            this.selection.emit(application);
        });
        this.handleCreateCaseClick = (/**
         * @return {?}
         */
        () => {
            // console.log('Select Case: ' + this.applications.casetypes.length);
            if (this.applications.casetypes.length === 1) {
                // Open Case Directly
                // console.log('Opening Case Directly...');
                this.handleCreatorAppSelection(this.applications.casetypes[0]);
            }
        });
    }
}
LiveAppsApplicationCreateButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-application-create-button',
                template: "<button mat-button [matMenuTriggerFor]=\"creatorApplicationMenu\" (click)=\"handleCreateCaseClick()\">\n  <mat-icon [svgIcon]=\"'tcs-case-start-icon'\"></mat-icon>\n  <span class=\"tcs-case-create-button-text\"> Create Case</span>\n</button>\n<mat-menu #creatorApplicationMenu=\"matMenu\">\n  <button mat-menu-item *ngFor=\"let application of applications.casetypes\" (click)=\"handleCreatorAppSelection(application)\">{{application.applicationName}}</button>\n</mat-menu>\n",
                styles: [".tcs-case-create-button-text{font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:normal;text-align:center;color:#0081cb}"]
            }] }
];
if (false) {
    /** @type {?} */
    LiveAppsApplicationCreateButtonComponent.prototype.handleCreatorAppSelection;
    /** @type {?} */
    LiveAppsApplicationCreateButtonComponent.prototype.handleCreateCaseClick;
}
