// @tibco-tcstk/tc-core-lib/esm5/lib/components/tibco-cloud-new-element/tibco-cloud-new-element.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
var TibcoCloudNewElementComponent = /** @class */ (function () {
    function TibcoCloudNewElementComponent(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        this.idReadOnly = false;
        this.elementType = data.resourceType;
        this.idReadOnly = data.idReadOnly;
        this.id = data.id;
    }
    /**
     * @return {?}
     */
    TibcoCloudNewElementComponent.prototype.onCreate = /**
     * @return {?}
     */
    function () {
        this.dialogRef.close({ event: 'submit', id: this.id, name: this.name });
    };
    /**
     * @return {?}
     */
    TibcoCloudNewElementComponent.prototype.onCancel = /**
     * @return {?}
     */
    function () {
        this.dialogRef.close();
    };
    TibcoCloudNewElementComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tc-tibco-cloud-new-element',
                    template: "<mat-toolbar class=\"tcs-case-creation-dialog-toolbar\">\n  <div fxLayout=\"row\" fxLayoutAlign=\"space-between center\" fxFlex>\n    <span class=\"tcs-case-creation-dialog-toolbar-text\">Create New {{elementType}}</span>\n  </div>\n</mat-toolbar>\n<div class=\"tcs-dialog-outer-div\">\n    <div mat-dialog-content class=\"tcs-new-element-form\" fxLayout=\"column\" fxLayoutAlign=\"space-around none\">\n        <mat-form-field>\n            <input matInput placeholder=\"Id\" [(ngModel)]=\"id\" [readonly]=\"idReadOnly\">\n        </mat-form-field>\n        <mat-form-field>\n            <input matInput placeholder=\"Name\" [(ngModel)]=\"name\" required>\n        </mat-form-field>\n    </div>\n    <mat-dialog-actions class=\"tcs-dialog-actions\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n        <button mat-button mat-flat-button color=\"secondary\" (click)=\"onCancel()\">Cancel</button>\n        <button mat-button mat-flat-button color=\"primary\" [disabled]=\"!name || !id\" (click)=\"onCreate()\">Create</button>\n    </mat-dialog-actions>\n</div>\n",
                    styles: [".tcs-dialog-outer-div{height:100%;display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;padding:30px}.tcs-new-element-form{margin-top:5px}"]
                }] }
    ];
    /** @nocollapse */
    TibcoCloudNewElementComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    return TibcoCloudNewElementComponent;
}());
export { TibcoCloudNewElementComponent };
if (false) {
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.id;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.name;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.elementType;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.idReadOnly;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.dialogRef;
    /** @type {?} */
    TibcoCloudNewElementComponent.prototype.data;
}
