// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-settings-summary-cards/live-apps-settings-summary-cards.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { LiveAppsSettingsComponent } from '../live-apps-settings/live-apps-settings.component';
import { map, take, takeUntil } from 'rxjs/operators';
import { TcCaseCardConfigService } from '../../services/tc-case-card-config.service';
import { ActivatedRoute } from '@angular/router';
import { TcLiveAppsConfigService } from '../../services/tc-live-apps-config.service';
import { MatSnackBar } from '@angular/material';
/**
 * Configuration of summary cards
 *
 * ![alt-text](../live-apps-settings-summary-cards.png "")
 *
 * \@example <tcla-live-apps-settings-summary-cards></tcla-live-apps-settings-summary-cards>
 */
export class LiveAppsSettingsSummaryCardsComponent extends LiveAppsSettingsComponent {
    /**
     * @param {?} caseCardConfigService
     * @param {?} routeExt
     * @param {?} liveAppsConfigServiceExt
     * @param {?} snackBar
     */
    constructor(caseCardConfigService, routeExt, liveAppsConfigServiceExt, snackBar) {
        super(routeExt, liveAppsConfigServiceExt, snackBar);
        this.caseCardConfigService = caseCardConfigService;
        this.routeExt = routeExt;
        this.liveAppsConfigServiceExt = liveAppsConfigServiceExt;
        this.snackBar = snackBar;
        this.handleConfigChanged = (/**
         * @param {?} caseCardConfig
         * @return {?}
         */
        (caseCardConfig) => {
            this.caseCardConfig = caseCardConfig;
            console.log(caseCardConfig);
        });
        this.handleAppSelection = (/**
         * @param {?} application
         * @return {?}
         */
        (application) => {
            this.selectedApp = application;
        });
        this.runSaveFunction = (/**
         * @return {?}
         */
        () => {
            this.caseCardConfigService.updateCaseCardConfig(this.sandboxId, this.selectedApp.applicationId, this.generalConfig.uiAppId, this.caseCardConfig)
                .pipe(take(1), takeUntil(this._destroyed$), map((/**
             * @param {?} caseCardConfig
             * @return {?}
             */
            caseCardConfig => {
                this.caseCardConfig = caseCardConfig;
            }))).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            result => {
                this.snackBar.open('Summary cards settings saved', 'OK', {
                    duration: 3000
                });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => {
                this.snackBar.open('Error saving Summary Cards settings', 'OK', {
                    duration: 3000
                });
                console.log('Unable to update case card config: ' + error.errorMsg);
            }));
        });
    }
}
LiveAppsSettingsSummaryCardsComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-settings-summary-cards',
                template: "<div fxLayout=\"column\" class=\"tcs-live-apps-summary-cards-settings\" fxFill>\n  <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\" [headerText]=\"'Live Apps Summary Card Configuration'\"></tc-tibco-cloud-widget-header>\n\n  <div style=\"padding: 20px 20px 20px 20px;\" fxLayout=\"column\" fxFlex>\n    <div class=\"tcs-card-config-widget-pane\" fxFlex fxLayout=\"column\" fxLayoutAlign=\"start stretch\">\n      <tcla-live-apps-app-configuration-widget fxFlex=\"grow\" [sandboxId]=\"sandboxId\" [appIds]=\"liveAppsConfig.applicationIds\" [uiAppId]=\"generalConfig.uiAppId\" [folderId]=\"liveAppsConfig.caseIconsFolderId\" (configChanged)=\"handleConfigChanged($event)\" (appSelected)=\"handleAppSelection($event)\"></tcla-live-apps-app-configuration-widget>\n    </div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n      <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\" [disabled]=\"!caseCardConfig || !selectedApp\">Save</button>\n    </div>\n  </div>\n</div>\n",
                styles: [".tcs-card-config-widget-pane,.tcs-live-apps-summary-cards-settings{overflow:hidden}"]
            }] }
];
/** @nocollapse */
LiveAppsSettingsSummaryCardsComponent.ctorParameters = () => [
    { type: TcCaseCardConfigService },
    { type: ActivatedRoute },
    { type: TcLiveAppsConfigService },
    { type: MatSnackBar }
];
if (false) {
    /** @type {?} */
    LiveAppsSettingsSummaryCardsComponent.prototype.caseCardConfig;
    /** @type {?} */
    LiveAppsSettingsSummaryCardsComponent.prototype.selectedApp;
    /** @type {?} */
    LiveAppsSettingsSummaryCardsComponent.prototype.handleConfigChanged;
    /** @type {?} */
    LiveAppsSettingsSummaryCardsComponent.prototype.handleAppSelection;
    /** @type {?} */
    LiveAppsSettingsSummaryCardsComponent.prototype.runSaveFunction;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsSummaryCardsComponent.prototype.caseCardConfigService;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsSummaryCardsComponent.prototype.routeExt;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsSummaryCardsComponent.prototype.liveAppsConfigServiceExt;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsSummaryCardsComponent.prototype.snackBar;
}
