// @tibco-tcstk/tc-core-lib/esm2015/lib/components/tibco-cloud-select-table/tibco-cloud-select-table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
import { TibcoCloudTableComponent } from '../tibco-cloud-table/tibco-cloud-table.component';
import { LogService } from '../../services/tc-logging.service';
/**
 * A table that allows rows to be selected
 *
 * \@example <tc-tibco-cloud-select-table></tc-tibco-cloud-select-table>
 */
export class TibcoCloudSelectTableComponent extends TibcoCloudTableComponent {
    /**
     * @param {?} logger
     */
    constructor(logger) {
        super();
        this.logger = logger;
        /**
         * ~event selectedlines : the lines selected
         * ~payload any : The lines that are selected in the table by the user
         */
        this.selectedlines = new EventEmitter();
        this.selected = {};
        logger.info('Select Table Started... ');
    }
    /**
     * @param {?} element
     * @return {?}
     */
    highlight(element) {
        element.highlighted = !element.highlighted;
    }
    /**
     * @param {?} row
     * @return {?}
     */
    clicked(row) {
        // console.log(row);
        /** @type {?} */
        const myData = JSON.parse(JSON.stringify(this.dataSource.data));
        // console.log(this.dataSource);
        /** @type {?} */
        const selectedArray = new Array();
        for (const line of myData) {
            // console.log('s:' , line);
            if (line.highlighted) {
                delete line['hovered'];
                delete line['highlighted'];
                selectedArray.push(line);
            }
        }
        // console.log('Selected Array: ' , selectedArray);
        this.selected = JSON.stringify(selectedArray);
        this.selectedlines.emit(this.selected);
    }
}
TibcoCloudSelectTableComponent.decorators = [
    { type: Component, args: [{
                selector: 'tc-tibco-cloud-select-table',
                template: "<div class=\"mat-elevation-z8\">\n  <table id=\"{{id}}\" mat-table class=\"tcst-full-width-table\" [dataSource]=\"dataSource\" matSort aria-label=\"Elements\">\n    <ng-container *ngFor=\"let column of columns\" [matColumnDef]='column.columnDef'>\n      <mat-header-cell *matHeaderCellDef class=\"tcst-header\" mat-sort-header>{{camelCaseTW(column.header)}}</mat-header-cell>\n      <mat-cell *matCellDef=\"let row\" >{{ column.cell(row) }}</mat-cell>\n    </ng-container>\n    <mat-header-row *matHeaderRowDef=\"displayedColumns\"></mat-header-row>\n    <mat-row *matRowDef=\"let row; columns: displayedColumns;  let even = even;\"\n             [ngClass]=\"{hovered: row.hovered, highlighted: row.highlighted, tcstgray: even}\"\n             (click)=\"row.highlighted = !row.highlighted; clicked(row);\" (mouseover)=\"row.hovered = true\" (mouseout)=\"row.hovered = false\"></mat-row>\n\n\n  </table> <!-- *ngIf=\"dataSource.data.length > 10\" -->\n<!-- <div class='' ngClass=\"{'tcst-hide': dataSource.data.length < 10}\"> -->\n  <mat-paginator [ngClass]=\"{'tcst-hide': dataSource.data.length < 10}\"  #paginator\n                 [length]=\"dataSource.data.length\"\n                 [pageIndex]=\"0\"\n                 [pageSize]=\"10\"\n                 [pageSizeOptions]=\"[5, 10, 20, 25]\">\n  </mat-paginator>\n</div>\n<!--</div>-->\n",
                styles: [".tcst-full-width-table{width:100%}.mat-row.hovered{background:#eee;cursor:pointer}.mat-row.highlighted{background:#999}.tcst-header{background:#0081cb;color:#fff;padding:5px 30px;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center}.tcst-header:hover{background-color:#229be0}.tcstgray{background-color:#f5f5f5}.tcst-hide{display:none}"]
            }] }
];
/** @nocollapse */
TibcoCloudSelectTableComponent.ctorParameters = () => [
    { type: LogService }
];
TibcoCloudSelectTableComponent.propDecorators = {
    selectedlines: [{ type: Output }]
};
if (false) {
    /**
     * ~event selectedlines : the lines selected
     * ~payload any : The lines that are selected in the table by the user
     * @type {?}
     */
    TibcoCloudSelectTableComponent.prototype.selectedlines;
    /** @type {?} */
    TibcoCloudSelectTableComponent.prototype.selected;
    /**
     * @type {?}
     * @protected
     */
    TibcoCloudSelectTableComponent.prototype.logger;
}
