// @tibco-tcstk/tc-core-lib/esm2015/lib/directives/on-create.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// simple directive that can be used to fire an event when a component is created
import { Directive, EventEmitter, Input, Output } from '@angular/core';
export class OnCreateDirective {
    constructor() {
        this.fireCreated = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.fireCreated.emit();
    }
}
OnCreateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tcOnCreate]',
                exportAs: 'onCreate'
            },] }
];
/** @nocollapse */
OnCreateDirective.ctorParameters = () => [];
OnCreateDirective.propDecorators = {
    idx: [{ type: Input }],
    fireCreated: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    OnCreateDirective.prototype.idx;
    /** @type {?} */
    OnCreateDirective.prototype.fireCreated;
}
