// @tibco-tcstk/tc-core-lib/esm5/lib/services/tc-general-landing-page.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { TcGeneralLandingPageConfigService } from './tc-general-landing-page-config.service';
import * as i0 from "@angular/core";
import * as i1 from "./tc-general-landing-page-config.service";
var TcGeneralLandingPageService = /** @class */ (function () {
    function TcGeneralLandingPageService(generalLandingPageConfigService) {
        var _this = this;
        this.generalLandingPageConfigService = generalLandingPageConfigService;
        this.getLandingPage = (/**
         * @param {?} key
         * @param {?} uiAppId
         * @return {?}
         */
        function (key, uiAppId) {
            /** @type {?} */
            var landingPage = _this.generalLandingPageConfigService.getGeneralLandingPageConfig(uiAppId, true, true).pipe(map((/**
             * @param {?} landingPages
             * @return {?}
             */
            function (landingPages) {
                /** @type {?} */
                var page = landingPages.landingPage.filter((/**
                 * @param {?} element
                 * @return {?}
                 */
                function (element) { return element.key === key; }))[0];
                return page;
            })));
            return landingPage;
        });
        this.getLandingPageForRole = (/**
         * @param {?} roleId
         * @param {?} uiAppId
         * @return {?}
         */
        function (roleId, uiAppId) {
            return _this.generalLandingPageConfigService.getGeneralLandingPageConfig(uiAppId, true, true).pipe(map((/**
             * @param {?} landingPages
             * @return {?}
             */
            function (landingPages) {
                /** @type {?} */
                var candidatePages = landingPages.landingPage.filter((/**
                 * @param {?} element
                 * @return {?}
                 */
                function (element) { return element.roles.indexOf(roleId) != -1; }));
                switch (candidatePages.length) {
                    case 0:
                        return undefined;
                    case 1:
                        return candidatePages[0];
                    default:
                        return candidatePages[0];
                }
            })));
        });
    }
    TcGeneralLandingPageService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TcGeneralLandingPageService.ctorParameters = function () { return [
        { type: TcGeneralLandingPageConfigService }
    ]; };
    /** @nocollapse */ TcGeneralLandingPageService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcGeneralLandingPageService_Factory() { return new TcGeneralLandingPageService(i0.ɵɵinject(i1.TcGeneralLandingPageConfigService)); }, token: TcGeneralLandingPageService, providedIn: "root" });
    return TcGeneralLandingPageService;
}());
export { TcGeneralLandingPageService };
if (false) {
    /** @type {?} */
    TcGeneralLandingPageService.prototype.getLandingPage;
    /** @type {?} */
    TcGeneralLandingPageService.prototype.getLandingPageForRole;
    /**
     * @type {?}
     * @private
     */
    TcGeneralLandingPageService.prototype.generalLandingPageConfigService;
}
