// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-case-data-display/live-apps-case-data-display.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { JsonSchema } from '../../models/liveappsdata';
import { CustomFormDefs } from '@tibco-tcstk/tc-forms-lib';
/**
 * Renders case data in a form
 *
 * ![alt-text](../live-apps-case-data-display.png "Image")
 *
 * \@example <tcla-live-apps-case-data-display></tcla-live-apps-case-data-display>
 */
export class LiveAppsCaseDataDisplayComponent {
    constructor() {
        this.DEFAULT_CASE_DATA_LAYOUT = [
            '*',
            { type: 'submit', title: 'Save', condition: '1===2' },
            { type: 'actions', title: 'hello', condition: '1===2' }
        ];
        /**
         * Layout object that can be passed to override default layout of the form renderer
         */
        this.layout = this.layout ? this.layout : this.DEFAULT_CASE_DATA_LAYOUT;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // since this is the 'display' of case data we set to readonly and no validation
        // legacy makes the labels appear above the data
        this.options = {
            defaultOptions: {
                'readonly': true,
                'appearance': 'legacy'
            },
            'validateOnRender': false
        };
        console.log(this.schema);
    }
}
LiveAppsCaseDataDisplayComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-case-data-display',
                template: "<tcfrm-rendered-form *ngIf=\"schema && caseData && options\" [schema]=\"schema\" [layout]=\"layout\" [data]=\"caseData\" [caseRef]=\"caseRef\" [appId]=\"appId\" [options]=\"options\" [customFormDefs]=\"customFormDefs\" [formRef]=\"formRef\"></tcfrm-rendered-form>\n",
                styles: [":host ::ng-deep button{visibility:hidden}"]
            }] }
];
/** @nocollapse */
LiveAppsCaseDataDisplayComponent.ctorParameters = () => [];
LiveAppsCaseDataDisplayComponent.propDecorators = {
    caseData: [{ type: Input }],
    appId: [{ type: Input }],
    caseRef: [{ type: Input }],
    schema: [{ type: Input }],
    layout: [{ type: Input }],
    customFormDefs: [{ type: Input }],
    formRef: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    LiveAppsCaseDataDisplayComponent.prototype.DEFAULT_CASE_DATA_LAYOUT;
    /**
     * The case data
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.caseData;
    /**
     * The case reference
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.appId;
    /**
     * The applicationId
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.caseRef;
    /**
     * The schema of the case type
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.schema;
    /**
     * Layout object that can be passed to override default layout of the form renderer
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.layout;
    /**
     * Custom Form configuration file
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.customFormDefs;
    /**
     * Custom Form Reference
     * @type {?}
     */
    LiveAppsCaseDataDisplayComponent.prototype.formRef;
    /** @type {?} */
    LiveAppsCaseDataDisplayComponent.prototype.options;
}
