// @tibco-tcstk/tc-core-lib/esm5/lib/models/tc-login.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/* Models used by login and authorization services */
var Group = /** @class */ (function () {
    function Group(id, type) {
        this.id = id;
        this.type = type;
    }
    return Group;
}());
export { Group };
if (false) {
    /** @type {?} */
    Group.prototype.id;
    /** @type {?} */
    Group.prototype.type;
}
var Claim = /** @class */ (function () {
    function Claim() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Claim.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Claim;
}());
export { Claim };
if (false) {
    /** @type {?} */
    Claim.prototype.email;
    /** @type {?} */
    Claim.prototype.firstName;
    /** @type {?} */
    Claim.prototype.id;
    /** @type {?} */
    Claim.prototype.lastName;
    /** @type {?} */
    Claim.prototype.sandboxes;
    /** @type {?} */
    Claim.prototype.subscriptionId;
    /** @type {?} */
    Claim.prototype.username;
    /** @type {?} */
    Claim.prototype.primaryProductionSandbox;
}
var Sandbox = /** @class */ (function () {
    function Sandbox(groups, id, type, name, subscriptionId, ownerId) {
        this.groups = groups;
        this.id = id;
        this.type = type;
        this.name = name;
        this.subscriptionId = subscriptionId;
        this.ownerId = ownerId;
    }
    return Sandbox;
}());
export { Sandbox };
if (false) {
    /** @type {?} */
    Sandbox.prototype.groups;
    /** @type {?} */
    Sandbox.prototype.id;
    /** @type {?} */
    Sandbox.prototype.type;
    /** @type {?} */
    Sandbox.prototype.name;
    /** @type {?} */
    Sandbox.prototype.subscriptionId;
    /** @type {?} */
    Sandbox.prototype.ownerId;
}
var SandboxList = /** @class */ (function () {
    function SandboxList() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    SandboxList.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).sandboxes = [];
        Object.assign((/** @type {?} */ (this)).sandboxes, input);
        return (/** @type {?} */ (this));
    };
    return SandboxList;
}());
export { SandboxList };
if (false) {
    /** @type {?} */
    SandboxList.prototype.sandboxes;
}
var AccessToken = /** @class */ (function () {
    function AccessToken() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    AccessToken.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return AccessToken;
}());
export { AccessToken };
if (false) {
    /** @type {?} */
    AccessToken.prototype.access_token;
    /** @type {?} */
    AccessToken.prototype.token_type;
    /** @type {?} */
    AccessToken.prototype.expires_in;
}
var AuthInfo = /** @class */ (function () {
    function AuthInfo() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    AuthInfo.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return AuthInfo;
}());
export { AuthInfo };
if (false) {
    /** @type {?} */
    AuthInfo.prototype.userName;
    /** @type {?} */
    AuthInfo.prototype.firstName;
    /** @type {?} */
    AuthInfo.prototype.lastName;
    /** @type {?} */
    AuthInfo.prototype.userId;
    /** @type {?} */
    AuthInfo.prototype.ts;
    /** @type {?} */
    AuthInfo.prototype.orgName;
}
var Subscription = /** @class */ (function () {
    function Subscription(accountId, accountDisplayName, ownerName, loggedInUserRole, regions) {
        this.accountId = accountId;
        this.accountDisplayName = accountDisplayName;
        this.ownerName = ownerName;
        this.loggedInUserRole = loggedInUserRole;
        this.regions = regions;
    }
    return Subscription;
}());
export { Subscription };
if (false) {
    /** @type {?} */
    Subscription.prototype.accountId;
    /** @type {?} */
    Subscription.prototype.accountDisplayName;
    /** @type {?} */
    Subscription.prototype.ownerName;
    /** @type {?} */
    Subscription.prototype.loggedInUserRole;
    /** @type {?} */
    Subscription.prototype.regions;
}
var Owner = /** @class */ (function () {
    function Owner() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Owner.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Owner;
}());
export { Owner };
if (false) {
    /** @type {?} */
    Owner.prototype.email;
    /** @type {?} */
    Owner.prototype.firstName;
    /** @type {?} */
    Owner.prototype.lastName;
    /** @type {?} */
    Owner.prototype.role;
}
var AccountInfo = /** @class */ (function () {
    function AccountInfo() {
    }
    return AccountInfo;
}());
export { AccountInfo };
if (false) {
    /** @type {?} */
    AccountInfo.prototype.accountDisplayName;
    /** @type {?} */
    AccountInfo.prototype.accountId;
    /** @type {?} */
    AccountInfo.prototype.loggedInUserRole;
    /** @type {?} */
    AccountInfo.prototype.ownerInfo;
    /** @type {?} */
    AccountInfo.prototype.regions;
}
var AccountsInfo = /** @class */ (function () {
    function AccountsInfo() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    AccountsInfo.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        (/** @type {?} */ (this)).accountInfos = [];
        Object.assign((/** @type {?} */ (this)).accountInfos, input);
        return (/** @type {?} */ (this));
    };
    return AccountsInfo;
}());
export { AccountsInfo };
if (false) {
    /** @type {?} */
    AccountsInfo.prototype.accountInfos;
}
var LoginPrefill = /** @class */ (function () {
    function LoginPrefill() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    LoginPrefill.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return LoginPrefill;
}());
export { LoginPrefill };
if (false) {
    /** @type {?} */
    LoginPrefill.prototype.emailId;
    /** @type {?} */
    LoginPrefill.prototype.clientId;
}
