// @tibco-tcstk/tc-core-lib/esm2015/lib/components/tibco-cloud-configuration/tibco-cloud-configuration.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Router, ActivatedRoute } from '@angular/router';
import { Location } from '@angular/common';
import { TcButtonsHelperService } from '../../services/tc-buttons-helper.service';
import { RouteAction } from '../../models/tc-routing-actions';
/**
 * TIBCO Cloud Configuration Component
 *
 * ![alt-text](../tibco-cloud-configuration.png "")
 *
 * \@example <tc-tibco-cloud-configuration></tc-tibco-cloud-configuration>
 */
export class TibcoCloudConfigurationComponent {
    /**
     * @param {?} router
     * @param {?} route
     * @param {?} buttonsHelper
     * @param {?} location
     */
    constructor(router, route, buttonsHelper, location) {
        this.router = router;
        this.route = route;
        this.buttonsHelper = buttonsHelper;
        this.location = location;
        /**
         * ~event routeAction : Component requests route to another page
         * ~payload RouteAction : RouteAction object to tell caller to navigate somewhere
         */
        this.routeAction = new EventEmitter();
        this.createToolbarButtons = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const homeButton = this.buttonsHelper.createButton('close', 'tcs-close-icon', true, 'Close', true, true);
            /** @type {?} */
            const buttons = [homeButton];
            return buttons;
        });
        this.handleSelectionEvent = (/**
         * @param {?} id
         * @return {?}
         */
        (id) => {
            this.configName = id;
            /** @type {?} */
            const url = this.baseRoute + id.toLowerCase().split(' ').join('-');
            this.router.navigate([url]);
        });
        this.handleToolbarButtonEvent = (/**
         * @param {?} buttonId
         * @return {?}
         */
        (buttonId) => {
            if (buttonId === 'close') {
                this.routeAction.emit(new RouteAction('backClicked', null));
            }
        });
        this.showConfig = (/**
         * @param {?} option
         * @return {?}
         */
        (option) => {
            this.configName = option;
            console.log('Setting selected: ' + option);
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.toolbarButtons = this.createToolbarButtons();
        // this.configName = this.route.component;
    }
}
TibcoCloudConfigurationComponent.decorators = [
    { type: Component, args: [{
                selector: 'tc-tibco-cloud-configuration',
                template: "<div fxFill fxLayout=\"column\">\n  <div>\n    <tc-tibco-cloud-menu-bar [toolbarButtons]=\"toolbarButtons\" [headerText]=\"'Manage Capabilities'\"\n                             (toolbarButtonEvent)=\"handleToolbarButtonEvent($event)\">\n      <ng-content select=\"[customToolbarContent]\" fxFlex></ng-content>\n    </tc-tibco-cloud-menu-bar>\n  </div>\n\n  <div fxFlex fxLayout=\"row\" fxLayoutGap=\"20px\" style=\"margin: 20px;\">\n\n    <div fxLayout=\"column\" fxLayoutAlign=\"start start\" style=\"min-width: 220px;\" fxLayoutGap>\n      <tc-tibco-cloud-setting-menu-entry style=\"padding-bottom: 20px\" *ngFor=\"let menu of configMenuPages\"\n                                         [icon]=\"menu.icon\" [title]=\"menu.entry\" [options]=\"menu.options\"\n                                         (configureOption)=\"handleSelectionEvent($event)\"></tc-tibco-cloud-setting-menu-entry>\n    </div>\n\n    <!-- configure Box -->\n    <div fxFlex class=\"tcs-configuration-outer-pane\" fxLayout=\"column\" fxFlex>\n      <div class=\"tcs-configuration-router-pane\" fxFlex>\n        <router-outlet></router-outlet>\n      </div>\n    </div>\n\n    <div fxShow.lt-lg=\"false\" style=\"min-width: 100px;\"  fxLayout=\"column\" fxLayoutAlign=\"start center\">\n      <a href=\"https://community.tibco.com/wiki/tibco-labs\" target=\"_blank\">\n        <mat-icon style=\"height: 64px; width: 64px;\" [svgIcon]=\"'tibco-labs'\"></mat-icon>\n      </a>\n    </div>\n\n  </div>\n</div>\n",
                styles: [".starters-custom-layout-panel-text{line-height:1.5;letter-spacing:.3px;text-align:left;color:#212121;font-weight:600;font-size:18px;box-shadow:0 0 0 #000}.starters-custom-layout-box-header{height:40px;min-height:40px;border-radius:3px 3px 0 0;background-color:#fff;box-shadow:0 2px 0 0 #dedede}.starters-custom-layout-panel-icon{max-width:24px;max-height:24px;margin-left:8px;margin-right:8px;padding:4px}.tcs-configuration-router-pane{overflow:hidden}.tcs-configuration-outer-pane{border-radius:3px;box-shadow:0 2px 4px 0 rgba(0,0,0,.33);background-color:#fff;border:1px solid #dedede;overflow:hidden}"]
            }] }
];
/** @nocollapse */
TibcoCloudConfigurationComponent.ctorParameters = () => [
    { type: Router },
    { type: ActivatedRoute },
    { type: TcButtonsHelperService },
    { type: Location }
];
TibcoCloudConfigurationComponent.propDecorators = {
    baseRoute: [{ type: Input }],
    configMenuPages: [{ type: Input }],
    routeAction: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TibcoCloudConfigurationComponent.prototype.baseRoute;
    /** @type {?} */
    TibcoCloudConfigurationComponent.prototype.configMenuPages;
    /**
     * ~event routeAction : Component requests route to another page
     * ~payload RouteAction : RouteAction object to tell caller to navigate somewhere
     * @type {?}
     */
    TibcoCloudConfigurationComponent.prototype.routeAction;
    /** @type {?} */
    TibcoCloudConfigurationComponent.prototype.configName;
    /** @type {?} */
    TibcoCloudConfigurationComponent.prototype.toolbarButtons;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudConfigurationComponent.prototype.createToolbarButtons;
    /** @type {?} */
    TibcoCloudConfigurationComponent.prototype.handleSelectionEvent;
    /** @type {?} */
    TibcoCloudConfigurationComponent.prototype.handleToolbarButtonEvent;
    /** @type {?} */
    TibcoCloudConfigurationComponent.prototype.showConfig;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudConfigurationComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudConfigurationComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudConfigurationComponent.prototype.buttonsHelper;
    /**
     * @type {?}
     * @private
     */
    TibcoCloudConfigurationComponent.prototype.location;
}
