// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-home-cockpit/live-apps-home-cockpit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { TcButtonsHelperService, RouteAction } from '@tibco-tcstk/tc-core-lib';
import { LiveAppsFavoriteCasesComponent } from '../live-apps-favorite-cases/live-apps-favorite-cases.component';
import { LiveAppsRecentCasesComponent } from '../live-apps-recent-cases/live-apps-recent-cases.component';
import { LiveAppsSearchWidgetComponent } from '../live-apps-search-widget/live-apps-search-widget.component';
import { MatDialog } from '@angular/material';
import { LiveAppsCreatorDialogComponent } from '../live-apps-creator-dialog/live-apps-creator-dialog.component';
import { CaseCreatorSelectionContext } from '../../models/tc-case-creator';
import { LiveAppsNotesComponent } from '../live-apps-notes/live-apps-notes.component';
import { LiveAppsDocumentsComponent } from '../live-apps-documents/live-apps-documents.component';
import { TcRolesService } from '../../services/tc-roles-service.ts.service';
import { Roles, RouteAccessControlConfigurationElement } from '../../models/tc-groups-data';
import { LiveAppsActiveCasesWidgetComponent } from '../live-apps-active-cases-widget/live-apps-active-cases-widget.component';
import { CustomFormDefs } from '@tibco-tcstk/tc-forms-lib';
import { LiveAppsWorkitemsComponent } from '../live-apps-workitems/live-apps-workitems.component';
/**
 * High level component to allow home page view of system
 *
 * ![alt-text](../live-apps-home-cockpit.png "Documents Component Image")
 *
 * \@example <tcla-live-apps-home-cockpit></tcla-live-apps-home-cockpit>
 */
export class LiveAppsHomeCockpitComponent {
    /**
     * @param {?} buttonsHelper
     * @param {?} dialog
     * @param {?} rolesService
     */
    constructor(buttonsHelper, dialog, rolesService) {
        this.buttonsHelper = buttonsHelper;
        this.dialog = dialog;
        this.rolesService = rolesService;
        /**
         * Enable legacy workitems
         */
        this.legacyWorkitems = this.legacyWorkitems ? this.legacyWorkitems : false;
        /**
         * Enable legacy creators
         */
        this.legacyCreators = this.legacyCreators ? this.legacyCreators : false;
        /**
         * Allow override of forms framework
         * Options: bootstrap-4 or material-design
         */
        this.formsFramework = this.formsFramework ? this.formsFramework : 'material-design';
        /**
         * ~event routeAction : Component requests route to another page
         * ~payload RouteAction : RouteAction object to tell caller to navigate somewhere
         */
        this.routeAction = new EventEmitter();
        this.toolbarButtons = [];
        this.incConfigButton = true;
        this.incRefreshButton = true;
        this.cockpitReady = false;
        this.clickCaseAction = (/**
         * @param {?} caseRoute
         * @return {?}
         */
        (caseRoute) => {
            // case clicked - tell parent (will pass caseRef and appId)
            this.routeAction.emit(new RouteAction('caseClicked', caseRoute));
        });
        this.createToolbarButtons = (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const buttons = [];
            // you can use the rolesService to either disable or hide the button as required - it returns true if the user has the roleId specified
            // const configButton = this.buttonsHelper.createButton('config', 'tcs-capabilities', true, 'Config', true, this.rolesService.checkRole('Partner Portal Configurator', this.roles));
            if (this.incConfigButton) {
                /** @type {?} */
                const configButton = this.buttonsHelper.createButton('config', 'tcs-capabilities', true, 'Config', (this.access ? this.rolesService.checkButton('configure', this.access) : true), true);
                buttons.push(configButton);
            }
            if (this.incRefreshButton) {
                /** @type {?} */
                const refreshButton = this.buttonsHelper.createButton('refresh', 'tcs-refresh-icon', true, 'Refresh', (this.access ? this.rolesService.checkButton('refresh', this.access) : true), true);
                buttons.push(refreshButton);
            }
            return buttons;
        });
        this.handleToolbarButtonEvent = (/**
         * @param {?} buttonId
         * @return {?}
         */
        (buttonId) => {
            if (buttonId === 'config') {
                this.routeAction.emit(new RouteAction('configClicked', null));
            }
            if (buttonId === 'refresh') {
                this.refresh();
            }
        });
        this.handleCreatorAppSelection = (/**
         * @param {?} application
         * @return {?}
         */
        (application) => {
            /** @type {?} */
            const EXAMPLE_INITIAL_DATA = {
                PartnerRequest: {
                    Customer_v1: {
                        CustomerReference_v1: 'CST-1111',
                        Name_v1: 'Roger Willis',
                    },
                    RequestDescription_v1: 'Where is my order?',
                    RequestDetails_v1: {
                        OrderReference_v1: 'ORD-55333',
                        PartReference_v1: 'PRT-102020',
                    },
                    RequestType_v1: 'Packaging Supplies'
                }
            }
            // this.openCreatorDialog(application, EXAMPLE_INITIAL_DATA, this.sandboxId);
            ;
            // this.openCreatorDialog(application, EXAMPLE_INITIAL_DATA, this.sandboxId);
            this.openCreatorDialog(application, undefined, this.sandboxId, this.customFormDefs, this.legacyCreators, this.formsFramework);
        });
        this.handleReportCaseTypeSelection = (/**
         * @param {?} caseTypeReportRecord
         * @return {?}
         */
        (caseTypeReportRecord) => {
            this.selectedCaseTypeReportRecord = caseTypeReportRecord;
        });
        this.handleReportCaseTypeStateSelection = (/**
         * @param {?} state
         * @return {?}
         */
        (state) => {
            if (this.selectedCaseTypeReportRecord) {
                this.searchComponent.searchCasesByState(Number(state.id), state.label, this.selectedCaseTypeReportRecord.applicationId, this.selectedCaseTypeReportRecord.caseTypeInfo.id, this.selectedCaseTypeReportRecord.caseTypeInfo.label + '|' + state.label);
            }
        });
        this.openCreatorDialog = (/**
         * @param {?} application
         * @param {?} initialData
         * @param {?} sandboxId
         * @param {?} customFormDefs
         * @param {?} legacyCreators
         * @param {?} formsFramework
         * @return {?}
         */
        (application, initialData, sandboxId, customFormDefs, legacyCreators, formsFramework) => {
            /** @type {?} */
            const dialogRef = this.dialog.open(LiveAppsCreatorDialogComponent, {
                width: '60%',
                height: '80%',
                maxWidth: '100vw',
                maxHeight: '100vh',
                panelClass: 'tcs-style-dialog',
                data: new CaseCreatorSelectionContext(application, initialData, sandboxId, customFormDefs, legacyCreators, formsFramework)
            });
            dialogRef.afterClosed().subscribe((/**
             * @param {?} result
             * @return {?}
             */
            result => {
                if (result) {
                    this.routeAction.emit(new RouteAction('caseClicked', result));
                }
            }));
        });
        this.refresh = (/**
         * @return {?}
         */
        () => {
            if (this.favoritesComponent) {
                this.favoritesComponent.refresh();
            }
            if (this.recentsComponent) {
                this.recentsComponent.refresh();
            }
            if (this.searchComponent) {
                this.searchComponent.refresh();
            }
            if (this.collaborationComponent) {
                this.collaborationComponent.refresh();
            }
            if (this.documentsComponent) {
                this.documentsComponent.refresh();
            }
            if (this.activeCasesComponent) {
                this.activeCasesComponent.refresh();
            }
            if (this.workitemsComponent) {
                this.workitemsComponent.refresh();
            }
        });
    }
    /**
     * @return {?}
     */
    initialize() {
        this.toolbarButtons = this.toolbarButtons.concat(this.createToolbarButtons());
        this.caseStartButtonActive = this.access ? this.rolesService.checkButton('caseStart', this.access) : true;
        this.cockpitReady = true;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        // on first set call initialize
        if (changes.appIds.firstChange && changes.sandboxId.firstChange && changes.uiAppId.firstChange) {
            this.initialize();
        }
    }
}
LiveAppsHomeCockpitComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-home-cockpit',
                template: "<div fxLayout=\"column\" fxFlex *ngIf=\"cockpitReady\">\n  <div>\n    <tc-tibco-cloud-menu-bar [toolbarButtons]=\"toolbarButtons\" [headerText]=\"title\"\n                             (toolbarButtonEvent)=\"handleToolbarButtonEvent($event)\">\n      <div>\n        <ng-content select=\"[customToolbarContent]\" fxFlex></ng-content>\n        <tcla-live-apps-application-create-button *ngIf=\"caseStartButtonActive\"\n                                                  style=\"margin-right: 20px; margin-left: 20px\" [sandboxId]=\"sandboxId\"\n                                                  [appIds]=\"appIds\"\n                                                  (selection)=\"handleCreatorAppSelection($event)\"></tcla-live-apps-application-create-button>\n      </div>\n    </tc-tibco-cloud-menu-bar>\n  </div>\n\n  <div class=\"tcs-home-content\"\n       style=\"padding-left: 20px; padding-right: 20px; margin-top: 10px; padding-bottom: 120px; height: 100%; display: block;\">\n    <div style=\"height: 100%;\">\n      <div class=\"searchCasesWidget\" style=\"min-height: 86px;\">\n        <tcla-live-apps-search-widget fxFlex [appIds]=\"appIds\" [sandboxId]=\"sandboxId\" [uiAppId]=\"uiAppId\"\n                                      (caseSelected)=\"clickCaseAction($event)\"></tcla-live-apps-search-widget>\n      </div>\n      <!--div style=\"background-color: red; height: 600px; width: 100%;\">TEST BOX</div-->\n\n      <!--div fxFlex=\"noshrink\" style=\"margin-bottom: 20px; overflow: auto; min-height: 400px\" fxLayoutGap=\"20px\" fxLayout=\"row\" fxLayout.lt-md=\"column\"-->\n      <div style=\"margin-bottom: 20px; overflow: hidden; height: 100%; min-height: 750px; padding-bottom: 10px;\" fxLayoutGap=\"20px\"\n           fxLayout=\"row\" fxLayout.lt-md=\"column\">\n        <div fxFlex.gt-sm=\"grow\" fxLayout=\"row\" style=\"\" [ngStyle.lt-md]=\"'min-height: 400px;'\">\n          <mat-tab-group fxFlex>\n            <mat-tab label=\"Recent Cases\">\n              <ng-template matTabContent>\n                <div class=\"recentCasesWidget\" fxFill>\n                  <tcla-live-apps-recent-cases class=\"padded-widget\" [showHeader]=\"true\" [sandboxId]=\"sandboxId\"\n                                               [uiAppId]=\"uiAppId\"\n                                               (clickCase)=\"clickCaseAction($event)\"></tcla-live-apps-recent-cases>\n                </div>\n              </ng-template>\n            </mat-tab>\n            <mat-tab label=\"Favorite Cases\">\n              <ng-template matTabContent>\n                <div class=\"favoriteCasesWidget\" fxFill>\n                  <tcla-live-apps-favorite-cases class=\"padded-widget\" [showHeader]=\"true\" [sandboxId]=\"sandboxId\"\n                                                 [uiAppId]=\"uiAppId\"\n                                                 (clickCase)=\"clickCaseAction($event)\"></tcla-live-apps-favorite-cases>\n                </div>\n              </ng-template>\n            </mat-tab>\n            <mat-tab *ngIf=\"legacyWorkitems\" label=\"Workitems\">\n              <ng-template matTabContent>\n                <div class=\"favoriteCasesWidget\" fxFill>\n                  <tcla-live-apps-workitems #componentDiv class=\"padded-widget\" [showHeader]=\"true\"\n                                            [sandboxId]=\"sandboxId\"\n                                            [uiAppId]=\"uiAppId\" [appIds]=\"appIds\"\n                                            (clickCase)=\"clickCaseAction($event)\"></tcla-live-apps-workitems>\n                </div>\n              </ng-template>\n            </mat-tab>\n          </mat-tab-group>\n        </div>\n\n        <div fxFlex.gt-sm=\"40\" fxLayout=\"column\" [ngStyle.lt-md]=\"'min-height: 400px'\">\n          <div style=\"padding-top: 48px; height: 350px;\">\n            <tcla-live-apps-active-cases-widget [sandboxId]=\"sandboxId\" [appIds]=\"appIds\" [showHeader]=\"true\"\n                                                [showPercentages]=\"true\" [uiAppId]=\"uiAppId\"\n                                                (selectedCaseTypeState)=\"handleReportCaseTypeStateSelection($event)\"\n                                                (selectedCaseType)=\"handleReportCaseTypeSelection($event)\"></tcla-live-apps-active-cases-widget>\n          </div>\n          <div fxFlex=\"noshrink\" style=\"height: 350px;\">\n            <mat-tab-group fxFlex>\n              <mat-tab label=\"Collaboration\">\n                <div fxFill class=\"collaborationWidget\">\n                  <tcla-live-apps-notes [userId]=\"userId\" [showHeader]=\"true\" [relatedItemId]=\"uiAppId\"\n                                        [relatedItemType]=\"'CASE_APP'\"></tcla-live-apps-notes>\n                </div>\n              </mat-tab>\n              <mat-tab label=\"Documents\">\n                <div fxFill class=\"documentWidget\">\n                  <tcla-live-apps-documents [showHeader]=\"true\" [sandboxId]=\"sandboxId\" [folderId]=\"uiAppId + '_Docs'\"\n                                            [folderType]=\"'orgFolders'\"\n                                            [folderDescription]=\"'Document Library'\"></tcla-live-apps-documents>\n                </div>\n              </mat-tab>\n            </mat-tab-group>\n          </div>\n        </div>\n      </div>\n\n    </div>\n  </div>\n</div>\n",
                styles: [".tc-home-welcome-text{font-family:Source Sans Pro;font-size:30px;font-weight:400;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:normal;color:#212121}.tcs-home-content{overflow:auto}"]
            }] }
];
/** @nocollapse */
LiveAppsHomeCockpitComponent.ctorParameters = () => [
    { type: TcButtonsHelperService },
    { type: MatDialog },
    { type: TcRolesService }
];
LiveAppsHomeCockpitComponent.propDecorators = {
    uiAppId: [{ type: Input }],
    appIds: [{ type: Input }],
    sandboxId: [{ type: Input }],
    userName: [{ type: Input }],
    userId: [{ type: Input }],
    email: [{ type: Input }],
    title: [{ type: Input }],
    roles: [{ type: Input }],
    access: [{ type: Input }],
    customFormDefs: [{ type: Input }],
    legacyWorkitems: [{ type: Input }],
    legacyCreators: [{ type: Input }],
    formsFramework: [{ type: Input }],
    routeAction: [{ type: Output }],
    favoritesComponent: [{ type: ViewChild, args: [LiveAppsFavoriteCasesComponent, { static: false },] }],
    recentsComponent: [{ type: ViewChild, args: [LiveAppsRecentCasesComponent, { static: false },] }],
    searchComponent: [{ type: ViewChild, args: [LiveAppsSearchWidgetComponent, { static: false },] }],
    collaborationComponent: [{ type: ViewChild, args: [LiveAppsNotesComponent, { static: false },] }],
    documentsComponent: [{ type: ViewChild, args: [LiveAppsDocumentsComponent, { static: false },] }],
    activeCasesComponent: [{ type: ViewChild, args: [LiveAppsActiveCasesWidgetComponent, { static: false },] }],
    workitemsComponent: [{ type: ViewChild, args: [LiveAppsWorkitemsComponent, { static: false },] }]
};
if (false) {
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.uiAppId;
    /**
     * The list of LA Application IDs you want to handle
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.appIds;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.sandboxId;
    /**
     * The name of the logged user
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.userName;
    /**
     * The ID of the logged user
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.userId;
    /**
     * * Email address of the user (comes from resolver)
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.email;
    /**
     * page title comes from config resolver
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.title;
    /**
     * Roles - The users current roles
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.roles;
    /**
     * RouteAccessControlConfig - basically the config for access control
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.access;
    /**
     * Custom Form configuration file
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.customFormDefs;
    /**
     * Enable legacy workitems
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.legacyWorkitems;
    /**
     * Enable legacy creators
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.legacyCreators;
    /**
     * Allow override of forms framework
     * Options: bootstrap-4 or material-design
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.formsFramework;
    /**
     * ~event routeAction : Component requests route to another page
     * ~payload RouteAction : RouteAction object to tell caller to navigate somewhere
     * @type {?}
     */
    LiveAppsHomeCockpitComponent.prototype.routeAction;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.favoritesComponent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.recentsComponent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.searchComponent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.collaborationComponent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.documentsComponent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.activeCasesComponent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.workitemsComponent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.toolbarButtons;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.caseStartButtonActive;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.selectedCaseTypeReportRecord;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.incConfigButton;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.incRefreshButton;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.cockpitReady;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.clickCaseAction;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsHomeCockpitComponent.prototype.createToolbarButtons;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.handleToolbarButtonEvent;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.handleCreatorAppSelection;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.handleReportCaseTypeSelection;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.handleReportCaseTypeStateSelection;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.openCreatorDialog;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.refresh;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsHomeCockpitComponent.prototype.buttonsHelper;
    /** @type {?} */
    LiveAppsHomeCockpitComponent.prototype.dialog;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsHomeCockpitComponent.prototype.rolesService;
}
