// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-favorite-cases/live-apps-favorite-cases.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { take, takeUntil } from 'rxjs/operators';
import { LiveAppsService } from '../../services/live-apps.service';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { TcCoreCommonFunctions } from '@tibco-tcstk/tc-core-lib';
/**
 * This Component allows to list your Favorite Cases.
 *
 * ![alt-text](../live-apps-favorite-cases.png "Documents Component Image")
 *
 * \@example <tcla-live-apps-favorite-cases></tcla-live-apps-favorite-cases>
 */
export class LiveAppsFavoriteCasesComponent extends LiveAppsComponent {
    /**
     * @param {?} liveapps
     */
    constructor(liveapps) {
        super();
        this.liveapps = liveapps;
        /**
         * case card format - list, card, miniCard, staticList (no click event)
         */
        this.displayType = this.displayType ? this.displayType : 'miniCard'; // miniCard, card, list
        // miniCard, card, list
        /**
         * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
         */
        this.showHeader = this.showHeader ? this.showHeader : true;
        /**
         * ~event clickCase : Case clicked
         * ~payload CaseRoute : CaseRoute object output when case is clicked so calling component can route accordingly - ie. route to case
         */
        this.clickCase = new EventEmitter();
        this.clickCaseAction = (/**
         * @param {?} caseRoute
         * @return {?}
         */
        (caseRoute) => {
            this.clickCase.emit(caseRoute);
        });
        this.refresh = (/**
         * @return {?}
         */
        () => {
            this.favoriteCases = [];
            this.liveapps.getFavoriteCases(this.uiAppId, this.sandboxId)
                .pipe(take(1), takeUntil(this._destroyed$)).subscribe((/**
             * @param {?} favoriteCases
             * @return {?}
             */
            favoriteCases => {
                this.favoriteCases = favoriteCases.caseRefs || [];
            }), (/**
             * @param {?} error
             * @return {?}
             */
            error => { this.errorMessage = 'Error retrieving favorite cases: ' + error.error.errorMsg; }));
        });
        this.clearFavoriteCases = (/**
         * @return {?}
         */
        () => {
            this.liveapps.setFavoriteCase('-1', this.uiAppId, this.sandboxId);
            this.favoriteCases = [];
        });
        this.handleDeleted = (/**
         * @param {?} caseRef
         * @return {?}
         */
        (caseRef) => {
            this.favoriteCases.splice(this.favoriteCases.indexOf(caseRef), 1);
            this.liveapps.setFavoriteCase(caseRef, this.uiAppId, this.sandboxId);
        });
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this.containerChanges$.subscribe((/**
         * @param {?} widget
         * @return {?}
         */
        widget => {
            this.cardWidthPct = TcCoreCommonFunctions.calcSummaryCardPct(widget);
        }));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.refresh();
    }
}
LiveAppsFavoriteCasesComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-favorite-cases',
                template: "<div #componentDiv class=\"tcs-case-favorites-box\" fxLayout=\"column\" fxFill>\n  <div *ngIf=\"showHeader\" class=\"tcs-case-favorites-header\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n    <div fxLayoutAlign=\"start center\">\n      <mat-icon class=\"tcs-icon tcs-favorites-icon\" svgIcon=\"tcs-favorites-icon\"></mat-icon>\n      <div class=\"tcs-case-favorites-header-text\">Favorites</div>\n    </div>\n    <mat-icon (click)=\"clearFavoriteCases()\" class=\"tcs-icon tcs-icon-active tcs-clear-favorite-icon\" svgIcon=\"tcs-clear-icon\" matTooltip=\"Clear Favorites\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\"></mat-icon>\n  </div>\n  <div *ngIf=\"!showHeader\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n    <mat-icon style=\"margin: 5px\" (click)=\"clearFavoriteCases()\" class=\"tcs-icon tcs-icon-active tcs-clear-favorite-icon\" svgIcon=\"tcs-clear-icon\" matTooltip=\"Clear Favorites\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\"></mat-icon>\n  </div>\n  <div *ngIf=\"favoriteCases.length > 0\" class=\"tcs-case-favorites-flow-list\" fxLayout=\"column\" fxLayout=\"row wrap\">\n    <div class=\"tcs-case-favorite-item-box\" *ngFor=\"let case of favoriteCases\" [fxFlex]=\"cardWidthPct\">\n      <tcla-live-apps-case-summary [uiAppId]=\"uiAppId\" [typeBar]=\"true\" [borderCard]=\"false\" [displayType]=\"displayType\" [sandboxId]=\"sandboxId\" [caseRef]=\"case\" (clickCase)=\"clickCaseAction($event)\" (deleted)=\"handleDeleted($event)\"></tcla-live-apps-case-summary>\n      <div class=\"tcs-case-summary-line\"></div>\n    </div>\n  </div>\n  <div *ngIf=\"!(favoriteCases.length > 0)\" fxLayout=\"row\" fxLayoutAlign=\"center start\" fxLayoutGap=\"10px\" style=\"margin: 20px;\">\n    <mat-icon [svgIcon]=\"'ic-no-cases-icon'\" style=\"height: 48px; width: 48px;\"></mat-icon>\n    <div style=\"height: 100%; margin: 5px\" fxLayoutAlign=\"start center\">\n      <span class=\"tcs-no-item-text\">No cases found</span>\n    </div>\n  </div>\n</div>\n",
                styles: [".tcs-case-favorites-box{border-radius:3px;box-shadow:0 2px 8px 0 #dedede;background-color:#fff}.tcs-case-favorites-header{min-height:40px;border-radius:3px 3px 0 0;box-shadow:0 1px 2px 0 #dedede;padding-left:20px;padding-right:20px}.tcs-case-favorites-header-text{font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;text-align:left;color:#000;margin-left:10px}.tcs-case-favorites-flow-list{margin:10px;overflow-y:auto;padding:4px}.tcs-case-favorite-item-box{padding:5px}.tcs-icon.tcs-icon-active:hover{cursor:pointer}:host ::ng-deep .tcs-icon.tcs-icon-active:hover .svg-content{fill:#0081cb}.tcs-case-summary-line{padding:0;margin:4px 5px 0;border-bottom:1.1px solid #f4f4f4}.tcs-no-item-text{font-family:Source Sans Pro;font-size:16px;font-weight:400;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:normal;color:#b6b6b6}"]
            }] }
];
/** @nocollapse */
LiveAppsFavoriteCasesComponent.ctorParameters = () => [
    { type: LiveAppsService }
];
LiveAppsFavoriteCasesComponent.propDecorators = {
    sandboxId: [{ type: Input }],
    uiAppId: [{ type: Input }],
    displayType: [{ type: Input }],
    showHeader: [{ type: Input }],
    clickCase: [{ type: Output }]
};
if (false) {
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsFavoriteCasesComponent.prototype.sandboxId;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsFavoriteCasesComponent.prototype.uiAppId;
    /**
     * case card format - list, card, miniCard, staticList (no click event)
     * @type {?}
     */
    LiveAppsFavoriteCasesComponent.prototype.displayType;
    /**
     * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
     * @type {?}
     */
    LiveAppsFavoriteCasesComponent.prototype.showHeader;
    /**
     * ~event clickCase : Case clicked
     * ~payload CaseRoute : CaseRoute object output when case is clicked so calling component can route accordingly - ie. route to case
     * @type {?}
     */
    LiveAppsFavoriteCasesComponent.prototype.clickCase;
    /** @type {?} */
    LiveAppsFavoriteCasesComponent.prototype.favoriteCases;
    /** @type {?} */
    LiveAppsFavoriteCasesComponent.prototype.errorMessage;
    /** @type {?} */
    LiveAppsFavoriteCasesComponent.prototype.cardWidthPct;
    /** @type {?} */
    LiveAppsFavoriteCasesComponent.prototype.clickCaseAction;
    /** @type {?} */
    LiveAppsFavoriteCasesComponent.prototype.refresh;
    /** @type {?} */
    LiveAppsFavoriteCasesComponent.prototype.clearFavoriteCases;
    /** @type {?} */
    LiveAppsFavoriteCasesComponent.prototype.handleDeleted;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsFavoriteCasesComponent.prototype.liveapps;
}
