// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-legacy-form/live-apps-legacy-form.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { LegacyIframeService } from '@tibco-tcstk/tc-core-lib';
import { TcVisibilityService } from '@tibco-tcstk/tc-core-lib';
/**
 * Wraps legacy angularjs form renderer
 *
 * \@example <tcla-live-apps-legacy-form></tcla-live-apps-legacy-form>
 */
var LiveAppsLegacyFormComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsLegacyFormComponent, _super);
    function LiveAppsLegacyFormComponent(visibilityService, host, legacyIframeService) {
        var _this = _super.call(this) || this;
        _this.visibilityService = visibilityService;
        _this.host = host;
        _this.legacyIframeService = legacyIframeService;
        _this.legacyIframeId = _this.legacyIframeId ? _this.legacyIframeId : 'legacyWorkitemFrame';
        _this.workitemComplete = new EventEmitter();
        _this.wiActive = false;
        _this.renderWi = (/**
         * @param {?} wiId
         * @return {?}
         */
        function (wiId) {
            if (_this.wiActive && _this.openWiId) {
                // opening a new workitem so cancel old one
                _this.cancelWi(wiId);
            }
            // send message to external form app
            _this.formDiv.contentWindow.postMessage({ 'action': 'openWI', 'wiId': wiId }, window.location.origin);
            // position the form iframe over the workitemDiv placeholder
            _this.wiActive = true;
            _this.openWiId = wiId;
            _this.resizeWi();
        });
        _this.receiveMessage = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (event.data.action === 'wiCompleted') {
                _this.wiActive = false;
                /** @type {?} */
                var wiId = _this.openWiId;
                _this.openWiId = undefined;
                _this.hideWi();
                _this.workitemComplete.emit(wiId);
            }
        });
        _this.cancelWi = (/**
         * @param {?} wiId
         * @return {?}
         */
        function (wiId) {
            _this.formDiv.contentWindow.postMessage({ 'action': 'cancelWI', 'wiId': _this.openWiId }, window.location.origin);
            _this.wiActive = false;
            _this.openWiId = undefined;
            _this.hideWi();
        });
        _this.hideWi = (/**
         * @return {?}
         */
        function () {
            // hide the iframe!
            _this.formDiv.style.top = '-1000px';
            _this.formDiv.style.left = '-1000px';
            _this.formDiv.style.zIndex = '1';
        });
        _this.resizeWi = (/**
         * @return {?}
         */
        function () {
            // position the form iframe over the workitemDiv placeholder
            setTimeout((/**
             * @param {?} handler
             * @return {?}
             */
            function (handler) {
                /** @type {?} */
                var rect = _this.target.getBoundingClientRect();
                _this.formDiv.style.top = rect.top.toString() + 'px';
                _this.formDiv.style.left = rect.left.toString() + 'px';
                _this.formDiv.style.height = rect.height.toString() + 'px';
                _this.formDiv.style.width = rect.width.toString() + 'px';
                _this.formDiv.style.position = 'absolute';
                _this.formDiv.style.zIndex = '1000';
            }), 1000);
        });
        _this.initialize = (/**
         * @return {?}
         */
        function () {
            // listen for messages from the iframe
            window.addEventListener('message', _this.receiveMessage, false);
            // we only want to render the workitem when the element is visible
            /** @type {?} */
            var inSight$ = _this.visibilityService.elementInSight(_this.host);
            inSight$.subscribe((/**
             * @param {?} next
             * @return {?}
             */
            function (next) {
                if (!next) {
                    // element not visible so hide the workitem if it is showing
                    if (_this.wiActive && _this.target) {
                        _this.hideWi();
                    }
                }
                else {
                    // element is visible
                    if (!_this.wiActive) {
                        // if workitem not already showing - show it
                        // first get a handle on the target div and the legacy iframe
                        _this.target = document.getElementById('workitemDiv');
                        _this.formDiv = document.getElementById(_this.legacyIframeId);
                        if (_this.workitemId) {
                            // trigger display of the workitem
                            _this.renderWi(_this.workitemId);
                        }
                    }
                    else {
                        _this.resizeWi();
                    }
                }
            }));
        });
        return _this;
    }
    /**
     * @return {?}
     */
    LiveAppsLegacyFormComponent.prototype.isFormActive = /**
     * @return {?}
     */
    function () {
        return this.wiActive;
    };
    /**
     * @return {?}
     */
    LiveAppsLegacyFormComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngAfterViewInit.call(this);
        this.containerChanges$.subscribe((/**
         * @param {?} widget
         * @return {?}
         */
        function (widget) {
            if (_this.wiActive && _this.target) {
                _this.resizeWi();
            }
        }));
        // wait for iframe to load before doing anything else
        // this observable will emit true once the iFrame is loaded
        this.legacyIframeService.workitemStatus.subscribe((/**
         * @param {?} loaded
         * @return {?}
         */
        function (loaded) {
            if (loaded) {
                // iframe is loaded
                _this.initialize();
            }
        }));
    };
    /**
     * @return {?}
     */
    LiveAppsLegacyFormComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.openWiId) {
            this.cancelWi(this.workitemId);
        }
    };
    LiveAppsLegacyFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-legacy-form',
                    template: "<div #componentDiv fxFill id=\"workitemDiv\" style=\"border: none;\"></div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsLegacyFormComponent.ctorParameters = function () { return [
        { type: TcVisibilityService },
        { type: ElementRef },
        { type: LegacyIframeService }
    ]; };
    LiveAppsLegacyFormComponent.propDecorators = {
        legacyIframeId: [{ type: Input }],
        workitemId: [{ type: Input }],
        workitemComplete: [{ type: Output }]
    };
    return LiveAppsLegacyFormComponent;
}(LiveAppsComponent));
export { LiveAppsLegacyFormComponent };
if (false) {
    /** @type {?} */
    LiveAppsLegacyFormComponent.prototype.legacyIframeId;
    /** @type {?} */
    LiveAppsLegacyFormComponent.prototype.workitemId;
    /** @type {?} */
    LiveAppsLegacyFormComponent.prototype.workitemComplete;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.target;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.formDiv;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.wiActive;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.openWiId;
    /** @type {?} */
    LiveAppsLegacyFormComponent.prototype.renderWi;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.receiveMessage;
    /** @type {?} */
    LiveAppsLegacyFormComponent.prototype.cancelWi;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.hideWi;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.resizeWi;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyFormComponent.prototype.initialize;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsLegacyFormComponent.prototype.visibilityService;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsLegacyFormComponent.prototype.host;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsLegacyFormComponent.prototype.legacyIframeService;
}
