// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-settings-landing/live-apps-settings-landing.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { MatSnackBar, MatDialog } from '@angular/material';
import { TcGeneralLandingPageConfigService, TibcoCloudNewElementComponent, TibcoCloudUploadDialogComponent } from '@tibco-tcstk/tc-core-lib';
import { TcDocumentService } from '../../services/tc-document.service';
import { LandingPageConfig, LandingPageItemConfig } from '../../models/tc-live-apps-landing-page-config';
/**
 * Configuration page home
 *
 * \@example <tcla-live-apps-setting-landing></tcla-live-apps-setting-landing>
 */
var LiveAppsSettingsLandingComponent = /** @class */ (function () {
    function LiveAppsSettingsLandingComponent(route, generalLandingPageConfigService, snackBar, dialog, documentsService) {
        var _this = this;
        this.route = route;
        this.generalLandingPageConfigService = generalLandingPageConfigService;
        this.snackBar = snackBar;
        this.dialog = dialog;
        this.documentsService = documentsService;
        /**
         * New Configuration
         */
        this.runNewConfiguration = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var dialogRef = _this.dialog.open(TibcoCloudNewElementComponent, {
                width: '50%',
                height: '30%',
                maxWidth: '100vw',
                maxHeight: '100vh',
                panelClass: 'tcs-style-dialog',
                data: { resourceType: 'Landing Page' }
            });
            dialogRef.afterClosed().subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                if (result) {
                    /** @type {?} */
                    var newElement = new LandingPageConfig().deserialize({
                        key: result.id,
                        description: result.name,
                        highlights: [new LandingPageItemConfig(), new LandingPageItemConfig(), new LandingPageItemConfig()]
                    });
                    _this.landingPages.push(newElement);
                    _this.selectedWelcomePage = newElement;
                }
            }));
        });
        /**
         * Helper to Compare Objects
         */
        this.compareObjects = (/**
         * @param {?} o1
         * @param {?} o2
         * @return {?}
         */
        function (o1, o2) {
            return o1 === o2;
        });
    }
    /**
    * @ignore
    */
    /**
     * @ignore
     * @return {?}
     */
    LiveAppsSettingsLandingComponent.prototype.ngOnInit = /**
     * @ignore
     * @return {?}
     */
    function () {
        this.landingPagesConfig = this.route.snapshot.data.landingPagesConfigHolder;
        this.landingPages = this.landingPagesConfig.landingPage;
        this.allRoles = this.route.snapshot.data.allRolesHolder.roles.filter((/**
         * @param {?} element
         * @return {?}
         */
        function (element) { return !element.configuration; }));
        this.sandboxId = this.route.snapshot.data.claims.primaryProductionSandbox.id;
        this.uiAppId = this.route.snapshot.data.landingPagesConfigHolder.uiAppId;
        // If there is only one landing page selects it automatically
        if (this.landingPages.length == 1) {
            this.selectedWelcomePage = this.landingPages[0];
        }
    };
    /**
     * Save Configuration
     */
    /**
     * Save Configuration
     * @return {?}
     */
    LiveAppsSettingsLandingComponent.prototype.runSaveFunction = /**
     * Save Configuration
     * @return {?}
     */
    function () {
        var _this = this;
        this.generalLandingPageConfigService.updateGeneralLandingPageConfig(this.sandboxId, this.uiAppId, this.landingPagesConfig, this.landingPagesConfig.id).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            _this.snackBar.open('Landing Pages configuration saved', 'OK', {
                duration: 3000
            });
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            _this.snackBar.open('Error saving Landing Pages configuration', 'OK', {
                duration: 3000
            });
        }));
    };
    /**
     * Delete Configuration
     */
    /**
     * Delete Configuration
     * @return {?}
     */
    LiveAppsSettingsLandingComponent.prototype.runDeleteConfiguration = /**
     * Delete Configuration
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var pages = this.landingPages;
        pages.forEach((/**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            if (element == _this.selectedWelcomePage) {
                /** @type {?} */
                var index = pages.indexOf(element, 0);
                pages.splice(index, 1);
                _this.selectedWelcomePage = undefined;
            }
        }));
    };
    /**
     * @param {?} imageFormat
     * @param {?} location
     * @param {?=} index
     * @return {?}
     */
    LiveAppsSettingsLandingComponent.prototype.openDialog = /**
     * @param {?} imageFormat
     * @param {?} location
     * @param {?=} index
     * @return {?}
     */
    function (imageFormat, location, index) {
        var _this = this;
        /** @type {?} */
        var dialogRef = this.dialog.open(TibcoCloudUploadDialogComponent, {
            width: '500px',
            data: {
                title: 'Upload New ' + location + ' image',
                allowedExtensions: imageFormat,
                location: location,
                index: index
            }
        });
        dialogRef.componentInstance.fileevent.subscribe((/**
         * @param {?} $e
         * @return {?}
         */
        function ($e) {
            _this.uploadFile($e.file, location, index);
        }));
        dialogRef.afterClosed().subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
        }));
    };
    /**
     * @param {?} file
     * @param {?} location
     * @param {?=} index
     * @return {?}
     */
    LiveAppsSettingsLandingComponent.prototype.uploadFile = /**
     * @param {?} file
     * @param {?} location
     * @param {?=} index
     * @return {?}
     */
    function (file, location, index) {
        if (file) {
            switch (location) {
                case 'background':
                    this.selectedWelcomePage.backgroundURL = file.name;
                    break;
                case 'hightlight':
                    this.selectedWelcomePage.highlights[index].iconURL = file.name;
                    break;
                default:
                    break;
            }
            this.documentsService.uploadDocument('orgFolders', this.uiAppId, this.sandboxId, file, location + '%2F' + file.name, '')
                .subscribe((/**
             * @param {?} val
             * @return {?}
             */
            function (val) {
                console.log("*********** DONE");
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) { console.log('error', error.errorMsg); })); //); this.errorMessage = 'Error uploading state icon: ' +
        }
    };
    LiveAppsSettingsLandingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-settings-landing',
                    template: "<div fxLayout=\"column\" fxFill>\n    <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\"\n        [headerText]=\"'Landing Page Configuration'\"></tc-tibco-cloud-widget-header>\n    <div fxFlex style=\"padding: 20px; overflow: hidden\" fxLayout=\"column\">\n        <div style=\"overflow: auto;\">\n            <div fxLayout=\"column\" fxFlex>\n                <p>You can manage the configuration for the landing pages here.</p>\n                <br>\n                <mat-form-field>\n                    <mat-label>Welcome Page Configuration</mat-label>\n                    <mat-select [(value)]=\"selectedWelcomePage\">\n                        <mat-option *ngFor=\"let landingPage of landingPages\" [value]=\"landingPage\">\n                            {{landingPage.key}}-{{landingPage.description}}\n                        </mat-option>\n                    </mat-select>\n                </mat-form-field>\n\n                <div *ngIf=\"selectedWelcomePage\">\n                    <div fxLayoutGap=\"10px\">\n                        <mat-form-field fxFlex=\"50\">\n                            <input matInput placeholder=\"Title\" [(ngModel)]=\"selectedWelcomePage.title\">\n                        </mat-form-field>\n                        <mat-form-field fxFlex=\"50\">\n                            <input matInput placeholder=\"Subtitle\" [(ngModel)]=\"selectedWelcomePage.subtitle\">\n                        </mat-form-field>\n                    </div>\n                    <div fxLayoutGap=\"10px\">\n                        <div fxFlex=\"50\">\n                            <mat-form-field fxFlex>\n                                <input matInput placeholder=\"Background URL\" [(ngModel)]=\"selectedWelcomePage.backgroundURL\">\n                            </mat-form-field>\n                            <mat-icon (click)=\"openDialog('.jpeg,.png', 'background')\"\n                                class=\"tcs-icon tcs-icon-active tcs-document-edit-buttons tcs-document-upload-icon\"\n                                matTooltip=\"Upload Document\"\n                                matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\" svgIcon=\"tcs-document-upload\"></mat-icon>\n                        </div>\n                        <mat-form-field fxFlex=\"50\">\n                            <input matInput placeholder=\"Margin above highlight panel (px)\" type=\"number\" min=\"0\" [(ngModel)]=\"selectedWelcomePage.topMargin\">\n                        </mat-form-field>\n                    </div>\n                    <div fxLayoutGap=\"10px\">\n                        <mat-form-field fxFlex=\"50\">\n                            <mat-label>Used in roles</mat-label>\n                            <mat-select [(ngModel)]=\"selectedWelcomePage.roles\" [compareWith]=\"compareObjects\" multiple>\n                                <mat-option *ngFor=\"let role of allRoles\" [value]=\"role.id\">{{role.display}}</mat-option>\n                            </mat-select>\n                        </mat-form-field>\n                        <mat-form-field fxFlex=\"50\">\n                            <input matInput placeholder=\"Home route\" [(ngModel)]=\"selectedWelcomePage.homeRoute\">\n                        </mat-form-field>\n                    </div>\n                    <div>Hightlights</div>\n                    <div *ngFor=\"let highlight of selectedWelcomePage.highlights; let i = index\" fxLayoutGap=\"10px\">\n                        <div fxFlex=\"15\">\n                            <mat-form-field fxFlex=\"80\">\n                                <input matInput placeholder=\"Icon URL\" [(ngModel)]=\"highlight.iconURL\">\n                            </mat-form-field>\n                            <mat-icon fxFlex=\"20\" (click)=\"openDialog('.svg', 'hightlight', i)\"\n                                class=\"tcs-icon tcs-icon-active tcs-document-edit-buttons tcs-document-upload-icon\" matTooltip=\"Upload Document\"\n                                matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\" svgIcon=\"tcs-document-upload\"></mat-icon>\n                        </div>\n                        <mat-form-field fxFlex=\"30\">\n                            <input matInput placeholder=\"Title\" [(ngModel)]=\"highlight.title\">\n                        </mat-form-field>\n                        <mat-form-field fxFlex=\"55\">\n                            <textarea matInput placeholder=\"Subtitle\" [(ngModel)]=\"highlight.content\" cdkAutosizeMinRows=\"3\" cdkAutosizeMaxRows=\"3\"></textarea>\n                        </mat-form-field>\n                    </div>\n                </div>\n            </div>\n        </div>\n        <div fxFlex class=\"tcs-filler-panel\"></div>\n        <div fxLayout=\"row\" fxLayoutGap=\"10px\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n            <button mat-raised-button color=\"secundary\" (click)=\"runDeleteConfiguration()\">Delete</button>\n            <button mat-raised-button color=\"secundary\" (click)=\"runNewConfiguration()\">New</button>\n            <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n        </div>\n    </div>\n</div>",
                    styles: [".tcs-document-upload-icon{margin-top:10px}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsSettingsLandingComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: TcGeneralLandingPageConfigService },
        { type: MatSnackBar },
        { type: MatDialog },
        { type: TcDocumentService }
    ]; };
    return LiveAppsSettingsLandingComponent;
}());
export { LiveAppsSettingsLandingComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.landingPagesConfig;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.sandboxId;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.uiAppId;
    /** @type {?} */
    LiveAppsSettingsLandingComponent.prototype.landingPages;
    /** @type {?} */
    LiveAppsSettingsLandingComponent.prototype.selectedWelcomePage;
    /** @type {?} */
    LiveAppsSettingsLandingComponent.prototype.allRoles;
    /** @type {?} */
    LiveAppsSettingsLandingComponent.prototype.selectedRole;
    /**
     * New Configuration
     * @type {?}
     */
    LiveAppsSettingsLandingComponent.prototype.runNewConfiguration;
    /**
     * Helper to Compare Objects
     * @type {?}
     */
    LiveAppsSettingsLandingComponent.prototype.compareObjects;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.generalLandingPageConfigService;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.snackBar;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    LiveAppsSettingsLandingComponent.prototype.documentsService;
}
