// @tibco-tcstk/tc-liveapps-lib/esm5/lib/models/tc-workitems.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DeserializableClass } from '@tibco-tcstk/tc-core-lib';
var WiDistributionStrategy = /** @class */ (function (_super) {
    tslib_1.__extends(WiDistributionStrategy, _super);
    function WiDistributionStrategy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return WiDistributionStrategy;
}(DeserializableClass));
export { WiDistributionStrategy };
if (false) {
    /** @type {?} */
    WiDistributionStrategy.prototype.stringKey;
}
var WiScheduleStatus = /** @class */ (function (_super) {
    tslib_1.__extends(WiScheduleStatus, _super);
    function WiScheduleStatus() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return WiScheduleStatus;
}(DeserializableClass));
export { WiScheduleStatus };
if (false) {
    /** @type {?} */
    WiScheduleStatus.prototype.description;
    /** @type {?} */
    WiScheduleStatus.prototype.stringKey;
}
var WiItemContext = /** @class */ (function (_super) {
    tslib_1.__extends(WiItemContext, _super);
    function WiItemContext() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return WiItemContext;
}(DeserializableClass));
export { WiItemContext };
if (false) {
    /** @type {?} */
    WiItemContext.prototype.activityId;
    /** @type {?} */
    WiItemContext.prototype.activityName;
    /** @type {?} */
    WiItemContext.prototype.appId;
    /** @type {?} */
    WiItemContext.prototype.appInstance;
    /** @type {?} */
    WiItemContext.prototype.appName;
    /** @type {?} */
    WiItemContext.prototype.caseRef;
}
var WiFlags = /** @class */ (function (_super) {
    tslib_1.__extends(WiFlags, _super);
    function WiFlags() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return WiFlags;
}(DeserializableClass));
export { WiFlags };
if (false) {
    /** @type {?} */
    WiFlags.prototype.scheduleStatus;
}
var WiHeader = /** @class */ (function (_super) {
    tslib_1.__extends(WiHeader, _super);
    function WiHeader() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return WiHeader;
}(DeserializableClass));
export { WiHeader };
if (false) {
    /** @type {?} */
    WiHeader.prototype.description;
    /** @type {?} */
    WiHeader.prototype.distributionStrategy;
    /** @type {?} */
    WiHeader.prototype.flags;
    /** @type {?} */
    WiHeader.prototype.itemContext;
    /** @type {?} */
    WiHeader.prototype.name;
    /** @type {?} */
    WiHeader.prototype.priority;
    /** @type {?} */
    WiHeader.prototype.startDate;
}
var WiState = /** @class */ (function (_super) {
    tslib_1.__extends(WiState, _super);
    function WiState() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return WiState;
}(DeserializableClass));
export { WiState };
if (false) {
    /** @type {?} */
    WiState.prototype.description;
    /** @type {?} */
    WiState.prototype.stringKey;
}
var Workitem = /** @class */ (function (_super) {
    tslib_1.__extends(Workitem, _super);
    function Workitem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    Workitem.prototype.deserialize = /**
     * @template THIS
     * @this {THIS}
     * @param {?} input
     * @return {THIS}
     */
    function (input) {
        Object.assign((/** @type {?} */ (this)), input);
        return (/** @type {?} */ (this));
    };
    return Workitem;
}(DeserializableClass));
export { Workitem };
if (false) {
    /** @type {?} */
    Workitem.prototype.header;
    /** @type {?} */
    Workitem.prototype.id;
    /** @type {?} */
    Workitem.prototype.state;
    /** @type {?} */
    Workitem.prototype.version;
    /** @type {?} */
    Workitem.prototype.visible;
}
