// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-notes-editor/live-apps-notes-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
/**
 * Editor component to edit a note
 *
 * ![alt-text](../live-apps-notes-editor.png "")
 *
 * \@example <tcla-live-apps-notes-editor></tcla-live-apps-notes-editor>
 */
export class LiveAppsNotesEditorComponent extends LiveAppsComponent {
    // notify text submitted
    constructor() {
        super();
        // use 'CASE_APP' to share notes with case manager
        this.isReply = false;
        /**
         * ~event noteSubmitted : Internal
         */
        this.noteSubmitted = new EventEmitter(); // notify text submitted
        this.submitNote = (/**
         * @param {?} noteText
         * @return {?}
         */
        (noteText) => {
            this.noteSubmitted.emit(noteText);
            this.noteText = undefined;
        });
    }
}
LiveAppsNotesEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-notes-editor',
                template: "<div class=\"tcs-collaboration-new-thread\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\" fxFlex>\n  <mat-icon *ngIf=\"isReply\" class=\"tcs-icon tcs-collaboration-reply-item-icon\" svgIcon=\"tcs-collaboration-reply\"></mat-icon>\n  <form class=\"tcs-collaboration-new-thread-form\" #newThreadForm=\"ngForm\" fxFlex>\n    <textarea fxFlex class=\"tcs-collaboration-input\" [(ngModel)]=\"noteText\" name=\"noteText\" md-maxlength=\"10000\" placeholder=\"New Note... \" rows=\"2\"></textarea>\n  </form>\n  <mat-icon class=\"tcs-icon tcs-collaboration-send-icon\" svgIcon=\"tcs-collaboration-send\" (click)=\"submitNote(noteText)\"></mat-icon>\n</div>\n",
                styles: [".tcs-collaboration-new-thread-form{max-width:90%;max-height:100%}.tcs-collaboration-input{height:55px;max-width:100%;border-color:#d5e6ff;font-size:11px;font-weight:400;font-style:normal;font-stretch:normal;line-height:normal;letter-spacing:-.1px;color:#b6b6b6;resize:none;border-radius:3px}.tcs-icon.tcs-collaboration-send-icon{height:24px;width:24px;margin-left:12px}.tcs-icon{height:20px;width:20px}.tcs-collaboration-reply-item-icon{-webkit-transform:scale(1,-1);transform:scale(1,-1);margin-top:10px}"]
            }] }
];
/** @nocollapse */
LiveAppsNotesEditorComponent.ctorParameters = () => [];
LiveAppsNotesEditorComponent.propDecorators = {
    noteText: [{ type: Input }],
    isReply: [{ type: Input }],
    noteSubmitted: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LiveAppsNotesEditorComponent.prototype.noteText;
    /** @type {?} */
    LiveAppsNotesEditorComponent.prototype.isReply;
    /**
     * ~event noteSubmitted : Internal
     * @type {?}
     */
    LiveAppsNotesEditorComponent.prototype.noteSubmitted;
    /** @type {?} */
    LiveAppsNotesEditorComponent.prototype.submitNote;
}
