// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-settings/live-apps-settings.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { LiveAppsConfig } from '../../models/tc-liveapps-config';
import { ActivatedRoute } from '@angular/router';
import { TcLiveAppsConfigService } from '../../services/tc-live-apps-config.service';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { MatSnackBar } from '@angular/material';
/**
 * Allow Selection of live apps applications (part of config)
 *
 * ![alt-text](../live-apps-settings.png "")
 *
 * \@example <tcla-live-apps-settings></tcla-live-apps-settings>
 */
var LiveAppsSettingsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsSettingsComponent, _super);
    function LiveAppsSettingsComponent(route, liveAppsConfigService, snackBar) {
        var _this = _super.call(this) || this;
        _this.route = route;
        _this.liveAppsConfigService = liveAppsConfigService;
        _this.snackBar = snackBar;
        _this.LIVE_APPS_URL = '/apps/dt-app/index.html#/application-content/';
        _this.handleEditLiveAppClick = (/**
         * @return {?}
         */
        function () {
            // window.open(this.LIVE_APPS_URL + applicationId);
            // currently no public API to get the appId version for design time to populate the URL
            window.open(_this.LIVE_APPS_URL);
        });
        _this.runSaveFunction = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var liveAppsConfig = new LiveAppsConfig().deserialize({
                applicationIds: _this.selectedAppIds,
                caseIconsFolderId: _this.liveAppsConfig.caseIconsFolderId,
                documentAppId: _this.liveAppsConfig.documentAppId,
                collaborationAppId: _this.liveAppsConfig.documentAppId
            });
            _this.liveAppsConfigService.updateLiveAppsConfig(_this.sandboxId, _this.generalConfig.uiAppId, liveAppsConfig, _this.liveAppsConfig.id).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                _this.snackBar.open('Live Apps App Selection Configuration settings saved', 'OK', {
                    duration: 3000
                });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                _this.snackBar.open('Error saving Live Apps App Selection Configuration', 'OK', {
                    duration: 3000
                });
            }));
        });
        return _this;
    }
    /**
     * @protected
     * @return {?}
     */
    LiveAppsSettingsComponent.prototype.getLiveAppsConfigService = /**
     * @protected
     * @return {?}
     */
    function () {
        return this.liveAppsConfigService;
    };
    /**
     * @param {?} appIds
     * @return {?}
     */
    LiveAppsSettingsComponent.prototype.handleAppIdSelection = /**
     * @param {?} appIds
     * @return {?}
     */
    function (appIds) {
        console.log('Selected: ', appIds);
        this.selectedAppIds = appIds;
    };
    /**
     * @return {?}
     */
    LiveAppsSettingsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.generalConfig = this.route.snapshot.data.laConfigHolder.generalConfig;
        this.liveAppsConfig = this.route.snapshot.data.laConfigHolder.liveAppsConfig;
        this.claims = this.route.snapshot.data.claims;
        this.sandboxId = Number(this.claims.primaryProductionSandbox.id).valueOf();
        this.selectedAppIds = this.liveAppsConfig.applicationIds;
    };
    LiveAppsSettingsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-settings',
                    template: "<div fxLayout=\"column\" class=\"tcs-live-apps-settings-pane\" fxFill>\n  <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\" [headerText]=\"'Live Apps App Selection Configuration'\">\n</tc-tibco-cloud-widget-header>\n  <div fxFlex style=\"padding: 20px;\" fxLayout=\"column\">\n    <div fxLayout=\"row\" fxLayoutAlign=\"start start\" fxLayoutGap=\"5px\" style=\"height: 20px\">\n      <p>Choose the TIBCO Cloud Live Apps Applications to be used within this application:</p>\n      <mat-icon class=\"tcs-designer-link\" [svgIcon]=\"'tcs-application-edit'\" (click)=\"handleEditLiveAppClick()\" matTooltip=\"Open Live Apps Designer\" matTooltipShowDelay=\"1000\" matTooltipPosition=\"below\"></mat-icon>\n    </div>\n    <tcla-live-apps-application-list class=\"tcs-app-selection-list-widget\" fxFlex=\"grow\" [sandboxId]=\"sandboxId\" [selectedAppIds]=\"selectedAppIds\" (appIdsSelected)=\"handleAppIdSelection($event)\"></tcla-live-apps-application-list>\n    <div fxFlex class=\"tcs-filler-panel\"></div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n      <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n    </div>\n  </div>\n</div>\n",
                    styles: [".tcs-designer-link{margin-bottom:10px}.tcs-designer-link:hover{fill:#0081cb;cursor:pointer}.tcs-app-selection-list-widget,.tcs-live-apps-settings-pane{overflow:hidden}"]
                }] }
    ];
    /** @nocollapse */
    LiveAppsSettingsComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: TcLiveAppsConfigService },
        { type: MatSnackBar }
    ]; };
    return LiveAppsSettingsComponent;
}(LiveAppsComponent));
export { LiveAppsSettingsComponent };
if (false) {
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.LIVE_APPS_URL;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.sandboxId;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.selectedAppIds;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.liveAppsConfig;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.generalConfig;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.claims;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.handleEditLiveAppClick;
    /** @type {?} */
    LiveAppsSettingsComponent.prototype.runSaveFunction;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsComponent.prototype.route;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsComponent.prototype.liveAppsConfigService;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsSettingsComponent.prototype.snackBar;
}
