// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-active-cases-widget/live-apps-active-cases-widget.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { LiveAppsActiveCasesReportComponent } from '../live-apps-active-cases-report/live-apps-active-cases-report.component';
import { LiveAppsActiveCasesForTypeReportComponent } from '../live-apps-active-cases-for-type-report/live-apps-active-cases-for-type-report.component';
/**
 * Home page active cases widget main component
 *
 * ![alt-text](../live-apps-active-cases-widget.png "Image")
 *
 * \@example <tcla-live-apps-active-cases-widget></tcla-live-apps-active-cases-widget>
 */
export class LiveAppsActiveCasesWidgetComponent extends LiveAppsComponent {
    constructor() {
        super(...arguments);
        /**
         * Maximum rows to show in legend before it is hidden (otherwise would take whole widget)
         */
        this.maxLegendItems = this.maxLegendItems ? this.maxLegendItems : 8;
        /**
         * Whether to show percentages or raw case numbers on the doughnut chart
         */
        this.showPercentages = this.showPercentages ? this.showPercentages : false;
        /**
         * ~event selectedCaseType : Case Type selected in vizualization
         * ~payload CaseTypeReportRecord : CaseTypeReportRecord object selected from component
         */
        this.selectedCaseType = new EventEmitter();
        /**
         * ~event selectedCaseTypeState : Case Type state selected in vizualization
         * ~payload CaseTypeStateReportStateInfo : CaseTypeStateReportStateInfo object selected from component (to drive caller to display something different (drill down))
         */
        this.selectedCaseTypeState = new EventEmitter();
        this.handleCaseTypeSelected = (/**
         * @param {?} selected
         * @return {?}
         */
        (selected) => {
            this.selectedCaseTypeReport = selected;
            this.selectedCaseType.emit(selected);
        });
        this.handleCaseTypeStateSelected = (/**
         * @param {?} selectedState
         * @return {?}
         */
        (selectedState) => {
            this.selectedCaseTypeState.emit(selectedState);
        });
        this.clearSelectedCaseType = (/**
         * @return {?}
         */
        () => {
            this.selectedCaseTypeReport = undefined;
        });
        this.refresh = (/**
         * @return {?}
         */
        () => {
            if (this.activeCasesComp) {
                this.activeCasesComp.refresh();
            }
            if (this.activeCaseTypeComp) {
                this.activeCaseTypeComp.refresh();
            }
        });
    }
}
LiveAppsActiveCasesWidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-active-cases-widget',
                template: "<div class=\"tcs-active-case-widget-pane\" fxFill fxLayout=\"column\">\n  <div *ngIf=\"showHeader\" class=\"tcs-active-case-widget-header\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\">\n    <div fxLayoutAlign=\"start center\">\n      <mat-icon class=\"tcs-icon tcs-active-case-widget-icon\" svgIcon=\"ic-active-case-widget-icon\"></mat-icon>\n      <div class=\"tcs-active-case-widget-header-text\">Active Cases</div>\n    </div>\n    <div fxLayoutGap=\"10px\">\n      <div fxFlex fxLayout=\"row\" fxLayoutAlign=\"start center\">\n        <div class=\"tcs-type-filter-text\" *ngIf=\"selectedCaseTypeReport\" fxLayout=\"row\">\n          [\n          <div class=\"tcs-type-filter-text-highlight\">{{selectedCaseTypeReport.caseTypeInfo.label}}</div>\n          ]\n        </div>\n      </div>\n      <mat-icon *ngIf=\"selectedCaseTypeReport\" (click)=\"clearSelectedCaseType()\" class=\"tcs-icon tcs-icon-active tcs-clear-selected-case-type-icon\" svgIcon=\"tcs-clear-icon\" matTooltip=\"Clear Selected Case Type\" matTooltipPosition=\"left\" matTooltipShowDelay=\"1000\"></mat-icon>\n    </div>\n  </div>\n  <div *ngIf=\"!showHeader\" fxLayout=\"row\" fxLayoutAlign=\"end center\">\n  </div>\n  <tcla-live-apps-active-cases-report *ngIf=\"!selectedCaseTypeReport\" fxFlex [maxLegendItems]=\"maxLegendItems\" [sandboxId]=\"sandboxId\" [appIds]=\"appIds\" [showPercentages]=\"showPercentages\" [uiAppId]=\"uiAppId\" (selectedCaseType)=\"handleCaseTypeSelected($event)\"></tcla-live-apps-active-cases-report>\n  <tcla-live-apps-active-cases-for-type-report *ngIf=\"selectedCaseTypeReport\" fxFlex [maxLegendItems]=\"maxLegendItems\" [sandboxId]=\"sandboxId\" [typeId]=\"selectedCaseTypeReport.caseTypeInfo.id\" [appId]=\"selectedCaseTypeReport.applicationId\" [showPercentages]=\"showPercentages\" [uiAppId]=\"uiAppId\" (selectedCaseTypeState)=\"handleCaseTypeStateSelected($event)\"></tcla-live-apps-active-cases-for-type-report>\n</div>\n",
                styles: [".tcs-active-case-widget-pane{border-radius:3px;box-shadow:0 2px 4px 0 rgba(0,0,0,.33);background-color:#fff}.tcs-active-case-widget-header{height:40px;border-radius:3px 3px 0 0;box-shadow:0 1px 2px 0 #dedede;padding-left:20px;padding-right:20px}.tcs-active-case-widget-header-text{font-family:Source Sans Pro;font-size:18px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;text-align:left;color:#000;margin-left:10px}.tcs-icon.tcs-icon-active:hover{cursor:pointer}:host ::ng-deep .tcs-icon.tcs-icon-active:hover .svg-content{fill:#0081cb}.tcs-type-filter-text{font-family:Source Sans Pro;font-size:12px;font-weight:600;font-style:normal;font-stretch:normal;line-height:1.5;letter-spacing:.3px;text-align:left;color:#000;margin-left:10px}.tcs-type-filter-text-highlight{color:#ff7800}"]
            }] }
];
LiveAppsActiveCasesWidgetComponent.propDecorators = {
    sandboxId: [{ type: Input }],
    appIds: [{ type: Input }],
    showHeader: [{ type: Input }],
    uiAppId: [{ type: Input }],
    maxLegendItems: [{ type: Input }],
    showPercentages: [{ type: Input }],
    selectedCaseType: [{ type: Output }],
    selectedCaseTypeState: [{ type: Output }],
    activeCasesComp: [{ type: ViewChild, args: [LiveAppsActiveCasesReportComponent, { static: false },] }],
    activeCaseTypeComp: [{ type: ViewChild, args: [LiveAppsActiveCasesForTypeReportComponent, { static: false },] }]
};
if (false) {
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.sandboxId;
    /**
     * The list of LA Application IDs you want to handle
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.appIds;
    /**
     * Whether to show the header bar in the widget - eg. favorites on home page (contains icon etc) - if off icons still appear without bar
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.showHeader;
    /**
     * The Application ID of the UI (should ideally be unique as it is shared state key)
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.uiAppId;
    /**
     * Maximum rows to show in legend before it is hidden (otherwise would take whole widget)
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.maxLegendItems;
    /**
     * Whether to show percentages or raw case numbers on the doughnut chart
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.showPercentages;
    /**
     * ~event selectedCaseType : Case Type selected in vizualization
     * ~payload CaseTypeReportRecord : CaseTypeReportRecord object selected from component
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.selectedCaseType;
    /**
     * ~event selectedCaseTypeState : Case Type state selected in vizualization
     * ~payload CaseTypeStateReportStateInfo : CaseTypeStateReportStateInfo object selected from component (to drive caller to display something different (drill down))
     * @type {?}
     */
    LiveAppsActiveCasesWidgetComponent.prototype.selectedCaseTypeState;
    /** @type {?} */
    LiveAppsActiveCasesWidgetComponent.prototype.activeCasesComp;
    /** @type {?} */
    LiveAppsActiveCasesWidgetComponent.prototype.activeCaseTypeComp;
    /** @type {?} */
    LiveAppsActiveCasesWidgetComponent.prototype.selectedCaseTypeReport;
    /** @type {?} */
    LiveAppsActiveCasesWidgetComponent.prototype.handleCaseTypeSelected;
    /** @type {?} */
    LiveAppsActiveCasesWidgetComponent.prototype.handleCaseTypeStateSelected;
    /** @type {?} */
    LiveAppsActiveCasesWidgetComponent.prototype.clearSelectedCaseType;
    /** @type {?} */
    LiveAppsActiveCasesWidgetComponent.prototype.refresh;
}
