// @tibco-tcstk/tc-core-lib/esm5/lib/components/tibco-cloud-table/tibco-cloud-table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild } from '@angular/core';
import { MatPaginator, MatSort } from '@angular/material';
import { TibcoCloudTableDataSource } from './tibco-cloud-table-datasource';
import { TcCoreCommonFunctions } from '../../common/tc-core-common-functions';
var TibcoCloudTableComponent = /** @class */ (function () {
    function TibcoCloudTableComponent() {
        this.id = '_' + Math.random().toString(36).substr(2, 9);
        /**
         * Columns displayed in the table. Columns IDs can be added, removed, or reordered.
         */
        this.displayedColumns = [];
        this.columns = [];
    }
    /**
     * @return {?}
     */
    TibcoCloudTableComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.updateTable();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    TibcoCloudTableComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        console.log('Table Changed');
        this.updateTable();
    };
    /**
     * @return {?}
     */
    TibcoCloudTableComponent.prototype.updateTable = /**
     * @return {?}
     */
    function () {
        // console.log('JSONSource: ', this.jsonSource);
        /** @type {?} */
        var parsedJsonSource = JSON.parse(this.jsonSource);
        // console.log('Parsed JSON Source: ', parsedJsonSource);
        this.dataSource = new TibcoCloudTableDataSource(this.paginator, this.sort, parsedJsonSource);
        /** @type {?} */
        var myColumns = [];
        if (parsedJsonSource.length != null) {
            if (parsedJsonSource.length > 0) {
                // console.log(parsedJsonSource[0]);
                if (parsedJsonSource[0] != null) {
                    /** @type {?} */
                    var myColumns_1 = [];
                    /** @type {?} */
                    var m = 0;
                    var _loop_1 = function (headerName) {
                        if (parsedJsonSource[0].hasOwnProperty(headerName)) {
                            // console.log(m + ' headerName:', headerName);
                            this_1.displayedColumns.push(headerName);
                            myColumns_1[m] = {};
                            myColumns_1[m]['columnDef'] = headerName;
                            myColumns_1[m]['header'] = headerName;
                            myColumns_1[m].cell = (/**
                             * @param {?} element
                             * @return {?}
                             */
                            function (element) {
                                return "" + element[headerName];
                            });
                            this_1.columns = myColumns_1;
                        }
                        m++;
                    };
                    var this_1 = this;
                    for (var headerName in parsedJsonSource[0]) {
                        _loop_1(headerName);
                    }
                }
                // console.log('myColums:', this.columns);
                // this.columDefArray = this.columns;
            }
        }
    };
    /*
    private log(...m){
      console.log('TIBCO CLOUD TABLE] ' , m);
    }*/
    /*
      private log(...m){
        console.log('TIBCO CLOUD TABLE] ' , m);
      }*/
    /**
     * @param {?} header
     * @return {?}
     */
    TibcoCloudTableComponent.prototype.camelCaseTW = /*
      private log(...m){
        console.log('TIBCO CLOUD TABLE] ' , m);
      }*/
    /**
     * @param {?} header
     * @return {?}
     */
    function (header) {
        return TcCoreCommonFunctions.camelCaseToWords(header);
    };
    TibcoCloudTableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tc-tibco-cloud-table',
                    template: "<div class=\"mat-elevation-z8\">\n  <table id=\"{{id}}\" mat-table class=\"tct-full-width-table\" [dataSource]=\"dataSource\" matSort aria-label=\"Elements\">\n    <ng-container *ngFor=\"let column of columns\" [matColumnDef]='column.columnDef'>\n      <mat-header-cell *matHeaderCellDef class=\"tct-header\"  mat-sort-header>{{camelCaseTW(column.header)}}</mat-header-cell>\n      <mat-cell *matCellDef=\"let row\">{{ column.cell(row) }}</mat-cell>\n    </ng-container>\n    <mat-header-row *matHeaderRowDef=\"displayedColumns\"  ></mat-header-row>\n    <mat-row *matRowDef=\"let row; columns: displayedColumns;\n                         let even = even;\n                         columns: displayedColumns;\"\n             [ngClass]=\"{gray: even}\"></mat-row>\n  </table>\n  <mat-paginator [ngClass]=\"{'tct-hide': dataSource.data.length < 10}\" #paginator\n                 [length]=\"dataSource.data.length\"\n                 [pageIndex]=\"0\"\n                 [pageSize]=\"10\"\n                 [pageSizeOptions]=\"[5, 10, 20, 25]\">\n  </mat-paginator>\n</div>\n",
                    styles: [".tct-full-width-table{width:100%}.tct-header{background:#0081cb;color:#fff;padding:5px 30px;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center}.tct-header:hover{background-color:#229be0}.tctgray{background-color:#f5f5f5}.tct-hide{display:none}"]
                }] }
    ];
    TibcoCloudTableComponent.propDecorators = {
        jsonSource: [{ type: Input }],
        paginator: [{ type: ViewChild, args: [MatPaginator, { static: true },] }],
        sort: [{ type: ViewChild, args: [MatSort, { static: true },] }]
    };
    return TibcoCloudTableComponent;
}());
export { TibcoCloudTableComponent };
if (false) {
    /** @type {?} */
    TibcoCloudTableComponent.prototype.jsonSource;
    /** @type {?} */
    TibcoCloudTableComponent.prototype.paginator;
    /** @type {?} */
    TibcoCloudTableComponent.prototype.sort;
    /** @type {?} */
    TibcoCloudTableComponent.prototype.dataSource;
    /** @type {?} */
    TibcoCloudTableComponent.prototype.id;
    /**
     * Columns displayed in the table. Columns IDs can be added, removed, or reordered.
     * @type {?}
     */
    TibcoCloudTableComponent.prototype.displayedColumns;
    /** @type {?} */
    TibcoCloudTableComponent.prototype.columns;
}
