// @tibco-tcstk/tc-core-lib/esm5/lib/guards/auth.guard.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// This guard will redirect to login when not authenticated against live apps.
// If hosted on Tibco cloud it will not be used since Tibco Cloud/Live Apps WRP resources are protected anyway.
// session is detected if API called in last 30 mins (checks local sessionTimestamp)
// alternative way to achieve this would be to make an API call - eg) live apps claims call
// NOTE: assumes the login route is /login
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
var AuthGuard = /** @class */ (function () {
    function AuthGuard(router) {
        this.router = router;
        this.TIBCO_CLOUD_DOMAIN = 'cloud.tibco.com';
        this.TIBCO_TEST_DOMAIN = 'tenant-integration.tcie.pro';
        this.TIBCO_DEV_DOMAIN = 'emea.tibco.com';
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuard.prototype.canActivate = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        // check if we are hosted on tibco.cloud.com
        /** @type {?} */
        var host = window.location.hostname.split('.');
        /** @type {?} */
        var hostDomain = host[host.length - 3] + '.' + host[host.length - 2] + '.' + host[host.length - 1];
        if (hostDomain === this.TIBCO_CLOUD_DOMAIN || hostDomain === this.TIBCO_TEST_DOMAIN || hostDomain === this.TIBCO_DEV_DOMAIN) {
            // delegate handling login/auth to Tibco Cloud since WRP resources are protected anyway
            return true;
        }
        else {
            // use the sessionTimestamp to decide whether to redirect to login (30 mins expiry of token if no API call)
            /** @type {?} */
            var tcsTimestamp = sessionStorage.getItem('tcsTimestamp');
            if (tcsTimestamp && Number(tcsTimestamp) > (Number(Date.now()) - (30 * 60000))) {
                // logged in and api called made in last 30 mins so should be OK
                return true;
            }
            else {
                // not logged in so redirect to login page
                this.router.navigate(['/login'], { queryParams: { returnUrl: state.url } });
            }
        }
    };
    AuthGuard.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AuthGuard.ctorParameters = function () { return [
        { type: Router }
    ]; };
    return AuthGuard;
}());
export { AuthGuard };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthGuard.prototype.TIBCO_CLOUD_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthGuard.prototype.TIBCO_TEST_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthGuard.prototype.TIBCO_DEV_DOMAIN;
    /**
     * @type {?}
     * @private
     */
    AuthGuard.prototype.router;
}
