// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/services/tc-roles-service.ts.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class TcRolesService {
    constructor() {
        this.currentRole = undefined;
        this.hasRoleAccess = (/**
         * @param {?} roleReq
         * @param {?} roles
         * @return {?}
         */
        (roleReq, roles) => {
            /** @type {?} */
            const reqRole = roles.roles.find((/**
             * @param {?} role
             * @return {?}
             */
            role => {
                return (role.id === roleReq);
            }));
            return reqRole ? true : false;
        });
        this.hasButtonAccess = (/**
         * @param {?} buttonId
         * @param {?} roles
         * @param {?} access
         * @return {?}
         */
        (buttonId, roles, access) => {
            // check which role required
            /** @type {?} */
            const reqRole = access.buttons.find((/**
             * @param {?} buttonRec
             * @return {?}
             */
            buttonRec => {
                return (buttonId === buttonRec.buttonId);
            }));
            if (!reqRole) {
                // nothing defined so return true;
                return true;
            }
            else {
                /** @type {?} */
                const hasRole = roles.roles.find((/**
                 * @param {?} role
                 * @return {?}
                 */
                role => {
                    return (role.id === reqRole.requiredRoleId);
                }));
                return hasRole ? true : false;
            }
        });
        this.amIConfigurator = (/**
         * @param {?} roles
         * @return {?}
         */
        (roles) => {
            /** @type {?} */
            const reqRole = roles.roles.find((/**
             * @param {?} role
             * @return {?}
             */
            role => {
                return (role.configuration === true);
            }));
            return reqRole ? true : false;
        });
        this.getCurrentRole = (/**
         * @return {?}
         */
        () => {
            return this.currentRole;
        });
        this.setCurrentRole = (/**
         * @param {?} role
         * @return {?}
         */
        (role) => {
            this.currentRole = role;
        });
    }
    /**
     * @param {?} roleId
     * @param {?} roles
     * @return {?}
     */
    checkRole(roleId, roles) {
        return this.hasRoleAccess(roleId, roles);
    }
    /**
     * @param {?} buttonId
     * @param {?} roles
     * @param {?} access
     * @return {?}
     */
    checkButton(buttonId, roles, access) {
        return this.hasButtonAccess(buttonId, roles, access);
    }
}
TcRolesService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TcRolesService.ctorParameters = () => [];
/** @nocollapse */ TcRolesService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcRolesService_Factory() { return new TcRolesService(); }, token: TcRolesService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    TcRolesService.prototype.currentRole;
    /**
     * @type {?}
     * @private
     */
    TcRolesService.prototype.hasRoleAccess;
    /**
     * @type {?}
     * @private
     */
    TcRolesService.prototype.hasButtonAccess;
    /** @type {?} */
    TcRolesService.prototype.amIConfigurator;
    /** @type {?} */
    TcRolesService.prototype.getCurrentRole;
    /** @type {?} */
    TcRolesService.prototype.setCurrentRole;
}
