// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-settings-recent-cases/live-apps-settings-recent-cases.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { LiveAppsConfig } from '../../models/tc-liveapps-config';
import { LiveAppsSettingsComponent } from '../live-apps-settings/live-apps-settings.component';
/**
 * Allow Selection of live apps applications (part of config)
 *
 * ![alt-text](../live-apps-settings-recent-cases.png "")
 *
 * \@example <tcla-live-apps-settings-recent-cases></tcla-live-apps-settings-recent-cases>
 */
var LiveAppsSettingsRecentCasesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LiveAppsSettingsRecentCasesComponent, _super);
    function LiveAppsSettingsRecentCasesComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.LIVE_APPS_URL = '/apps/dt-app/index.html#/application-content/';
        _this.runSaveFunction = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var liveAppsConfig = new LiveAppsConfig().deserialize({
                applicationIds: _this.selectedAppIds,
                recentExcludedAppIds: _this.excludeRecentIds,
                caseIconsFolderId: _this.liveAppsConfig.caseIconsFolderId,
                documentAppId: _this.liveAppsConfig.documentAppId,
                collaborationAppId: _this.liveAppsConfig.documentAppId
            });
            // unable to use super.liveAppsConfigService
            _super.prototype.getLiveAppsConfigService.call(_this).updateLiveAppsConfig(_this.sandboxId, _this.generalConfig.uiAppId, liveAppsConfig, _this.liveAppsConfig.id).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                _this.snackBar.open('Recent cases settings saved', 'OK', {
                    duration: 3000
                });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                _this.snackBar.open('Error saving Recent cases settings', 'OK', {
                    duration: 3000
                });
            }));
        });
        return _this;
    }
    /**
     * @param {?} appIds
     * @return {?}
     */
    LiveAppsSettingsRecentCasesComponent.prototype.handleExcludeRecentAppIdSelection = /**
     * @param {?} appIds
     * @return {?}
     */
    function (appIds) {
        this.excludeRecentIds = appIds;
    };
    /**
     * @return {?}
     */
    LiveAppsSettingsRecentCasesComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.excludeRecentIds = this.liveAppsConfig.recentExcludedAppIds;
    };
    LiveAppsSettingsRecentCasesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-settings-recent-cases',
                    template: "<div fxLayout=\"column\" class=\"tcs-live-apps-settings-pane\" fxFill>\n  <tc-tibco-cloud-widget-header style=\"height: 40px;\" [icon]=\"'tcs-capabilities'\" [headerText]=\"'Live Apps Recent Cases Configuration'\">\n</tc-tibco-cloud-widget-header>\n  <div fxFlex style=\"padding: 20px;\" fxLayout=\"column\" fxFill>\n    <div fxLayout=\"row\" fxLayoutAlign=\"start start\" fxLayoutGap=\"5px\" style=\"height: 20px\">\n      <p>Select any TIBCO Cloud Live Apps Applications you wish to EXCLUDE from recent cases:</p>\n      <mat-icon class=\"tcs-designer-link\" [svgIcon]=\"'tcs-application-edit'\" (click)=\"handleEditLiveAppClick()\" matTooltip=\"Open Live Apps Designer\" matTooltipShowDelay=\"1000\" matTooltipPosition=\"below\"></mat-icon>\n    </div>\n    <tcla-live-apps-application-list class=\"tcs-app-selection-list-widget\" fxFlex=\"grow\" [sandboxId]=\"sandboxId\" [selectedAppIds]=\"excludeRecentIds\" (appIdsSelected)=\"handleExcludeRecentAppIdSelection($event)\"></tcla-live-apps-application-list>\n    <div fxFlex class=\"tcs-filler-panel\"></div>\n    <div fxLayout=\"column\" fxLayoutAlign=\"end end\" style=\"min-height: 50px\">\n      <button mat-raised-button color=\"primary\" (click)=\"runSaveFunction()\">Save</button>\n    </div>\n  </div>\n</div>\n",
                    styles: [".tcs-designer-link{margin-bottom:10px}.tcs-designer-link:hover{fill:#0081cb;cursor:pointer}.tcs-app-selection-list-widget,.tcs-live-apps-settings-pane{overflow:hidden}"]
                }] }
    ];
    return LiveAppsSettingsRecentCasesComponent;
}(LiveAppsSettingsComponent));
export { LiveAppsSettingsRecentCasesComponent };
if (false) {
    /** @type {?} */
    LiveAppsSettingsRecentCasesComponent.prototype.LIVE_APPS_URL;
    /** @type {?} */
    LiveAppsSettingsRecentCasesComponent.prototype.excludeRecentIds;
    /** @type {?} */
    LiveAppsSettingsRecentCasesComponent.prototype.runSaveFunction;
}
