// @tibco-tcstk/tc-liveapps-lib/esm5/lib/guards/role.guard.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// This guard is to check whether the user has appropriate role to access a route based on config settings
import { Injectable } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { LiveAppsService } from '../services/live-apps.service';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { forkJoin } from 'rxjs';
import { TcCoreCommonFunctions, TcGeneralConfigService, TcSharedStateService } from '@tibco-tcstk/tc-core-lib';
import { Location } from '@angular/common';
import { TcRolesService } from '../services/tc-roles-service.ts.service';
import { TcAccessControlService } from '../services/tc-access-control.service';
import { AccessControlConfigurationResolver } from '../resolvers/accessControlConfiguration.resolver';
import { AccessResolver } from '../resolvers/access.resolver';
var RoleGuard = /** @class */ (function () {
    function RoleGuard(liveapps, rolesService, http, location, route, sharedStateService, generalConfigService, accessControlService, router) {
        this.liveapps = liveapps;
        this.rolesService = rolesService;
        this.http = http;
        this.location = location;
        this.route = route;
        this.sharedStateService = sharedStateService;
        this.generalConfigService = generalConfigService;
        this.accessControlService = accessControlService;
        this.router = router;
        this.DEFAULT_CONFIG_URL = TcCoreCommonFunctions.prepareUrlForStaticResource(this.location, 'assets/config/routeAccessControl.json');
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    RoleGuard.prototype.canActivate = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        /** @type {?} */
        var guardRoute = '';
        route.pathFromRoot.map((/**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            if (element.url.length === 1) {
                guardRoute += '/' + element.url[0];
            }
        }));
        // we will need the active user role
        /** @type {?} */
        var activeResolver$ = new AccessResolver(this.location, this.http, this.accessControlService, this.rolesService, this.liveapps, this.route, this.sharedStateService, this.generalConfigService, this.router).resolve();
        // access control configuration to check current URL is on allowedRoutes
        /** @type {?} */
        var accessControlConfig$ = new AccessControlConfigurationResolver(this.location, this.http, this.accessControlService).resolve();
        // run both in parallel then check access
        return forkJoin(activeResolver$, accessControlConfig$).pipe(map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), activeResolver = _b[0], accessControlConfig = _b[1];
            // if guardRoute is not in allowedRoutes then OK
            if (accessControlConfig.allowedRoutes.indexOf(guardRoute) === -1) {
                return true;
            }
            // Check guardRoute in active user role routes
            if (activeResolver.routes.indexOf(guardRoute) > -1) {
                return true;
            }
            else {
                return false;
            }
        })));
    };
    RoleGuard.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    RoleGuard.ctorParameters = function () { return [
        { type: LiveAppsService },
        { type: TcRolesService },
        { type: HttpClient },
        { type: Location },
        { type: ActivatedRoute },
        { type: TcSharedStateService },
        { type: TcGeneralConfigService },
        { type: TcAccessControlService },
        { type: Router }
    ]; };
    return RoleGuard;
}());
export { RoleGuard };
if (false) {
    /** @type {?} */
    RoleGuard.prototype.DEFAULT_CONFIG_URL;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.liveapps;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.rolesService;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.http;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.location;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.route;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.sharedStateService;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.generalConfigService;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.accessControlService;
    /**
     * @type {?}
     * @private
     */
    RoleGuard.prototype.router;
}
