// @tibco-tcstk/tc-liveapps-lib/esm2015/lib/components/live-apps-legacy-process/live-apps-legacy-process.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { LiveAppsComponent } from '../live-apps-component/live-apps-component.component';
import { LegacyIframeService } from '@tibco-tcstk/tc-core-lib';
import { TcVisibilityService } from '@tibco-tcstk/tc-core-lib';
import { LaProcessSelection } from '../../models/tc-case-processes';
/**
 * Wraps legacy angularjs form renderer
 *
 * \@example <tcla-live-apps-legacy-form></tcla-live-apps-legacy-form>
 */
export class LiveAppsLegacyProcessComponent extends LiveAppsComponent {
    /**
     * @param {?} visibilityService
     * @param {?} host
     * @param {?} legacyIframeService
     */
    constructor(visibilityService, host, legacyIframeService) {
        super();
        this.visibilityService = visibilityService;
        this.host = host;
        this.legacyIframeService = legacyIframeService;
        this.legacyIframeId = this.legacyIframeId ? this.legacyIframeId : 'legacyProcessFrame';
        this.processComplete = new EventEmitter();
        this.processCancelled = new EventEmitter();
        this.processActive = false;
        this.changeProcess = (/**
         * @param {?} process
         * @return {?}
         */
        (process) => {
            // process has changed
            this.renderProcess(process);
        });
        this.renderProcess = (/**
         * @param {?} process
         * @return {?}
         */
        (process) => {
            if (this.processActive && this.openProcessId) {
                // opening a new process so cancel old one
                this.cancelProcess(false);
            }
            // send message to external form app
            this.formDiv.contentWindow.postMessage({
                action: this.type === 'creator' ? 'runCreator' : 'runAction',
                processId: process.process.id,
                caseRef: process.caseReference,
                caseState: this.caseState,
                applicationId: this.applicationId,
                typeId: this.typeId
            }, window.location.origin);
            // position the form iframe over the processDiv placeholder
            this.processActive = true;
            this.openProcessId = process.process.id;
            this.resizeProcess();
        });
        this.receiveMessage = (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            if (event.data.action === 'processComplete') {
                this.processActive = false;
                this.openProcessId = undefined;
                this.hideProcess();
                this.processComplete.emit();
            }
            else if (event.data.action === 'processCancelled') {
                this.processActive = false;
                this.openProcessId = undefined;
                this.hideProcess();
                this.processCancelled.emit();
            }
        });
        this.cancelProcess = (/**
         * @param {?=} emit
         * @return {?}
         */
        (emit) => {
            this.formDiv.contentWindow.postMessage({ 'action': 'cancelProcess' }, window.location.origin);
            this.processActive = false;
            this.openProcessId = undefined;
            this.hideProcess();
            if (emit) {
                this.processCancelled.emit();
            }
        });
        this.hideProcess = (/**
         * @return {?}
         */
        () => {
            // hide the iframe!
            this.formDiv.style.top = '-1000px';
            this.formDiv.style.left = '-1000px';
            this.formDiv.style.zIndex = '1';
        });
        this.resizeProcess = (/**
         * @return {?}
         */
        () => {
            // position the form iframe over the processDiv placeholder
            setTimeout((/**
             * @param {?} handler
             * @return {?}
             */
            handler => {
                /** @type {?} */
                const rect = this.target.getBoundingClientRect();
                this.formDiv.style.top = rect.top.toString() + 'px';
                this.formDiv.style.left = rect.left.toString() + 'px';
                this.formDiv.style.height = rect.height.toString() + 'px';
                this.formDiv.style.width = rect.width.toString() + 'px';
                this.formDiv.style.position = 'absolute';
                // creator needs 1001 so it is over the dialog box
                this.formDiv.style.zIndex = (this.type === 'creator') ? '1001' : '1000';
            }), 1000);
        });
        this.initialize = (/**
         * @return {?}
         */
        () => {
            // listen for messages from the iframe
            window.addEventListener('message', this.receiveMessage, false);
            // we only want to render the process when the element is visible
            /** @type {?} */
            const inSight$ = this.visibilityService.elementInSight(this.host);
            inSight$.subscribe((/**
             * @param {?} next
             * @return {?}
             */
            next => {
                if (!next) {
                    // element not visible so hide the process if it is showing
                    if (this.processActive && this.target) {
                        this.hideProcess();
                    }
                }
                else {
                    // element is visible
                    if (!this.processActive) {
                        // if process not already showing - show it
                        // first get a handle on the target div and the legacy iframe
                        this.target = document.getElementById('processDiv');
                        this.formDiv = document.getElementById(this.legacyIframeId);
                        if (this.openProcessId) {
                            // trigger display of the process
                            this.renderProcess(this.process);
                        }
                    }
                    else {
                        this.resizeProcess();
                    }
                }
            }));
        });
    }
    /**
     * @return {?}
     */
    isFormActive() {
        return this.processActive;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this.openProcessId = this.process.process.id;
        this.containerChanges$.subscribe((/**
         * @param {?} widget
         * @return {?}
         */
        widget => {
            if (this.processActive && this.target) {
                this.resizeProcess();
            }
        }));
        // wait for iframe to load before doing anything else
        // this observable will emit true once the iFrame is loaded
        this.legacyIframeService.processStatus.subscribe((/**
         * @param {?} loaded
         * @return {?}
         */
        loaded => {
            if (loaded) {
                // iframe is loaded
                this.initialize();
            }
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.openProcessId) {
            this.cancelProcess();
        }
    }
}
LiveAppsLegacyProcessComponent.decorators = [
    { type: Component, args: [{
                selector: 'tcla-live-apps-legacy-process',
                template: "<div #componentDiv fxFill id=\"processDiv\" style=\"border: none;\"></div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
LiveAppsLegacyProcessComponent.ctorParameters = () => [
    { type: TcVisibilityService },
    { type: ElementRef },
    { type: LegacyIframeService }
];
LiveAppsLegacyProcessComponent.propDecorators = {
    legacyIframeId: [{ type: Input }],
    process: [{ type: Input }],
    type: [{ type: Input }],
    applicationId: [{ type: Input }],
    typeId: [{ type: Input }],
    caseRef: [{ type: Input }],
    caseState: [{ type: Input }],
    processComplete: [{ type: Output }],
    processCancelled: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.legacyIframeId;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.process;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.type;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.applicationId;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.typeId;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.caseRef;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.caseState;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.processComplete;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.processCancelled;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.target;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.formDiv;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.processActive;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.openProcessId;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.changeProcess;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.renderProcess;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.receiveMessage;
    /** @type {?} */
    LiveAppsLegacyProcessComponent.prototype.cancelProcess;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.hideProcess;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.resizeProcess;
    /**
     * @type {?}
     * @private
     */
    LiveAppsLegacyProcessComponent.prototype.initialize;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsLegacyProcessComponent.prototype.visibilityService;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsLegacyProcessComponent.prototype.host;
    /**
     * @type {?}
     * @protected
     */
    LiveAppsLegacyProcessComponent.prototype.legacyIframeService;
}
