// @tibco-tcstk/tc-core-lib/esm5/lib/services/legacy-iframe.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @ngdoc component
 * @name LegacyIframeService
 *
 * @description
 *
 * Used to handle synchronization and loading of legacy form iframe
 *
 * Will emit false until the iFrame has posted it's ready message
 *
 *
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
var LegacyIframeService = /** @class */ (function () {
    function LegacyIframeService() {
        var _this = this;
        this.iWorkitemFrameReady = new BehaviorSubject(false);
        this.iProcessFrameReady = new BehaviorSubject(false);
        this.workitemStatus = this.iWorkitemFrameReady.asObservable();
        this.processStatus = this.iProcessFrameReady.asObservable();
        this.receiveMessage = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (event.data.action === 'workitemAppReady') {
                _this.iWorkitemFrameReady.next(true);
                console.log('*** workitems iframe ready');
            }
            else if (event.data.action === 'processAppReady') {
                _this.iProcessFrameReady.next(true);
                console.log('*** process iframe ready');
            }
        });
        console.log('*** service waiting for message');
        window.addEventListener('message', this.receiveMessage, false);
    }
    LegacyIframeService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    LegacyIframeService.ctorParameters = function () { return []; };
    /** @nocollapse */ LegacyIframeService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function LegacyIframeService_Factory() { return new LegacyIframeService(); }, token: LegacyIframeService, providedIn: "root" });
    return LegacyIframeService;
}());
export { LegacyIframeService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LegacyIframeService.prototype.iWorkitemFrameReady;
    /**
     * @type {?}
     * @private
     */
    LegacyIframeService.prototype.iProcessFrameReady;
    /** @type {?} */
    LegacyIframeService.prototype.workitemStatus;
    /** @type {?} */
    LegacyIframeService.prototype.processStatus;
    /**
     * @type {?}
     * @private
     */
    LegacyIframeService.prototype.receiveMessage;
}
