// @tibco-tcstk/tc-core-lib/esm5/lib/services/tc-buttons-helper.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ToolbarButton } from '../models/tc-widget-header';
import * as i0 from "@angular/core";
var TcButtonsHelperService = /** @class */ (function () {
    function TcButtonsHelperService() {
        this.createButton = (/**
         * @param {?} id
         * @param {?} icon
         * @param {?} filled
         * @param {?} tooltip
         * @param {?} visible
         * @param {?} enabled
         * @return {?}
         */
        function (id, icon, filled, tooltip, visible, enabled) {
            /** @type {?} */
            var newButton = new ToolbarButton().deserialize({
                id: id,
                icon: icon,
                filled: filled,
                tooltip: tooltip,
                visible: visible,
                enabled: enabled
            });
            return newButton;
        });
        this.updateButtons = (/**
         * @param {?} updatedToolbarButtons
         * @param {?} toolbarButtons
         * @return {?}
         */
        function (updatedToolbarButtons, toolbarButtons) {
            updatedToolbarButtons.forEach((/**
             * @param {?} updatedButton
             * @return {?}
             */
            function (updatedButton) {
                /** @type {?} */
                var idx = toolbarButtons.findIndex((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item.id === updatedButton.id; }));
                if (idx > -1) {
                    toolbarButtons[idx] = updatedButton;
                }
                else {
                    // no existing button so add it
                    toolbarButtons.push(updatedButton);
                }
            }));
            return toolbarButtons;
        });
        this.addButtons = (/**
         * @param {?} buttons
         * @param {?} toolbarButtons
         * @return {?}
         */
        function (buttons, toolbarButtons) {
            buttons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            function (button) {
                toolbarButtons.push(button);
            }));
            return toolbarButtons;
        });
        this.removeButtons = (/**
         * @param {?} buttons
         * @param {?} toolbarButtons
         * @return {?}
         */
        function (buttons, toolbarButtons) {
            buttons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            function (button) {
                /** @type {?} */
                var idx = toolbarButtons.findIndex((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item.id === button.id; }));
                toolbarButtons.splice(idx, 1);
            }));
            return toolbarButtons;
        });
    }
    TcButtonsHelperService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TcButtonsHelperService.ctorParameters = function () { return []; };
    /** @nocollapse */ TcButtonsHelperService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function TcButtonsHelperService_Factory() { return new TcButtonsHelperService(); }, token: TcButtonsHelperService, providedIn: "root" });
    return TcButtonsHelperService;
}());
export { TcButtonsHelperService };
if (false) {
    /** @type {?} */
    TcButtonsHelperService.prototype.createButton;
    /** @type {?} */
    TcButtonsHelperService.prototype.updateButtons;
    /** @type {?} */
    TcButtonsHelperService.prototype.addButtons;
    /** @type {?} */
    TcButtonsHelperService.prototype.removeButtons;
}
