// @tibco-tcstk/tc-core-lib/esm2015/lib/resolvers/configuration-menu-config.resolver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/* Used to resolve contents of configuration menu config file */
import { Injectable } from '@angular/core';
import { UiAppIdConfig } from '../models/tc-app-config';
import { map, flatMap } from 'rxjs/operators';
import { HttpClient } from '@angular/common/http';
import { ConfigurationMenuConfig } from '../models/tc-configuration-menu-config';
import { Location } from '@angular/common';
import { TcCoreCommonFunctions } from '../common/tc-core-common-functions';
export class ConfigurationMenuConfigResolver {
    /**
     * @param {?} http
     * @param {?} location
     */
    constructor(http, location) {
        this.http = http;
        this.location = location;
        // DEFAULT_CONFIG_URL = 'assets/config/<uiAppId>/configurationMenuConfig.json';
        // JS: Changed since we will use different GIT repo for different templates from now on
        this.DEFAULT_CONFIG_URL = 'assets/config/configurationMenuConfig.json';
        this.APP_ID_URL = TcCoreCommonFunctions.prepareUrlForStaticResource(this.location, 'assets/config/uiAppId.json');
        // can be used to load defaultAppConfig from a JSON config
        this.getConfigurationMenuConfig = (/**
         * @param {?} uiAppId
         * @return {?}
         */
        (uiAppId) => {
            return this.http.get(TcCoreCommonFunctions.prepareUrlForStaticResource(this.location, this.DEFAULT_CONFIG_URL.replace('<uiAppId>', uiAppId)));
        });
        // loads uiAppId from json file in assets (appId.json)
        this.getAppId = (/**
         * @return {?}
         */
        () => {
            return this.http.get(this.APP_ID_URL).pipe(map((/**
             * @param {?} uiAppId
             * @return {?}
             */
            uiAppId => {
                /** @type {?} */
                const uiAppIdConfig = new UiAppIdConfig().deserialize(uiAppId);
                return uiAppIdConfig;
            })));
        });
    }
    /**
     * @param {?} routeSnapshot
     * @return {?}
     */
    resolve(routeSnapshot) {
        /** @type {?} */
        const configurationMenuConfig$ = this.getAppId().pipe(flatMap((/**
         * @param {?} uiAppId
         * @return {?}
         */
        uiAppId => {
            return this.getConfigurationMenuConfig(uiAppId.uiAppId).pipe(map((/**
             * @param {?} config
             * @return {?}
             */
            config => {
                return new ConfigurationMenuConfig().deserialize(config);
            })));
        })));
        return configurationMenuConfig$;
    }
}
ConfigurationMenuConfigResolver.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ConfigurationMenuConfigResolver.ctorParameters = () => [
    { type: HttpClient },
    { type: Location }
];
if (false) {
    /** @type {?} */
    ConfigurationMenuConfigResolver.prototype.DEFAULT_CONFIG_URL;
    /** @type {?} */
    ConfigurationMenuConfigResolver.prototype.APP_ID_URL;
    /**
     * @type {?}
     * @private
     */
    ConfigurationMenuConfigResolver.prototype.getConfigurationMenuConfig;
    /**
     * @type {?}
     * @private
     */
    ConfigurationMenuConfigResolver.prototype.getAppId;
    /**
     * @type {?}
     * @private
     */
    ConfigurationMenuConfigResolver.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ConfigurationMenuConfigResolver.prototype.location;
}
