// angular6-json-schema-form/esm5/lib/widget-library/widget-library.module.js
import * as tslib_1 from "tslib";
import { BASIC_WIDGETS } from './index';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { JsonSchemaFormService } from '../json-schema-form.service';
import { NgModule } from '@angular/core';
import { OrderableDirective } from './orderable.directive';
var WidgetLibraryModule = /** @class */ (function () {
    function WidgetLibraryModule() {
    }
    WidgetLibraryModule_1 = WidgetLibraryModule;
    WidgetLibraryModule.forRoot = function () {
        return {
            ngModule: WidgetLibraryModule_1,
            providers: [JsonSchemaFormService]
        };
    };
    var WidgetLibraryModule_1;
    WidgetLibraryModule = WidgetLibraryModule_1 = tslib_1.__decorate([
        NgModule({
            imports: [CommonModule, FormsModule, ReactiveFormsModule],
            declarations: tslib_1.__spread(BASIC_WIDGETS, [OrderableDirective]),
            exports: tslib_1.__spread(BASIC_WIDGETS, [OrderableDirective]),
            entryComponents: tslib_1.__spread(BASIC_WIDGETS),
            providers: [JsonSchemaFormService]
        })
    ], WidgetLibraryModule);
    return WidgetLibraryModule;
}());
export { WidgetLibraryModule };
