// @tibco-tcstk/tc-liveapps-lib/esm5/lib/components/live-apps-document-viewer/live-apps-document-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { Document } from '../../models/tc-document';
import { TcDocumentService } from '../../services/tc-document.service';
import { TcCoreCommonFunctions } from '@tibco-tcstk/tc-core-lib';
import { Location } from '@angular/common';
/**
 * Display document
 *
 * \@example <tcla-live-apps-document-viewer></tcla-live-apps-document-viewer>
 */
var LiveAppsDocumentViewerComponent = /** @class */ (function () {
    function LiveAppsDocumentViewerComponent(sanitizer, documentsService, location) {
        this.sanitizer = sanitizer;
        this.documentsService = documentsService;
        this.location = location;
    }
    /**
     * @return {?}
     */
    LiveAppsDocumentViewerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var preparedUrl = TcCoreCommonFunctions.prepareUrlForNonStaticResource(this.location, this.documentsService.getUrlForDocument(this.folderType, this.folderId, this.document.name, this.document.artifactVersion, this.sandboxId));
        this.sanitizedResourceUrl = this.sanitizer.bypassSecurityTrustResourceUrl(preparedUrl);
    };
    LiveAppsDocumentViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'tcla-live-apps-document-viewer',
                    template: "<div fxFill>\n  <iframe name=\"DocViewerFrame\" class=\"md-dialog-content\" fxFill [src]=\"sanitizedResourceUrl\">\n  </iframe>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    LiveAppsDocumentViewerComponent.ctorParameters = function () { return [
        { type: DomSanitizer },
        { type: TcDocumentService },
        { type: Location }
    ]; };
    LiveAppsDocumentViewerComponent.propDecorators = {
        document: [{ type: Input }],
        folderType: [{ type: Input }],
        folderId: [{ type: Input }],
        sandboxId: [{ type: Input }]
    };
    return LiveAppsDocumentViewerComponent;
}());
export { LiveAppsDocumentViewerComponent };
if (false) {
    /**
     * Document metadata object from API - describes the document to display
     * @type {?}
     */
    LiveAppsDocumentViewerComponent.prototype.document;
    /**
     * orgFolders' or 'caseFolders' - different API calls made according to which one this is
     * @type {?}
     */
    LiveAppsDocumentViewerComponent.prototype.folderType;
    /**
     * The organisation folder to store/retrieve documents
     * @type {?}
     */
    LiveAppsDocumentViewerComponent.prototype.folderId;
    /**
     * sandboxId - this comes from claims resolver
     * @type {?}
     */
    LiveAppsDocumentViewerComponent.prototype.sandboxId;
    /** @type {?} */
    LiveAppsDocumentViewerComponent.prototype.sanitizedResourceUrl;
    /**
     * @type {?}
     * @private
     */
    LiveAppsDocumentViewerComponent.prototype.sanitizer;
    /**
     * @type {?}
     * @private
     */
    LiveAppsDocumentViewerComponent.prototype.documentsService;
    /**
     * @type {?}
     * @private
     */
    LiveAppsDocumentViewerComponent.prototype.location;
}
